#install.packages('caper')
library(caper)
###############################################################################
# 1) Example: create data frames (df1 and df2) from the two tables you posted
#    In practice, you might read these from files or have them as data frames
#    already. For simplicity, the code below includes only columns needed:
###############################################################################
df1 <- data.frame(
  Lineage = c("rfXP_032990674_1","paXP_024906169_1","pgXP_039732085_1",
              "pvXP_023386510_1","raXP_016009779_1","sbXP_066123222_1",
              "drXP_053766978_1","ajXP_037023005_2","phXP_045709102_1",
              "pdXP_028375203_1","moXP_036137889_1","mbXP_014389106_1",
              "mlXP_006095087_1","mdXP_059551924_1","mmXP_036168168_1",
              "pkXP_045441875_1","pnCAK6445218_1"),
  dN_dS   = c(0.038,0.033,0.033,0.030,0.033,0.030,0.075,0.077,
              0.118,0.118,0.035,0.335,0.111,0.081,0.301,0.092,0.052)
)

df2 <- data.frame(
  Lineage = c("rfXP_032944932_1","paXP_015445049_1","pgXP_039715790_1",
              "pvXP_011371998_1","raXP_016019355_1","sbXP_066090953_1",
              "ajXP_037013594_2","phXP_045692109_1","pdXP_028376576_1",
              "drXP_024409693_2","moXP_036124947_1","mbXP_014386752_1",
              "mlXP_006109297_1","mdXP_059527509_1","mmXP_036198054_1",
              "pkXP_036268269_1","pnCAK6438618_1"),
  dN_dS   = c(0.097,0.076,0.182,0.061,0.085,0.089,0.029,0.039,
              0.022,0.052,0.080,0.042,0.047,0.045,0.042,0.032,0.025)
)

df3 <- data.frame(
  Lineage = c(
    "rfXP_032986514_1", "paXP_015453119_1", "pgXP_039735734_1", 
    "pvXP_011385878_1", "raXP_036094006_1", "sbXP_066115246_1",
    "ajXP_037014053_2", "phXP_045708392_1", "pdXP_028359441_1",
    "drXP_045052603_1", "moXP_036099573_1", "mbXP_005878984_1",
    "mlXP_023601736_1", "mdXP_059542708_1", "mmXP_036201658_1",
    "pkXP_036301498_1", "pnCAK6442005_1"
  ),
  dN_dS = c(
    0.000, 0.047, 0.036, 0.183, 0.040, 0.014, 0.024, 0.027, 
    0.026, 0.124, 0.019, 0.179, 0.024, 0.026, 0.023, 0.022, 0.020
  )
)

df4 <- data.frame(
  Lineage = c(
    "rfXP_032987550_1", "paXP_006910842_1", "pgXP_039699242_1", 
    "pvXP_011363736_1", "raXP_015979423_2", "sbXP_066101817_1",
    "ajXP_037004116_2", "phXP_045713410_1", "pdXP_028374036_1",
    "drXP_024421621_2", "moXP_036123050_1", "mbXP_014385766_1",
    "mlXP_006099835_2", "mdXP_059520391_1", "mmXP_036189063_1",
    "pkXP_036287769_1", "pnCAK6435747_1"
  ),
  dN_dS = c(
    0.036, 0.098, 0.102, 0.102, 0.145, 0.077, 0.043, 0.053, 
    0.038, 0.040, 0.149, 0.329, 0.291, 0.075, 0.077, 0.069, 0.055
  )
)

df5 <- data.frame(
  Lineage = c(
    "rfXP_032984110_1",
    "paXP_006908795_1",
    "pgXP_039736709_1",
    "pvXP_011387201_1",
    "raXP_063715412_1",
    "sbXP_066099488_1",
    "ajXP_037050297_1",
    "phXP_024637532_1",
    "pdXP_028359860_1",
    "drXP_045025217_1",
    "moXP_036103709_1",
    "mbXP_014383900_1",
    "mlXP_014306449_1",
    "mdXP_059524129_1",
    "mmXP_036202361_1",
    "pkXP_036288622_1",
    "pnCAK6436556_1"
  ),
  dN_dS = c(
    0.042,
    0.024,
    0.021,
    0.018,
    0.028,
    0.102,
    0.038,
    0.028,
    0.029,
    0.045,
    0.085,
    0.125,
    0.116,
    0.060,
    0.052,
    0.041,
    0.047
  )
)

df6 <- data.frame(
  Lineage = c(
    "rfXP_032968530_1",
    "paXP_006906806_1",
    "pgXP_039700651_1",
    "pvXP_011356944_1",
    "raXP_016012252_2",
    "sbXP_066135523_1",
    "ajXP_036988173_2",
    "phXP_045709143_1",
    "pdXP_028366069_1",
    "drXP_024419668_2",
    "moXP_036110717_1",
    "mbXP_005882353_1",
    "mlXP_006082553_1",
    "mdXP_059559054_1",
    "mmXP_036175349_1",
    "pkXP_036277090_1",
    "pnCAK6448113_1"
  ),
  dN_dS = c(
    0.524,
    0.421,
    0.427,
    0.427,
    0.626,
    0.500,
    0.534,
    0.465,
    0.448,
    0.417,
    0.501,
    0.543,
    0.419,
    0.452,
    0.412,
    0.525,
    0.693
  )
)

df7 <- data.frame(
  Lineage = c(
    "rfXP_032956285_1",
    "paXP_006926372_1",
    "pgXP_039708889_1",
    "pvXP_023390598_1",
    "raXP_015996101_1",
    "sbXP_066110360_1",
    "ajXP_036987125_1",
    "phXP_045698783_1",
    "pdXP_028365299_1",
    "drXP_053769761_1",
    "moXP_036104716_1",
    "mbXP_005872144_1",
    "mlXP_023612273_1",
    "mdXP_059556959_1",
    "mmXP_036173428_1",
    "pkXP_036271131_1",
    "pnCAK6446769_1"
  ),
  dN_dS = c(
    0.000,
    0.070,
    0.070,
    0.070,
    0.021,
    0.027,
    0.025,
    0.019,
    0.019,
    0.424,
    0.049,
    0.037,
    0.034,
    0.024,
    0.026,
    0.022,
    0.018
  )
)

df8 <- data.frame(
  Lineage = c(
    "rfXP_032952366_1",
    "paXP_015455893_1",
    "pgXP_039710410_1",
    "pvXP_011353022_1",
    "raXP_036074329_1",
    "sbXP_066107492_1",
    "drXP_024413494_2",
    "ajXP_053524271_1",
    "phXP_045698747_1",
    "pdXP_035878485_1",
    "moXP_036104241_1",
    "mdXP_059557556_1",
    "mmXP_036172778_1",
    "mbXP_014390861_1",
    "mlXP_006100653_1",
    "pkXP_045429115_1",
    "pnCAK6446945_1"
  ),
  dN_dS = c(
    0.063,
    0.056,
    0.056,
    0.061,
    0.068,
    0.080,
    0.106,
    0.064,
    0.052,
    0.056,
    0.195,
    0.216,
    0.215,
    0.342,
    0.168,
    0.176,
    0.222
  )
)

df9 <- data.frame(
  Lineage = c(
    "rfXP_032953592_1",
    "paXP_024900246_1",
    "pgXP_039724240_1",
    "pvXP_011354167_1",
    "raXP_036077570_1",
    "sbXP_066116542_1",
    "ajXP_037012039_2",
    "phXP_045698041_1",
    "pdXP_028383838_1",
    "drXP_053773757_1",
    "moXP_036132246_1",
    "mbXP_005863827_1",
    "mlXP_023616189_1",
    "mdXP_059532519_1",
    "mmXP_036209432_1",
    "pkXP_045441337_1",
    "pnCAK6440418_1"
  ),
  dN_dS = c(
    0.042,
    0.143,
    0.017,
    0.017,
    0.024,
    0.038,
    0.059,
    0.036,
    0.030,
    0.034,
    0.090,
    0.229,
    0.086,
    0.105,
    0.107,
    0.063,
    0.075
  )
)

df10 <- data.frame(
  Lineage = c(
    "rfXP_032948942_1",
    "paXP_024896273_1",
    "pgXP_039737779_1",
    "pvXP_023394054_1",
    "raXP_015995384_2",
    "sbXP_066117732_1",
    "ajXP_053523895_1",
    "phXP_045697476_1",
    "pdXP_035871659_1",
    "drXP_045038263_2",
    "moXP_036129261_1",
    "mbXP_014405159_1",
    "mlXP_023604012_1",
    "mdXP_059530861_1",
    "mmXP_036198707_1",
    "pkXP_036298055_1",
    "pnCAK6439750_1"
  ),
  dN_dS = c(
    0.066,
    0.075,
    0.072,
    0.085,
    0.066,
    0.121,
    0.093,
    0.101,
    0.088,
    0.038,
    0.031,
    0.040,
    0.316,
    0.046,
    0.042,
    0.048,
    0.038
  )
)

df11 <- data.frame(
  Lineage = c(
    "rfXP_032974401_1",
    "paXP_006904563_1",
    "pgXP_039718178_1",
    "pvXP_011374072_1",
    "raXP_015998248_1",
    "sbXP_066114632_1",
    "ajXP_053517926_1",
    "phXP_045685631_1",
    "pdXP_028388683_1",
    "drXP_024431534_1",
    "moXP_036107339_1",
    "mbXP_005869330_1",
    "mlXP_023615498_1",
    "mdXP_059539103_1",
    "mmXP_036154503_1",
    "pkXP_045444000_1",
    "pnCAK6443372_1"
  ),
  dN_dS = c(
    0.047,
    0.000,
    0.000,
    0.000,
    0.000,
    0.125,
    0.109,
    0.129,
    0.129,
    0.160,
    0.301,
    0.665,
    0.266,
    0.207,
    0.265,
    0.108,
    0.108
  )
)

df12 <- data.frame(
  Lineage = c(
    "rfXP_032985148_1",
    "sbXP_066100214_1",
    "ajXP_037020213_2",
    "phXP_045674633_1",
    "pdXP_035868054_1",
    "drXP_024411640_2",
    "moXP_036131952_1",
    "mbXP_005881222_1",
    "mlXP_023611825_1",
    "mdXP_059523517_1",
    "mmXP_036201257_1",
    "pkXP_045438670_1",
    "pnCAK6436762_1",
    "paXP_024901073_1",
    "pgXP_039708001_1",
    "pvXP_023379420_1",
    "raXP_036087282_1"
  ),
  dN_dS = c(
    0.557,
    0.460,
    0.327,
    0.318,
    0.301,
    0.605,
    0.415,
    0.311,
    0.280,
    0.312,
    0.307,
    0.397,
    0.422,
    0.296,
    0.344,
    0.326,
    0.370
  )
)

df13 <- data.frame(
  Lineage = c(
    "rfXP_032977066_1",
    "paXP_024902071_1",
    "pgXP_039716318_1",
    "pvXP_011366977_1",
    "raXP_036076327_1",
    "sbXP_066107875_1",
    "ajXP_036996135_1",
    "phXP_045711209_1",
    "pdXP_028373211_1",
    "drXP_053781290_1",
    "moXP_036114824_1",
    "mdXP_059566352_1",
    "mmXP_036151612_1",
    "mbXP_014400770_1",
    "mlXP_023602542_1",
    "pkXP_036273162_1",
    "pnCAK6449637_1"
  ),
  dN_dS = c(
    0.012,
    0.013,
    0.013,
    0.012,
    0.010,
    0.008,
    0.033,
    0.024,
    0.036,
    0.044,
    0.027,
    0.018,
    0.014,
    0.125,
    0.140,
    0.027,
    0.024
  )
)

df14 <- data.frame(
  Lineage = c(
    "rfXP_032943899_1",
    "paXP_006912035_1",
    "pgXP_039718426_1",
    "pvXP_011361058_1",
    "raXP_016015076_2",
    "sbXP_066096122_1",
    "ajXP_037019810_2",
    "phXP_045684530_1",
    "pdXP_035866435_1",
    "drXP_024418899_2",
    "moXP_036103011_1",
    "mbXP_014400099_1",
    "mlXP_014307973_1",
    "mdXP_059568479_1",
    "mmXP_036183853_1",
    "pkXP_036285120_1",
    "pnCAK6434485_1"
  ),
  dN_dS = c(
    0.384,
    0.395,
    0.357,
    0.356,
    0.505,
    0.318,
    0.308,
    0.331,
    0.303,
    0.290,
    0.235,
    0.450,
    0.314,
    0.324,
    0.248,
    0.288,
    0.261
  )
)

df15 <- data.frame(
  Lineage = c(
    "rfXP_032957960_1",
    "paXP_006914229_1",
    "pgXP_039740349_1",
    "pvXP_011376890_1",
    "raXP_015988313_2",
    "sbXP_066131866_1",
    "ajXP_053527662_1",
    "phXP_045705205_1",
    "pdXP_028386087_2",
    "drXP_045051480_1",
    "moXP_036136016_1",
    "mbXP_014385277_1",
    "mlXP_023608434_1",
    "mdXP_059548147_1",
    "mmXP_036166096_1",
    "pkXP_036268927_1",
    "pnCAK6445017_1"
  ),
  dN_dS = c(
    0.099,
    0.089,
    0.084,
    0.084,
    0.177,
    0.137,
    0.117,
    0.146,
    0.131,
    0.083,
    0.097,
    0.383,
    0.100,
    0.100,
    0.092,
    0.138,
    0.141
  )
)

df16 <- data.frame(
  Lineage = c(
    "rfXP_032952715_1",
    "paXP_024905334_1",
    "pgXP_039735217_1",
    "pvXP_011376831_1",
    "raXP_015974584_2",
    "sbXP_066129640_1",
    "ajXP_036993761_2",
    "phXP_045680956_1",
    "pdXP_028384534_1",
    "drXP_024419672_2",
    "moXP_036097854_1",
    "pkXP_036288476_1",
    "pnCAK6446425_1",
    "mbXP_014390748_1",
    "mlXP_014304793_1",
    "mdXP_059555429_1",
    "mmXP_036169334_1"
  ),
  dN_dS = c(
    0.209,
    0.546,
    0.392,
    0.435,
    0.372,
    0.286,
    0.177,
    0.146,
    0.168,
    0.043,
    0.524,
    0.291,
    0.244,
    0.506,
    0.413,
    0.239,
    0.155
  )
)

df17 <- data.frame(
  Lineage = c(
    "rfXP_032952219_1",
    "paXP_015453419_1",
    "pgXP_039692344_1",
    "pvXP_011365804_1",
    "raXP_015979626_1",
    "sbXP_066128915_1",
    "ajXP_053522461_1",
    "phXP_045682059_1",
    "pdXP_028383783_1",
    "drXP_053768868_1",
    "moXP_036096732_1",
    "pkXP_045432831_1",
    "pnCAK6446173_1",
    "mbXP_014384547_1",
    "mlXP_014309403_1",
    "mdXP_059554721_1",
    "mmXP_036167860_1"
  ),
  dN_dS = c(
    0.030,
    0.115,
    0.073,
    0.104,
    0.054,
    0.080,
    0.017,
    0.012,
    0.011,
    0.012,
    0.105,
    0.265,
    0.242,
    0.510,
    0.239,
    0.200,
    0.301
  )
)

df18 <- data.frame(
  Lineage = c(
    "rfXP_032984101_1",
    "paXP_006916006_3",
    "pgXP_039713005_1",
    "pvXP_011377562_1",
    "raXP_016009715_1",
    "sbXP_066127095_1",
    "ajXP_037015728_1",
    "phXP_045686283_1",
    "pdXP_028385280_1",
    "drXP_024426129_1",
    "moXP_036130386_1",
    "pkXP_036292312_1",
    "pnCAK6437490_1",
    "mbXP_014392816_1",
    "mlXP_006109072_1",
    "mdXP_059521078_1",
    "mmXP_036162615_1"
  ),
  dN_dS = c(
    0.028,
    0.043,
    0.035,
    0.041,
    0.035,
    0.022,
    0.027,
    0.016,
    0.018,
    0.028,
    0.009,
    0.027,
    0.017,
    0.106,
    0.174,
    0.012,
    0.013
  )
)


df19 <- data.frame(
  Lineage = c(
    "rfXP_032963172_1",
    "paXP_006911708_1",
    "pgXP_039730182_1",
    "pvXP_023388071_1",
    "raXP_036082585_1",
    "sbXP_066105933_1",
    "ajXP_037013438_1",
    "phXP_045711989_1",
    "pdXP_028378622_1",
    "drXP_024425702_1",
    "moXP_036127494_1",
    "mbXP_014399783_1",
    "mlXP_023609510_1",
    "mdXP_059535550_1",
    "mmXP_036159909_1",
    "pkXP_036295446_1",
    "pnCAK6450645_1"
  ),
  dN_dS = c(
    0.197,
    1.089,
    0.831,
    0.621,
    0.678,
    0.198,
    0.447,
    0.423,
    0.492,
    0.521,
    0.161,
    0.125,
    0.736,
    0.109,
    0.318,
    0.109,
    0.125
  )
)


df20 <- data.frame(
  Lineage = c(
    "rfXP_032981290_1",
    "paXP_024901437_1",
    "pgXP_039743542_1",
    "pvXP_011354914_1",
    "raXP_015984538_2",
    "sbXP_066122809_1",
    "drXP_053781253_1",
    "ajXP_037002657_1",
    "phXP_045700675_1",
    "pdXP_028372231_1",
    "moXP_036119128_1",
    "mbXP_005875028_2",
    "mlXP_023619242_1",
    "mdXP_059516341_1",
    "mmXP_036191014_1",
    "pkXP_036290881_1",
    "pnCAK6434994_1"
  ),
  dN_dS = c(
    0.000,
    0.280,
    0.009,
    0.173,
    0.010,
    0.008,
    0.007,
    0.007,
    0.008,
    0.006,
    0.005,
    0.072,
    0.047,
    0.008,
    0.009,
    0.008,
    0.008
  )
)


df21 <- data.frame(
  Lineage = c(
    "rfXP_032966741_1",
    "paXP_006913445_2",
    "pgXP_039704346_1",
    "pvXP_023389375_1",
    "raXP_015992720_2",
    "sbXP_066130067_1",
    "ajXP_036994512_1",
    "phXP_045693842_1",
    "pdXP_035876529_1",
    "drXP_045053573_1",
    "moXP_036134736_1",
    "mbXP_005866503_1",
    "mlXP_014320125_1",
    "mdXP_059533970_1",
    "mmXP_036211667_1",
    "pkXP_036264828_1",
    "pnCAK6444417_1"
  ),
  dN_dS = c(
    0.044,
    0.125,
    0.100,
    0.111,
    0.140,
    0.108,
    0.376,
    0.092,
    0.109,
    0.410,
    0.222,
    0.449,
    0.425,
    0.097,
    0.097,
    0.314,
    0.258
  )
)


df22 <- data.frame(
  Lineage = c(
    "rfXP_032959652_1",
    "paXP_006916498_1",
    "pgXP_039695278_1",
    "pvXP_011370328_1",
    "raXP_036084494_1",
    "sbXP_066093336_1",
    "ajXP_053521618_1",
    "phXP_045695266_1",
    "pdXP_035867098_1",
    "drXP_045057025_1",
    "moXP_036105905_1",
    "mbXP_005861452_1",
    "mlXP_023604729_1",
    "mdXP_059537179_1",
    "mmXP_036157125_1",
    "pkXP_036312814_1",
    "pnCAK6442919_1"
  ),
  dN_dS = c(
    0.062,
    0.047,
    0.044,
    0.044,
    0.014,
    0.009,
    0.032,
    0.041,
    0.045,
    0.043,
    0.061,
    0.051,
    0.052,
    0.336,
    0.075,
    0.061,
    0.061
  )
)


df23 <- data.frame(
  Lineage = c(
    "rfXP_032945656_1",
    "paXP_024897428_1",
    "pgXP_039729568_1",
    "pvXP_023381296_1",
    "raXP_016020261_1",
    "sbXP_066119860_1",
    "ajXP_036996508_2",
    "phXP_045693123_1",
    "pdXP_028376577_1",
    "drXP_024428789_2",
    "moXP_036124875_1",
    "mdXP_059526825_1",
    "mmXP_036197616_1",
    "mbXP_005861965_1",
    "mlXP_023620139_1",
    "pkXP_036263823_1",
    "pnCAK6438378_1"
  ),
  dN_dS = c(
    0.043,
    0.183,
    0.076,
    0.013,
    0.011,
    0.033,
    0.023,
    0.025,
    0.020,
    0.017,
    0.033,
    0.017,
    0.009,
    0.010,
    0.423,
    0.037,
    0.060
  )
)


df24 <- data.frame(
  Lineage = c(
    "rfXP_032984803_1",
    "paXP_024902552_1",
    "pgXP_039722132_1",
    "pvXP_023391126_1",
    "raXP_016005974_1",
    "sbXP_066127896_1",
    "ajXP_037003923_1",
    "phXP_045701093_1",
    "pdXP_028385790_1",
    "drXP_024425231_1",
    "moXP_036130990_1",
    "mbXP_005863442_1",
    "mlXP_023604919_1",
    "mdXP_059522145_1",
    "mmXP_036201853_1",
    "pkXP_036272320_1",
    "pnCAK6437196_1"
  ),
  dN_dS = c(
    0.043,
    0.040,
    0.036,
    0.151,
    0.050,
    0.053,
    0.042,
    0.033,
    0.040,
    0.047,
    0.032,
    0.170,
    0.176,
    0.028,
    0.013,
    0.014,
    0.012
  )
)


df25 <- data.frame(
  Lineage = c(
    "rfXP_032980923_1",
    "paXP_006909702_1",
    "pgXP_039743223_1",
    "pvXP_011359652_1",
    "raXP_015997721_1",
    "sbXP_066123106_1",
    "ajXP_037017968_2",
    "phXP_045684175_1",
    "pdXP_028370358_1",
    "drXP_024408647_1",
    "moXP_036118962_1",
    "mdXP_059515956_1",
    "mmXP_036190621_1",
    "mbXP_014393980_1",
    "mlXP_006083538_1",
    "pkXP_045436649_1",
    "pnCAK6435041_1"
  ),
  dN_dS = c(
    0.077,
    0.036,
    0.038,
    0.038,
    0.042,
    0.031,
    0.025,
    0.043,
    0.059,
    0.029,
    0.019,
    0.037,
    0.037,
    0.037,
    0.201,
    0.060,
    0.049
  )
)


df26 <- data.frame(
  Lineage = c(
    "rfXP_032951697_1",
    "paXP_006923326_1",
    "pgXP_039736235_1",
    "pvXP_011370244_1",
    "raXP_015997867_1",
    "sbXP_066122380_1",
    "ajXP_036996368_1",
    "phXP_045706503_1",
    "pdXP_028390372_1",
    "drXP_024416371_2",
    "moXP_036097381_1",
    "mdXP_059551797_1",
    "mmXP_036170290_1",
    "mbXP_014390342_1",
    "mlXP_014314153_1",
    "pkXP_036306508_1",
    "pnCAK6445173_1"
  ),
  dN_dS = c(
    0.004,
    0.018,
    0.022,
    0.018,
    0.017,
    0.034,
    0.028,
    0.020,
    0.023,
    0.032,
    0.045,
    0.024,
    0.025,
    0.266,
    0.032,
    0.036,
    0.031
  )
)


df27 <- data.frame(
  Lineage = c(
    "rfXP_032960709_1",
    "paXP_015448287_1",
    "pgXP_039709588_1",
    "pvXP_011370975_1",
    "raXP_015982926_1",
    "sbXP_066091509_1",
    "ajXP_036998406_1",
    "phXP_045677479_1",
    "pdXP_028382756_1",
    "drXP_024434723_2",
    "moXP_036105675_1",
    "mdXP_059541322_1",
    "mmXP_036155836_1",
    "mbXP_005856578_1",
    "mlXP_014318392_1",
    "pkXP_036312787_1",
    "pnCAK6442662_1"
  ),
  dN_dS = c(
    0.018,
    0.102,
    0.000,
    0.000,
    0.000,
    0.014,
    0.000,
    0.000,
    0.000,
    0.000,
    0.000,
    0.048,
    0.051,
    0.054,
    0.255,
    0.045,
    0.047
  )
)


df28 <- data.frame(
  Lineage = c(
    "rfXP_032971451_1",
    "raXP_016018760_2",
    "paXP_015442747_1",
    "pgXP_039722376_1",
    "pvXP_011379840_1",
    "sbXP_066124384_1",
    "ajXP_036987172_1",
    "phXP_045703082_1",
    "pdXP_028368171_1",
    "drXP_053776704_1",
    "moXP_036112604_1",
    "mbXP_014398159_1",
    "mlXP_006082371_1",
    "mdXP_059544744_1",
    "mmXP_036205592_1",
    "pkXP_036301197_1",
    "pnCAK6441367_1"
  ),
  dN_dS = c(
    0.000,
    0.017,
    0.025,
    0.025,
    0.184,
    0.000,
    0.009,
    0.008,
    0.008,
    0.010,
    0.000,
    0.189,
    0.027,
    0.012,
    0.012,
    0.021,
    0.022
  )
)


df29 <- data.frame(
  Lineage = c(
    "rfXP_032943426_1",
    "paXP_006919954_1",
    "pgXP_039697315_1",
    "pvXP_011358783_1",
    "raXP_016004267_1",
    "sbXP_066103598_1",
    "ajXP_037020940_1",
    "phXP_045705579_1",
    "pdXP_028379738_1",
    "drXP_045043885_1",
    "moXP_036100110_1",
    "mbXP_005866327_1",
    "mlXP_023606582_1",
    "mdXP_059563307_1",
    "mmXP_036177818_1",
    "pkXP_036274608_1",
    "pnCAK6447942_1"
  ),
  dN_dS = c(
    0.011,
    0.000,
    0.000,
    0.000,
    0.035,
    0.000,
    0.011,
    0.032,
    0.028,
    0.000,
    0.017,
    0.345,
    0.011,
    0.000,
    0.000,
    0.019,
    0.333
  )
)

df30 <- data.frame(
  Lineage = c(
    "rfXP_032978124_1",
    "paXP_015449707_1",
    "pgXP_039714942_1",
    "pvXP_011368171_1",
    "raXP_036095142_1",
    "sbXP_066111675_1",
    "ajXP_053523636_1",
    "phXP_045702103_1",
    "pdXP_035878412_1",
    "drXP_053774411_1",
    "moXP_036116331_1",
    "mbXP_014385182_1",
    "mlXP_006107863_1",
    "mdXP_059514522_1",
    "mmXP_036162789_1",
    "pkXP_036272688_2",
    "pnCAK6433452_1"
  ),
  dN_dS = c(
    0.148,
    0.140,
    0.155,
    0.136,
    0.100,
    0.111,
    0.114,
    0.101,
    0.109,
    0.124,
    0.113,
    0.230,
    0.124,
    0.124,
    0.127,
    0.113,
    0.104
  )
)

df31 <- data.frame(
  Lineage = c(
    "rfXP_032952841_1",
    "paXP_006923224_1",
    "pgXP_039735909_1",
    "pvXP_011360574_1",
    "raXP_015973922_1",
    "sbXP_066132626_1",
    "ajXP_036985862_1",
    "phXP_045684402_1",
    "pdXP_028384641_1",
    "drXP_024412873_1",
    "moXP_036097190_1",
    "mbXP_014394659_1",
    "mlXP_006100205_1",
    "mdXP_059551317_1",
    "mmXP_036170037_1",
    "pkXP_036280254_1",
    "pnCAK6446028_1"
  ),
  dN_dS = c(
    1.151,
    0.299,
    0.302,
    0.303,
    0.415,
    0.739,
    0.420,
    0.575,
    0.533,
    0.361,
    1.333,
    0.792,
    1.028,
    0.880,
    0.920,
    0.569,
    0.707
  )
)

df32 <- data.frame(
  Lineage = c(
    "rfXP_032976778_1",
    "paXP_006911045_1",
    "pgXP_039696480_1",
    "pvXP_023393127_1",
    "raXP_036076881_1",
    "sbXP_066108410_1",
    "ajXP_037012239_2",
    "phXP_045677081_1",
    "pdXP_028371838_1",
    "drXP_053779900_1",
    "moXP_036137039_1",
    "mbXP_014388583_1",
    "mlXP_014309688_1",
    "mdXP_059566682_1",
    "mmXP_036151520_1",
    "pkXP_036284693_1",
    "pnCAK6449857_1"
  ),
  dN_dS = c(
    0.035,
    0.058,
    0.061,
    0.121,
    0.051,
    0.080,
    0.121,
    0.126,
    0.119,
    0.129,
    0.092,
    0.158,
    0.075,
    0.069,
    0.071,
    0.076,
    0.084
  )
)

df33 <- data.frame(
  Lineage = c(
    "rfXP_032947041_1",
    "paXP_024902888_1",
    "pgXP_039715718_1",
    "pvXP_023387407_1",
    "raXP_036078963_1",
    "sbXP_066092108_1",
    "ajXP_036983982_1",
    "phXP_045693444_1",
    "pdXP_028377739_1",
    "drXP_045038138_2",
    "moXP_036125083_1",
    "pkXP_036263537_1",
    "pnCAK6438473_1",
    "mdXP_059527031_1",
    "mmXP_036196336_1",
    "mbXP_014391581_1",
    "mlXP_006094976_1"
  ),
  dN_dS = c(
    1.000,
    0.344,
    0.341,
    0.342,
    0.208,
    1.410,
    0.318,
    0.316,
    0.460,
    0.343,
    0.319,
    0.398,
    0.362,
    0.605,
    0.338,
    0.288,
    0.424
  )
)

#should be removed
df34 <- data.frame(
  Lineage = c(
    "rfXP_032983396_1",
    "pgXP_039735801_1",
    "pvXP_023381986_1",
    "raXP_036088209_1",
    "sbXP_066120899_1",
    "ajXP_053512654_1",
    "phXP_045686254_1",
    "pdXP_035869418_1",
    "drXP_053771846_1",
    "moXP_036131609_1",
    "mdXP_059521748_1",
    "mmXP_036162439_1",
    "mbXP_014383607_1",
    "mlXP_023602881_1",
    "pkXP_045440158_1",
    "pnCAK6437448_1"
  ),
  dN_dS = c(
    0.446,
    0.305,
    0.334,
    0.336,
    0.442,
    0.471,
    1.020,
    0.738,
    0.405,
    0.709,
    0.650,
    0.455,
    0.573,
    0.360,
    0.392,
    0.905
  )
)

df35 <- data.frame(
  Lineage = c(
    "rfXP_032949397_1",
    "paXP_015453825_1",
    "pgXP_039723219_1",
    "pvXP_011364679_1",
    "raXP_016021670_2",
    "sbXP_066117897_1",
    "ajXP_036988541_2",
    "phXP_045673205_1",
    "pdXP_035871156_1",
    "drXP_053771477_1",
    "moXP_036129207_1",
    "mbXP_014396138_1",
    "mlXP_023617296_1",
    "mdXP_059529425_1",
    "mmXP_036198532_1",
    "pkXP_036297886_1",
    "pnCAK6439518_1"
  ),
  dN_dS = c(
    0.070,
    0.029,
    0.068,
    0.023,
    0.028,
    0.095,
    0.119,
    0.123,
    0.133,
    0.108,
    0.100,
    0.264,
    0.233,
    0.120,
    0.114,
    0.095,
    0.085
  )
)

df36 <- data.frame(
  Lineage = c(
    "rfXP_032943845_1",
    "paXP_006912181_2",
    "pgXP_039718656_1",
    "pvXP_011360776_2",
    "raXP_016015527_1",
    "sbXP_066095859_1",
    "ajXP_053517685_1",
    "phXP_045684836_1",
    "pdXP_028381514_1",
    "drXP_024418788_1",
    "moXP_036103095_1",
    "mbXP_005868138_1",
    "mlXP_014308246_2",
    "mdXP_059511815_1",
    "mmXP_036183242_1",
    "pkXP_036285779_1",
    "pnCAK6434686_1"
  ),
  dN_dS = c(
    0.067,
    0.117,
    0.106,
    0.124,
    0.120,
    0.164,
    0.096,
    0.090,
    0.112,
    0.129,
    0.156,
    0.096,
    0.309,
    0.087,
    0.087,
    0.116,
    0.119
  )
)

df37 <- data.frame(
  Lineage = c(
    "rfXP_032969713_1",
    "paXP_024901541_1",
    "pgXP_039732692_1",
    "pvXP_011384341_1",
    "raXP_016000014_2",
    "sbXP_066126850_1",
    "ajXP_037002757_2",
    "phXP_045714588_1",
    "pdXP_028368662_1",
    "drXP_024410640_3",
    "moXP_036137387_1",
    "mbXP_014399272_1",
    "mlXP_023610818_1",
    "mdXP_059547608_1",
    "mmXP_036163372_1",
    "pkXP_036292323_1",
    "pnCAK6440476_1"
  ),
  dN_dS = c(
    0.120,
    0.128,
    0.126,
    0.126,
    0.136,
    0.162,
    0.196,
    0.179,
    0.175,
    0.256,
    0.198,
    0.312,
    0.138,
    0.155,
    0.149,
    0.136,
    0.138
  )
)


df38 <- data.frame(
  Lineage = c(
    "rfXP_032984557_1",
    "paXP_015450180_1",
    "pgXP_039709385_1",
    "pvXP_023393672_1",
    "raXP_016022293_2",
    "sbXP_066098129_1",
    "ajXP_036989539_2",
    "phXP_045702539_1",
    "pdXP_028385248_1",
    "drXP_053771230_1",
    "moXP_036131046_1",
    "mbXP_014397180_1",
    "mlXP_023599734_1",
    "mdXP_059522496_1",
    "mmXP_036202140_1",
    "pkXP_036263879_1",
    "pnCAK6437069_1"
  ),
  dN_dS = c(
    0.097,
    0.052,
    0.056,
    0.155,
    0.084,
    0.077,
    0.099,
    0.104,
    0.104,
    0.097,
    0.084,
    0.157,
    0.283,
    0.074,
    0.085,
    0.089,
    0.082
  )
)

df39 <- data.frame(
  Lineage = c(
    "rfXP_032946250_1",
    "paXP_015443556_1",
    "pgXP_039702735_1",
    "pvXP_011375401_1",
    "raXP_015995306_1",
    "sbXP_066118805_1",
    "ajXP_037019581_1",
    "phXP_045692528_1",
    "pdXP_028377577_1",
    "drXP_053766779_1",
    "moXP_036124766_1",
    "mdXP_059525805_1",
    "mmXP_036196338_1",
    "mbXP_014387357_1",
    "mlXP_023612574_1",
    "pkXP_036307378_1",
    "pnCAK6442720_1"
  ),
  dN_dS = c(
    0.131,
    0.135,
    0.136,
    0.172,
    0.210,
    0.038,
    0.264,
    0.082,
    0.083,
    0.150,
    0.122,
    0.139,
    0.140,
    0.484,
    0.470,
    0.158,
    0.192
  )
)

df40 <- data.frame(
  Lineage = c(
    "rfXP_032950596_1",
    "paXP_006921927_1",
    "pgXP_039720380_1",
    "pvXP_011367629_1",
    "raXP_036085651_1",
    "sbXP_066093581_1",
    "ajXP_053527377_1",
    "phXP_045689604_1",
    "pdXP_028380283_2",
    "drXP_024415395_2",
    "moXP_036100336_1",
    "mbXP_005862201_1",
    "mlXP_006091444_1",
    "mdXP_059562906_1",
    "mmXP_036177124_1",
    "pkXP_036275578_1",
    "pnCAK6447882_1"
  ),
  dN_dS = c(
    0.497,
    0.982,
    0.860,
    0.906,
    0.633,
    0.152,
    0.410,
    0.182,
    0.165,
    0.270,
    0.177,
    0.208,
    0.202,
    0.271,
    0.245,
    0.268,
    0.289
  )
)

df41 <- data.frame(
  Lineage = c(
    "rfXP_032955351_1",
    "paXP_024906549_1",
    "pgXP_039736596_1",
    "pvXP_023393640_1",
    "raXP_016018915_1",
    "sbXP_066091194_1",
    "ajXP_037000003_2",
    "phXP_045673748_1",
    "pdXP_028386979_1",
    "drXP_053768297_1",
    "moXP_036133556_1",
    "pkXP_036278661_1",
    "pnCAK6442290_1",
    "mdXP_059533405_1",
    "mmXP_036210543_1",
    "mbXP_005868843_1",
    "mlXP_023601666_1"
  ),
  dN_dS = c(
    0.093,
    0.016,
    0.011,
    0.014,
    0.011,
    0.457,
    0.008,
    0.014,
    0.014,
    0.008,
    0.017,
    0.007,
    0.007,
    0.015,
    0.015,
    0.008,
    0.109
  )
)

df42 <- data.frame(
  Lineage = c(
    "rfXP_032973402_1",
    "paXP_006909034_1",
    "pgXP_039721777_1",
    "pvXP_011381206_1",
    "raXP_015983179_1",
    "sbXP_066114960_1",
    "ajXP_037023071_1",
    "phXP_045685543_1",
    "pdXP_028388006_1",
    "drXP_045040562_2",
    "moXP_036107151_1",
    "mdXP_059539407_1",
    "mmXP_036152694_1",
    "mbXP_014405549_1",
    "mlXP_006093600_1",
    "pkXP_036314025_1",
    "pnCAK6443250_1"
  ),
  dN_dS = c(
    0.083,
    0.257,
    0.240,
    0.255,
    0.214,
    0.127,
    0.044,
    0.075,
    0.091,
    0.294,
    0.158,
    0.025,
    0.393,
    0.059,
    0.028,
    0.055,
    0.047
  )
)

df43 <- data.frame(
  Lineage = c(
    "rfXP_032974873_1",
    "paXP_006911365_1",
    "pgXP_039696603_1",
    "pvXP_011379946_1",
    "raXP_036077508_1",
    "sbXP_066108321_1",
    "ajXP_053518882_1",
    "phXP_045710987_1",
    "pdXP_028370287_1",
    "drXP_053779719_1",
    "moXP_036116121_1",
    "mdXP_059567211_1",
    "mmXP_036182321_1",
    "mbXP_014393715_1",
    "mlXP_023599618_1",
    "pkXP_036272388_1",
    "pnCAK6437841_1"
  ),
  dN_dS = c(
    0.179,
    0.231,
    0.217,
    0.213,
    0.333,
    0.172,
    0.461,
    0.219,
    0.207,
    0.519,
    0.260,
    0.419,
    0.472,
    0.451,
    0.475,
    0.342,
    0.159
  )
)

df44 <- data.frame(
  Lineage = c(
    "rfXP_032984088_1",
    "paXP_015443182_1",
    "pgXP_039707776_1",
    "pvXP_011378838_1",
    "raXP_016020011_1",
    "sbXP_066099595_1",
    "ajXP_037018482_2",
    "phXP_045674420_1",
    "pdXP_035867952_1",
    "drXP_045047929_2",
    "moXP_036130407_1",
    "mbXP_014403303_1",
    "mlXP_023618870_1",
    "mdXP_059523668_1",
    "mmXP_036202463_1",
    "pkXP_036297338_1",
    "pnCAK6436718_1"
  ),
  dN_dS = c(
    0.156,
    0.307,
    0.279,
    0.269,
    0.495,
    0.257,
    0.136,
    0.221,
    0.224,
    0.173,
    0.156,
    0.189,
    0.355,
    0.170,
    0.358,
    0.142,
    0.140
  )
)

df45 <- data.frame(
  Lineage = c(
    "rfXP_032984557_1",
    "paXP_006923224_1",
    "pgXP_039735909_1",
    "pvXP_011360574_1",
    "raXP_015973922_1",
    "sbXP_066132626_1",
    "ajXP_036985862_1",
    "phXP_045684402_1",
    "pdXP_028384641_1",
    "drXP_024412873_1",
    "moXP_036097190_1",
    "mbXP_014394659_1",
    "mlXP_006100205_1",
    "mdXP_059551317_1",
    "mmXP_036170037_1",
    "pkXP_036280254_1",
    "pnCAK6446028_1"
  ),
  dN_dS = c(
    1.149,
    0.299,
    0.302,
    0.302,
    0.416,
    0.740,
    0.419,
    0.574,
    0.533,
    0.361,
    1.332,
    0.789,
    1.028,
    0.879,
    0.921,
    0.568,
    0.707
  )
)

df46 <- data.frame(
  Lineage = c(
    "rfXP_032961894_1",
    "paXP_006919081_1",
    "pgXP_039693955_1",
    "pvXP_011372100_1",
    "raXP_015979710_2",
    "sbXP_066088897_1",
    "ajXP_036982712_2",
    "phXP_045690034_1",
    "pdXP_028359644_1",
    "drXP_045038620_2",
    "moXP_036099575_1",
    "mbXP_014388937_1",
    "mlXP_023608941_1",
    "mdXP_059532761_1",
    "mmXP_036183558_1",
    "pkXP_036272999_1",
    "pnCAK6433338_1"
  ),
  dN_dS = c(
    0.105,
    0.116,
    0.117,
    0.126,
    0.126,
    0.073,
    0.124,
    0.101,
    0.110,
    0.124,
    0.095,
    0.224,
    0.078,
    0.092,
    0.073,
    0.068,
    0.087
  )
)

df47 <- data.frame(
  Lineage = c(
    "rfXP_032961190_1",
    "sbXP_066135880_1",
    "ajXP_037006109_2",
    "phXP_045714005_1",
    "pdXP_028363499_2",
    "drXP_024433098_1",
    "moXP_036137138_1",
    "mbXP_005874893_1",
    "mlXP_014316104_1",
    "mdXP_059524733_1",
    "mmXP_036192054_1",
    "pkXP_036269895_1",
    "pnCAK6433079_1",
    "paXP_015448989_1",
    "pgXP_039701326_1",
    "pvXP_011358924_1",
    "raXP_016009496_1"
  ),
  dN_dS = c(
    0.985,
    0.475,
    0.360,
    0.512,
    0.346,
    0.471,
    0.415,
    1.081,
    1.306,
    1.383,
    0.908,
    1.244,
    1.648,
    1.176,
    0.922,
    1.000,
    0.410
  )
)

df48 <- data.frame(
  Lineage = c(
    "rfXP_032960709_1",
    "paXP_015448287_1",
    "pgXP_039709588_1",
    "pvXP_011370975_1",
    "raXP_015982926_1",
    "sbXP_066091509_1",
    "ajXP_036998406_1",
    "phXP_045677479_1",
    "pdXP_028382756_1",
    "drXP_024434723_2",
    "moXP_036105675_1",
    "mdXP_059541322_1",
    "mmXP_036155836_1",
    "mbXP_005856578_1",
    "mlXP_014318392_1",
    "pkXP_036312787_1",
    "pnCAK6442662_1"
  ),
  dN_dS = c(
    0.018,
    0.102,
    0.000,
    0.000,
    0.000,
    0.014,
    0.000,
    0.000,
    0.000,
    0.000,
    0.000,
    0.048,
    0.051,
    0.054,
    0.255,
    0.045,
    0.047
  )
)

df49 <- data.frame(
  Lineage = c(
    "rfXP_032952366_1",
    "paXP_015455893_1",
    "pgXP_039710410_1",
    "pvXP_011353022_1",
    "raXP_036074329_1",
    "sbXP_066107492_1",
    "drXP_024413494_2",
    "ajXP_053524271_1",
    "phXP_045698747_1",
    "pdXP_035878485_1",
    "moXP_036104241_1",
    "mdXP_059557556_1",
    "mmXP_036172778_1",
    "mbXP_014390861_1",
    "mlXP_006100653_1",
    "pkXP_045429115_1",
    "pnCAK6446945_1"
  ),
  dN_dS = c(
    0.063,
    0.056,
    0.056,
    0.061,
    0.068,
    0.080,
    0.106,
    0.064,
    0.052,
    0.056,
    0.195,
    0.216,
    0.215,
    0.342,
    0.168,
    0.176,
    0.222
  )
)

df50 <- data.frame(
  Lineage = c(
    "rfXP_032956859_1",
    "paXP_024900830_1",
    "pgXP_039707589_1",
    "pvXP_023377703_1",
    "raXP_036080874_1",
    "sbXP_066120922_1",
    "ajXP_037006028_2",
    "phXP_045683233_1",
    "pdXP_028370191_2",
    "drXP_024414225_1",
    "moXP_036119267_1",
    "mbXP_014385142_1",
    "mlXP_023604869_1",
    "mdXP_059516695_1",
    "mmXP_036167544_1",
    "pkXP_036291893_1",
    "pnCAK6435304_1"
  ),
  dN_dS = c(
    0.292,
    0.375,
    0.407,
    0.374,
    0.604,
    0.499,
    0.378,
    0.805,
    1.408,
    0.414,
    0.315,
    0.373,
    0.943,
    0.993,
    0.903,
    0.536,
    0.465
  )
)

df51 <- data.frame(
  Lineage = c(
    "rfXP_032956859_1",
    "paXP_024900830_1",
    "pgXP_039707589_1",
    "pvXP_023377703_1",
    "raXP_036080948_1",
    "sbXP_066120923_1",
    "ajXP_037006028_2",
    "phXP_045683233_1",
    "pdXP_028370191_2",
    "drXP_024414225_1",
    "moXP_036119267_1",
    "mbXP_005852975_1",
    "mlXP_023600127_1",
    "mdXP_059516697_1",
    "mmXP_036167544_1",
    "pkXP_036291893_1",
    "pnCAK6435304_1"
  ),
  dN_dS = c(
    0.149,
    0.252,
    0.294,
    0.264,
    0.354,
    0.601,
    1.033,
    1.087,
    1.522,
    1.313,
    0.584,
    0.414,
    0.876,
    0.906,
    0.831,
    0.798,
    0.784
  )
)

df52 <- data.frame(
  Lineage = c(
    "rfXP_032977430_1",
    "paXP_015441441_1",
    "pgXP_039714710_1",
    "pvXP_011366413_1",
    "raXP_036076128_1",
    "sbXP_066103755_1",
    "ajXP_036997559_2",
    "phXP_045674257_1",
    "pdXP_028373200_1",
    "drXP_024426682_2",
    "moXP_036116076_1",
    "mbXP_005856932_1",
    "mlXP_006093578_1",
    "mdXP_059563908_1",
    "mmXP_036180315_1",
    "pkXP_036283099_1",
    "pnCAK6448827_1"
  ),
  dN_dS = c(
    0.073,
    0.068,
    0.048,
    0.051,
    0.377,
    0.198,
    0.115,
    0.136,
    0.133,
    0.172,
    0.108,
    0.157,
    0.412,
    0.143,
    0.149,
    0.127,
    0.124
  )
)

df53 <- data.frame(
  Lineage = c(
    "rfXP_032952548_1",
    "paXP_024905339_1",
    "pgXP_039735046_1",
    "pvXP_011362168_1",
    "raXP_015974630_1",
    "sbXP_066129684_1",
    "ajXP_053512958_1",
    "phXP_045682636_1",
    "pdXP_028384825_1",
    "drXP_053782653_1",
    "moXP_036096655_1",
    "mbXP_014390751_1",
    "mlXP_014305981_1",
    "mdXP_059555471_1",
    "mmXP_036169299_1",
    "pkXP_036288501_1",
    "pnCAK6446435_1"
  ),
  dN_dS = c(
    0.144,
    0.184,
    0.126,
    0.354,
    0.143,
    0.094,
    0.086,
    0.126,
    0.127,
    0.095,
    0.175,
    0.247,
    0.304,
    0.179,
    0.146,
    0.105,
    0.099
  )
)

df54 <- data.frame(
  Lineage = c(
    "rfXP_032951697_1",
    "paXP_006923326_1",
    "pgXP_039736235_1",
    "pvXP_011370244_1",
    "raXP_015997867_1",
    "sbXP_066122380_1",
    "ajXP_036996368_1",
    "phXP_045706503_1",
    "pdXP_028390372_1",
    "drXP_024416371_2",
    "moXP_036097381_1",
    "mdXP_059551797_1",
    "mmXP_036170290_1",
    "mbXP_014390342_1",
    "mlXP_014314153_1",
    "pkXP_036306508_1",
    "pnCAK6445173_1"
  ),
  dN_dS = c(
    0.004,
    0.018,
    0.022,
    0.018,
    0.017,
    0.034,
    0.028,
    0.020,
    0.023,
    0.032,
    0.045,
    0.024,
    0.025,
    0.266,
    0.032,
    0.036,
    0.031
  )
)

df55 <- data.frame(
  Lineage = c(
    "rfXP_032977431_1",
    "paXP_015441442_1",
    "pgXP_039714711_1",
    "pvXP_023391152_1",
    "raXP_036076128_1",
    "sbXP_066103756_1",
    "ajXP_036997560_2",
    "phXP_045674258_1",
    "pdXP_028373201_1",
    "drXP_024426683_2",
    "moXP_036116076_1",
    "mbXP_005856931_1",
    "mlXP_006093578_1",
    "mdXP_059563909_1",
    "mmXP_036180316_1",
    "pkXP_036283100_1",
    "pnCAK6448827_1"
  ),
  dN_dS = c(
    0.074,
    0.071,
    0.050,
    0.053,
    0.195,
    0.212,
    0.119,
    0.140,
    0.137,
    0.170,
    0.110,
    0.161,
    0.408,
    0.147,
    0.153,
    0.130,
    0.127
  )
)

df56 <- data.frame(
  Lineage = c(
    "rfXP_032947686_1",
    "paXP_006924998_1",
    "pgXP_039741866_1",
    "pvXP_011377741_1",
    "raXP_015995601_2",
    "sbXP_066111481_1",
    "drXP_024409715_1",
    "ajXP_037019475_2",
    "phXP_045692157_1",
    "pdXP_028375035_1",
    "moXP_036125020_1",
    "mdXP_059524263_1",
    "mmXP_036196284_1",
    "mbXP_014390250_1",
    "mlXP_006101969_1",
    "pkXP_036306824_1",
    "pnCAK6438256_1"
  ),
  dN_dS = c(
    0.385,
    0.164,
    0.164,
    0.211,
    0.643,
    0.214,
    1.519,
    0.415,
    1.304,
    1.145,
    0.165,
    0.156,
    0.183,
    0.192,
    0.150,
    0.184,
    0.232
  )
)

df57 <- data.frame(
  Lineage = c(
    "rfXP_032956859_1",
    "paXP_024900830_1",
    "pgXP_039707589_1",
    "pvXP_023377703_1",
    "raXP_036080873_1",
    "sbXP_066120922_1",
    "ajXP_037006028_2",
    "phXP_045683233_1",
    "pdXP_028370191_2",
    "drXP_024414225_1",
    "moXP_036119267_1",
    "mdXP_059516695_1",
    "mmXP_036167544_1",
    "mbXP_005859481_1",
    "mlXP_023600127_1",
    "pkXP_036291893_1",
    "pnCAK6435304_1"
  ),
  dN_dS = c(
    0.190,
    0.358,
    0.388,
    0.382,
    0.661,
    0.409,
    0.511,
    1.829,
    11.585,
    0.459,
    0.274,
    0.684,
    0.711,
    1.016,
    0.617,
    0.658,
    0.517
  )
)

df58 <- data.frame(
  Lineage = c(
    "rfXP_032956749_1",
    "paXP_006909524_1",
    "pgXP_039707576_1",
    "pvXP_023377685_1",
    "raXP_016008642_2",
    "sbXP_066120920_1",
    "drXP_024414403_2",
    "ajXP_053520509_1",
    "phXP_045684314_1",
    "pdXP_028370187_1",
    "moXP_036119267_1",
    "pkXP_036291893_1",
    "pnCAK6435304_1",
    "mbXP_014399719_1",
    "mlXP_023600126_1",
    "mdXP_059516697_1",
    "mmXP_036167545_1"
  ),
  dN_dS = c(
    0.334,
    0.313,
    0.503,
    0.553,
    0.814,
    0.742,
    0.815,
    0.497,
    0.785,
    0.587,
    0.427,
    0.215,
    0.195,
    0.163,
    0.131,
    0.227,
    0.361
  )
)

df59 <- data.frame(
  Lineage = c(
    "rfXP_032956749_1",
    "paXP_006909524_1",
    "pgXP_039707576_1",
    "pvXP_023377685_1",
    "raXP_016008642_2",
    "sbXP_066120920_1",
    "drXP_024414403_2",
    "ajXP_053520509_1",
    "phXP_045684314_1",
    "pdXP_028370187_1",
    "moXP_036119449_1",
    "mbXP_014385142_1",
    "mlXP_023600126_1",
    "mdXP_059516694_1",
    "mmXP_036167556_1",
    "pkXP_045436936_1",
    "pnCAK6435305_1"
  ),
  dN_dS = c(
    0.307,
    0.400,
    0.516,
    0.704,
    0.959,
    0.684,
    1.221,
    0.672,
    1.051,
    0.934,
    0.426,
    0.835,
    0.694,
    0.849,
    0.736,
    1.019,
    1.034
  )
)

df60 <- data.frame(
  Lineage = c(
    "rfXP_032956749_1",
    "paXP_006909524_1",
    "pgXP_039707576_1",
    "pvXP_023377685_1",
    "raXP_016008642_2",
    "sbXP_066120920_1",
    "drXP_024414403_2",
    "ajXP_053520509_1",
    "phXP_045684314_1",
    "pdXP_028370187_1",
    "moXP_036119267_1",
    "pkXP_036291893_1",
    "pnCAK6435304_1",
    "mbXP_014399719_1",
    "mlXP_023600126_1",
    "mdXP_059516697_1",
    "mmXP_036167545_1"
  ),
  dN_dS = c(
    0.334,
    0.313,
    0.503,
    0.553,
    0.814,
    0.742,
    0.815,
    0.497,
    0.785,
    0.587,
    0.427,
    0.215,
    0.195,
    0.163,
    0.131,
    0.227,
    0.361
  )
)

df61 <- data.frame(
  Lineage = c(
    "rfXP_032952381_1",
    "paXP_015455893_1",
    "pgXP_039710411_1",
    "pvXP_011353023_1",
    "raXP_015984227_1",
    "sbXP_066107492_1",
    "drXP_024413494_2",
    "ajXP_053524271_1",
    "phXP_045698747_1",
    "pdXP_035878485_1",
    "moXP_036104243_1",
    "mdXP_059557556_1",
    "mmXP_036172778_1",
    "mbXP_014390861_1",
    "mlXP_006100655_1",
    "pkXP_045429115_1",
    "pnCAK6446945_1"
  ),
  dN_dS = c(
    0.060,
    0.054,
    0.054,
    0.059,
    0.066,
    0.092,
    0.121,
    0.082,
    0.066,
    0.071,
    0.199,
    0.247,
    0.246,
    0.349,
    0.181,
    0.207,
    0.197
  )
)

df62 <- data.frame(
  Lineage = c(
    "rfXP_032952366_1",
    "paXP_015455893_1",
    "pgXP_039710410_1",
    "pvXP_011353022_1",
    "raXP_036074329_1",
    "sbXP_066107492_1",
    "drXP_024413494_2",
    "ajXP_053524271_1",
    "phXP_045698747_1",
    "pdXP_035878485_1",
    "moXP_036104241_1",
    "mdXP_059557556_1",
    "mmXP_036172778_1",
    "mbXP_014390861_1",
    "mlXP_006100653_1",
    "pkXP_045429115_1",
    "pnCAK6446945_1"
  ),
  dN_dS = c(
    0.063,
    0.056,
    0.056,
    0.061,
    0.068,
    0.080,
    0.106,
    0.064,
    0.052,
    0.056,
    0.196,
    0.216,
    0.216,
    0.340,
    0.168,
    0.175,
    0.222
  )
)
#SHOULD BE REMOVED
df63 <- data.frame(
  Lineage = c(
    "rfXP_032977140_1",
    "paXP_006912872_1",
    "pgXP_039743849_1",
    "pvXP_011363695_1",
    "raXP_016004504_2",
    "sbXP_066100890_1",
    "ajXP_037010898_2",
    "phXP_045696164_1",
    "pdXP_035884459_1",
    "drXP_045059922_1",
    "moXP_036114810_1",
    "mbXP_005856915_1",
    "mlXP_023613660_1",
    "mdXP_059563865_1",
    "mmXP_036180265_1",
    "pkXP_036283244_1"
  ),
  dN_dS = c(
    1.521,
    0.460,
    0.552,
    0.496,
    0.493,
    1.012,
    0.705,
    0.720,
    0.766,
    0.750,
    1.505,
    1.235,
    0.933,
    1.091,
    1.415,
    1.077
  )
)

df64 <- data.frame(
  Lineage = c(
    "rfXP_032962316_1",
    "paXP_006907716_1",
    "pgXP_039701389_1",
    "pvXP_011358836_1",
    "raXP_016009489_1",
    "sbXP_066088988_1",
    "drXP_024433061_1",
    "ajXP_053511624_1",
    "phXP_045713867_1",
    "pdXP_035877989_1",
    "moXP_036134616_1",
    "mbXP_014402084_1",
    "mlXP_006094480_1",
    "mdXP_059524609_1",
    "mmXP_036192245_1",
    "pkXP_036269822_1",
    "pnCAK6433074_1"
  ),
  dN_dS = c(
    0.166,
    0.190,
    0.362,
    0.159,
    0.173,
    0.728,
    0.623,
    0.359,
    0.686,
    0.518,
    0.740,
    0.541,
    0.465,
    0.464,
    0.416,
    0.503,
    0.551
  )
)

df65 <- data.frame(
  Lineage = c(
    "rfXP_032962316_1",
    "paXP_006907716_1",
    "pgXP_039701390_1",
    "pvXP_011358836_1",
    "raXP_016009489_1",
    "sbXP_066088988_1",
    "drXP_024433069_1",
    "ajXP_053511624_1",
    "phXP_045713868_1",
    "pdXP_035877998_1",
    "moXP_036134633_1",
    "mbXP_014402084_1",
    "mlXP_014316098_1",
    "mdXP_059524609_1",
    "mmXP_036192254_1",
    "pkXP_036269822_1",
    "pnCAK6433074_1"
  ),
  dN_dS = c(
    0.138,
    0.166,
    0.187,
    0.152,
    0.171,
    1.258,
    0.511,
    0.314,
    0.613,
    0.456,
    0.719,
    0.458,
    0.441,
    0.443,
    0.394,
    0.525,
    0.512
  )
)

df66 <- data.frame(
  Lineage = c(
    "rfXP_032961894_1",
    "paXP_006919081_1",
    "pgXP_039693954_1",
    "pvXP_011372100_1",
    "raXP_015979710_2",
    "sbXP_066088897_1",
    "ajXP_036982711_2",
    "phXP_045690033_1",
    "pdXP_028359644_1",
    "drXP_045038620_2",
    "moXP_036099575_1",
    "mbXP_014388937_1",
    "mlXP_023608941_1",
    "mdXP_059532751_1",
    "mmXP_036183558_1",
    "pkXP_036272999_1",
    "pnCAK6433338_1"
  ),
  dN_dS = c(
    0.105,
    0.116,
    0.117,
    0.126,
    0.126,
    0.073,
    0.123,
    0.101,
    0.110,
    0.124,
    0.095,
    0.225,
    0.079,
    0.092,
    0.074,
    0.068,
    0.087
  )
)

df67 <- data.frame(
  Lineage = c(
    "rfXP_032956859_1",
    "paXP_024900830_1",
    "pgXP_039707589_1",
    "pvXP_023377703_1",
    "raXP_036080948_1",
    "sbXP_066120923_1",
    "ajXP_037006028_2",
    "phXP_045683233_1",
    "pdXP_028370191_2",
    "drXP_024414225_1",
    "moXP_036119267_1",
    "mbXP_005852975_1",
    "mlXP_023600127_1",
    "mdXP_059516697_1",
    "mmXP_036167544_1",
    "pkXP_036291893_1",
    "pnCAK6435304_1"
  ),
  dN_dS = c(
    0.148,
    0.241,
    0.293,
    0.263,
    0.353,
    0.602,
    1.036,
    1.085,
    1.517,
    1.302,
    0.586,
    0.416,
    0.875,
    0.910,
    0.829,
    0.797,
    0.786
  )
)

df68 <- data.frame(
  Lineage = c(
    "rfXP_032987038_1",
    "paXP_006926960_2",
    "pgXP_039706560_1",
    "pvXP_011380217_1",
    "raXP_015991744_2",
    "sbXP_066097826_1",
    "ajXP_053513724_1",
    "phXP_045707470_1",
    "pdXP_035873711_1",
    "drXP_024416671_2",
    "moXP_036098392_1",
    "mbXP_014385872_1",
    "mlXP_023606248_1",
    "mdXP_059543308_1",
    "mmXP_036206078_1",
    "pkXP_036304099_1",
    "pnCAK6441817_1"
  ),
  dN_dS = c(
    0.680,
    1.120,
    1.046,
    0.955,
    0.757,
    0.545,
    0.569,
    0.635,
    0.566,
    0.752,
    0.542,
    0.981,
    0.995,
    1.183,
    1.174,
    0.519,
    0.534
  )
)

df69 <- data.frame(
  Lineage = c(
    "rfXP_032987716_1",
    "paXP_015449250_1",
    "pgXP_039723389_1",
    "pvXP_023378058_1",
    "raXP_015977307_2",
    "sbXP_066100855_1",
    "ajXP_036990612_2",
    "phXP_045688369_1",
    "pdXP_028374138_1",
    "drXP_045048322_2",
    "moXP_036123338_1",
    "mbXP_005857356_1",
    "mlXP_006090905_2",
    "mdXP_059519541_1",
    "mmXP_036187833_1",
    "pkXP_036286781_1",
    "pnCAK6435502_1"
  ),
  dN_dS = c(
    0.213,
    0.174,
    0.156,
    0.174,
    0.235,
    0.130,
    0.140,
    0.089,
    0.101,
    0.116,
    0.160,
    0.287,
    0.282,
    0.230,
    0.262,
    0.207,
    0.212
  )
)

df70 <- data.frame(
  Lineage = c(
    "rfXP_032954711_1",
    "paXP_006910730_1",
    "pgXP_039716603_1",
    "pvXP_011355555_1",
    "raXP_015980632_2",
    "sbXP_066110625_1",
    "ajXP_037023594_2",
    "phXP_045711546_1",
    "pdXP_028380932_1",
    "drXP_024410706_2",
    "moXP_036102222_1",
    "mbXP_005869650_1",
    "mlXP_006090572_2",
    "mdXP_059512037_1",
    "mmXP_036182960_1",
    "pkXP_036285819_1",
    "pnCAK6434282_1"
  ),
  dN_dS = c(
    0.771,
    0.858,
    0.755,
    0.713,
    1.139,
    1.021,
    0.510,
    0.691,
    0.766,
    0.759,
    0.870,
    0.917,
    0.922,
    0.826,
    0.793,
    0.769,
    0.739
  )
)
###############################################################################
# 2) Define a mapping from the two-letter tag to the full bat species name
###############################################################################
species_map <- c(
  "rf" = "Rhinolophus_ferrumequinum",
  "dr" = "Desmodus_rotundus",
  "ra" = "Rousettus_aegyptiacus",
  "pv" = "Pteropus_vampyrus",
  "pg" = "Pteropus_giganteus",
  "pa" = "Pteropus_alecto",
  "pk" = "Pipistrellus_kuhlii",
  "ph" = "Phyllostomus_hastatus",
  "pd" = "Phyllostomus_discolor",
  "mm" = "Myotis_myotis",
  "ml" = "Myotis_lucifugus",
  "mb" = "Myotis_brandtii",
  "mo" = "Molossus_molossus",
  "md" = "Myotis_daubentonii",
  "sb" = "Saccopteryx_bilineata",
  "aj" = "Artibeus_jamaicensis",
  "pn" = "Pipistrellus_nathusii"
)

###############################################################################
# 3) For each data frame (df1 for NP_004322.1 and df2 for P52815.2):
#    - Extract the two-letter code (e.g., "rf", "pa")
#    - Map it to the full species name
#    - Add a "Protein" identifier column
###############################################################################
df1$SpeciesCode  <- substr(df1$Lineage, 1, 2)          # e.g. "rf"
df1$SpeciesName  <- species_map[df1$SpeciesCode]       # e.g. "Rhinolophus_ferrumequinum"
df1$Protein      <- "NP_004322.1"

df2$SpeciesCode  <- substr(df2$Lineage, 1, 2)
df2$SpeciesName  <- species_map[df2$SpeciesCode]
df2$Protein      <- "P52815.2"

df3$SpeciesCode  <- substr(df3$Lineage, 1, 2)
df3$SpeciesName  <- species_map[df3$SpeciesCode]
df3$Protein      <- "NP_478054.2"

df4$SpeciesCode  <- substr(df4$Lineage, 1, 2)
df4$SpeciesName  <- species_map[df4$SpeciesCode]
df4$Protein      <- "Q99836.1"

df5$SpeciesCode  <- substr(df5$Lineage, 1, 2)
df5$SpeciesName  <- species_map[df5$SpeciesCode]
df5$Protein      <- "P47985.2"

df6$SpeciesCode  <- substr(df6$Lineage, 1, 2)
df6$SpeciesName  <- species_map[df6$SpeciesCode]
df6$Protein      <- "Q9BY49.2"

df7$SpeciesCode  <- substr(df7$Lineage, 1, 2)
df7$SpeciesName  <- species_map[df7$SpeciesCode]
df7$Protein      <- "Q9NQL2.1"

df8$SpeciesCode  <- substr(df8$Lineage, 1, 2)
df8$SpeciesName  <- species_map[df8$SpeciesCode]
df8$Protein      <- "P15692.3"

df9$SpeciesCode  <- substr(df9$Lineage, 1, 2)
df9$SpeciesName  <- species_map[df9$SpeciesCode]
df9$Protein      <- "P49023.3"

df10$SpeciesCode  <- substr(df10$Lineage, 1, 2)
df10$SpeciesName  <- species_map[df10$SpeciesCode]
df10$Protein      <- "Q9UN79.3"

df11$SpeciesCode  <- substr(df11$Lineage, 1, 2)
df11$SpeciesName  <- species_map[df11$SpeciesCode]
df11$Protein      <- "P09017.2"

df12$SpeciesCode  <- substr(df12$Lineage, 1, 2)
df12$SpeciesName  <- species_map[df12$SpeciesCode]
df12$Protein      <- "P17023.4"

df13$SpeciesCode  <- substr(df13$Lineage, 1, 2)
df13$SpeciesName  <- species_map[df13$SpeciesCode]
df13$Protein      <- "Q9Y6J9.1"

df14$SpeciesCode  <- substr(df14$Lineage, 1, 2)
df14$SpeciesName  <- species_map[df14$SpeciesCode]
df14$Protein      <- "Q3B726.1"

df15$SpeciesCode  <- substr(df15$Lineage, 1, 2)
df15$SpeciesName  <- species_map[df15$SpeciesCode]
df15$Protein      <- "Q6ZMU5.2"

df16$SpeciesCode  <- substr(df16$Lineage, 1, 2)
df16$SpeciesName  <- species_map[df16$SpeciesCode]
df16$Protein      <- "Q8N6M9.2"

df17$SpeciesCode  <- substr(df17$Lineage, 1, 2)
df17$SpeciesName  <- species_map[df17$SpeciesCode]
df17$Protein      <- "Q9BT67.1"

df18$SpeciesCode  <- substr(df18$Lineage, 1, 2)
df18$SpeciesName  <- species_map[df18$SpeciesCode]
df18$Protein      <- "Q16763.2"

df19$SpeciesCode  <- substr(df19$Lineage, 1, 2)
df19$SpeciesName  <- species_map[df19$SpeciesCode]
df19$Protein      <- "WHV00477.1"

df20$SpeciesCode  <- substr(df20$Lineage, 1, 2)
df20$SpeciesName  <- species_map[df20$SpeciesCode]
df20$Protein      <- "O14649.1"

df21$SpeciesCode  <- substr(df21$Lineage, 1, 2)
df21$SpeciesName  <- species_map[df21$SpeciesCode]
df21$Protein      <- "P07332.3"

df22$SpeciesCode  <- substr(df22$Lineage, 1, 2)
df22$SpeciesName  <- species_map[df22$SpeciesCode]
df22$Protein      <- "P28223.2"

df23$SpeciesCode  <- substr(df23$Lineage, 1, 2)
df23$SpeciesName  <- species_map[df23$SpeciesCode]
df23$Protein      <- "Q02641.3"

df24$SpeciesCode  <- substr(df24$Lineage, 1, 2)
df24$SpeciesName  <- species_map[df24$SpeciesCode]
df24$Protein      <- "P18074.1"

df25$SpeciesCode  <- substr(df25$Lineage, 1, 2)
df25$SpeciesName  <- species_map[df25$SpeciesCode]
df25$Protein      <- "P47974.3"

df26$SpeciesCode  <- substr(df26$Lineage, 1, 2)
df26$SpeciesName  <- species_map[df26$SpeciesCode]
df26$Protein      <- "P56704.2"

df27$SpeciesCode  <- substr(df27$Lineage, 1, 2)
df27$SpeciesName  <- species_map[df27$SpeciesCode]
df27$Protein      <- "Q8N474.1"

df28$SpeciesCode  <- substr(df28$Lineage, 1, 2)
df28$SpeciesName  <- species_map[df28$SpeciesCode]
df28$Protein      <- "Q06945.1"

df29$SpeciesCode  <- substr(df29$Lineage, 1, 2)
df29$SpeciesName  <- species_map[df29$SpeciesCode]
df29$Protein      <- "Q13485.1"

df30$SpeciesCode  <- substr(df30$Lineage, 1, 2)
df30$SpeciesName  <- species_map[df30$SpeciesCode]
df30$Protein      <- "Q9H9B1.4"

df31$SpeciesCode  <- substr(df31$Lineage, 1, 2)
df31$SpeciesName  <- species_map[df31$SpeciesCode]
df31$Protein      <- "P04141.1"

df32$SpeciesCode  <- substr(df32$Lineage, 1, 2)
df32$SpeciesName  <- species_map[df32$SpeciesCode]
df32$Protein      <- "P49589.3"

df33$SpeciesCode  <- substr(df33$Lineage, 1, 2)
df33$SpeciesName  <- species_map[df33$SpeciesCode]
df33$Protein      <- "Q8N1S5.3"

df34$SpeciesCode  <- substr(df34$Lineage, 1, 2)
df34$SpeciesName  <- species_map[df34$SpeciesCode]
df34$Protein      <- "Q9H7L2.2"

df35$SpeciesCode  <- substr(df35$Lineage, 1, 2)
df35$SpeciesName  <- species_map[df35$SpeciesCode]
df35$Protein      <- "NP_068767.3"

df36$SpeciesCode  <- substr(df36$Lineage, 1, 2)
df36$SpeciesName  <- species_map[df36$SpeciesCode]
df36$Protein      <- "NP_597725.1"

df37$SpeciesCode  <- substr(df37$Lineage, 1, 2)
df37$SpeciesName  <- species_map[df37$SpeciesCode]
df37$Protein      <- "NP_940978.2"

df38$SpeciesCode  <- substr(df38$Lineage, 1, 2)
df38$SpeciesName  <- species_map[df38$SpeciesCode]
df38$Protein      <- "NP_001036009.1"

df39$SpeciesCode  <- substr(df39$Lineage, 1, 2)
df39$SpeciesName  <- species_map[df39$SpeciesCode]
df39$Protein      <- "O14653.2"

df40$SpeciesCode  <- substr(df40$Lineage, 1, 2)
df40$SpeciesName  <- species_map[df40$SpeciesCode]
df40$Protein      <- "O14734.1"

df41$SpeciesCode  <- substr(df41$Lineage, 1, 2)
df41$SpeciesName  <- species_map[df41$SpeciesCode]
df41$Protein      <- "P20338.3"

df42$SpeciesCode  <- substr(df42$Lineage, 1, 2)
df42$SpeciesName  <- species_map[df42$SpeciesCode]
df42$Protein      <- "Q9UBP6.1"

df43$SpeciesCode  <- substr(df43$Lineage, 1, 2)
df43$SpeciesName  <- species_map[df43$SpeciesCode]
df43$Protein      <- "Q96HD9.1"

df44$SpeciesCode  <- substr(df44$Lineage, 1, 2)
df44$SpeciesName  <- species_map[df44$SpeciesCode]
df44$Protein      <- "Q92536.3"

df45$SpeciesCode  <- substr(df45$Lineage, 1, 2)
df45$SpeciesName  <- species_map[df45$SpeciesCode]
df45$Protein      <- "NP_000749.2"

df46$SpeciesCode  <- substr(df46$Lineage, 1, 2)
df46$SpeciesName  <- species_map[df46$SpeciesCode]
df46$Protein      <- "NP_001519.1"

df47$SpeciesCode  <- substr(df47$Lineage, 1, 2)
df47$SpeciesName  <- species_map[df47$SpeciesCode]
df47$Protein      <- "NP_002695.1"

df48$SpeciesCode  <- substr(df48$Lineage, 1, 2)
df48$SpeciesName  <- species_map[df48$SpeciesCode]
df48$Protein      <- "NP_003003.3"

df49$SpeciesCode  <- substr(df49$Lineage, 1, 2)
df49$SpeciesName  <- species_map[df49$SpeciesCode]
df49$Protein      <- "NP_003367.4"

df50$SpeciesCode  <- substr(df50$Lineage, 1, 2)
df50$SpeciesName  <- species_map[df50$SpeciesCode]
df50$Protein      <- "NP_055253.2"

df51$SpeciesCode  <- substr(df51$Lineage, 1, 2)
df51$SpeciesName  <- species_map[df51$SpeciesCode]
df51$Protein      <- "NP_055255.1"

df52$SpeciesCode  <- substr(df52$Lineage, 1, 2)
df52$SpeciesName  <- species_map[df52$SpeciesCode]
df52$Protein      <- "NP_079484.1"

df53$SpeciesCode  <- substr(df53$Lineage, 1, 2)
df53$SpeciesName  <- species_map[df53$SpeciesCode]
df53$Protein      <- "NP_148983.1"

df54$SpeciesCode  <- substr(df54$Lineage, 1, 2)
df54$SpeciesName  <- species_map[df54$SpeciesCode]
df54$Protein      <- "NP_149122.1"

df55$SpeciesCode  <- substr(df55$Lineage, 1, 2)
df55$SpeciesName  <- species_map[df55$SpeciesCode]
df55$Protein      <- "NP_149126.1"

df56$SpeciesCode  <- substr(df56$Lineage, 1, 2)
df56$SpeciesName  <- species_map[df56$SpeciesCode]
df56$Protein      <- "NP_665905.2"

df57$SpeciesCode  <- substr(df57$Lineage, 1, 2)
df57$SpeciesName  <- species_map[df57$SpeciesCode]
df57$Protein      <- "NP_775270.1"

df58$SpeciesCode  <- substr(df58$Lineage, 1, 2)
df58$SpeciesName  <- species_map[df58$SpeciesCode]
df58$Protein      <- "NP_775296.1"

df59$SpeciesCode  <- substr(df59$Lineage, 1, 2)
df59$SpeciesName  <- species_map[df59$SpeciesCode]
df59$Protein      <- "NP_775297.1"

df60$SpeciesCode  <- substr(df60$Lineage, 1, 2)
df60$SpeciesName  <- species_map[df60$SpeciesCode]
df60$Protein      <- "NP_001020538.2"

# this one has +7 isoforms
df61$SpeciesCode  <- substr(df61$Lineage, 1, 2)
df61$SpeciesName  <- species_map[df61$SpeciesCode]
df61$Protein      <- "NP_001020539.2"
# this one has +4 isoforms
df62$SpeciesCode  <- substr(df62$Lineage, 1, 2)
df62$SpeciesName  <- species_map[df62$SpeciesCode]
df62$Protein      <- "NP_001165095.1"

df63$SpeciesCode  <- substr(df63$Lineage, 1, 2)
df63$SpeciesName  <- species_map[df63$SpeciesCode]
df63$Protein      <- "NP_001230140.1"

df64$SpeciesCode  <- substr(df64$Lineage, 1, 2)
df64$SpeciesName  <- species_map[df64$SpeciesCode]
df64$Protein      <- "NP_001257920.1"

df65$SpeciesCode  <- substr(df65$Lineage, 1, 2)
df65$SpeciesName  <- species_map[df65$SpeciesCode]
df65$Protein      <- "NP_001257922.1"

df66$SpeciesCode  <- substr(df66$Lineage, 1, 2)
df66$SpeciesName  <- species_map[df66$SpeciesCode]
df66$Protein      <- "XP_005248023.1"

df67$SpeciesCode  <- substr(df67$Lineage, 1, 2)
df67$SpeciesName  <- species_map[df67$SpeciesCode]
df67$Protein      <- "XP_005263696.1"
# here ecmreg start
df68$SpeciesCode  <- substr(df68$Lineage, 1, 2)
df68$SpeciesName  <- species_map[df68$SpeciesCode]
df68$Protein      <- "NP_000884.1"

df69$SpeciesCode  <- substr(df69$Lineage, 1, 2)
df69$SpeciesName  <- species_map[df69$SpeciesCode]
df69$Protein      <- "NP_002209.2"
# this one has +4 isoforms
df70$SpeciesCode  <- substr(df70$Lineage, 1, 2)
df70$SpeciesName  <- species_map[df70$SpeciesCode]
df70$Protein      <- "NP_003108.2"
###############################################################################
# 4) Reshape each data frame from "long" (one row per lineage/species) to "wide"
#    We want species as columns, and the row to represent each protein.
#    We'll use tidyr::pivot_wider for convenience.
###############################################################################
library(tidyr)
library(dplyr)

df1_wide <- df1 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(
    names_from  = SpeciesName, 
    values_from = dN_dS
  )

df2_wide <- df2 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(
    names_from  = SpeciesName, 
    values_from = dN_dS
  )

df3_wide <- df3 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(
    names_from  = SpeciesName, 
    values_from = dN_dS
  )

df4_wide <- df4 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(
    names_from  = SpeciesName, 
    values_from = dN_dS
  )

df5_wide <- df5 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(
    names_from  = SpeciesName, 
    values_from = dN_dS
  )

df6_wide <- df6 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(
    names_from  = SpeciesName, 
    values_from = dN_dS
  )

df7_wide <- df7 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(
    names_from  = SpeciesName, 
    values_from = dN_dS
  )

df8_wide <- df8 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(
    names_from  = SpeciesName, 
    values_from = dN_dS
  )

df9_wide <- df9 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(
    names_from  = SpeciesName, 
    values_from = dN_dS
  )

df10_wide <- df10 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(
    names_from  = SpeciesName, 
    values_from = dN_dS
  )

df11_wide <- df11 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(
    names_from  = SpeciesName, 
    values_from = dN_dS
  )

df12_wide <- df12 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(
    names_from  = SpeciesName, 
    values_from = dN_dS
  )

df13_wide <- df13 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(
    names_from  = SpeciesName, 
    values_from = dN_dS
  )

df14_wide <- df14 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(
    names_from  = SpeciesName, 
    values_from = dN_dS
  )

df15_wide <- df15 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(
    names_from  = SpeciesName, 
    values_from = dN_dS
  )

df16_wide <- df16 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df17_wide <- df17 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df18_wide <- df18 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df19_wide <- df19 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df20_wide <- df20 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df21_wide <- df21 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df22_wide <- df22 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df23_wide <- df23 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df24_wide <- df24 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df25_wide <- df25 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df26_wide <- df26 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df27_wide <- df27 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df28_wide <- df28 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df29_wide <- df29 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df30_wide <- df30 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df31_wide <- df31 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df32_wide <- df32 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df33_wide <- df33 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df34_wide <- df34 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df35_wide <- df35 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df36_wide <- df36 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df37_wide <- df37 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df38_wide <- df38 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df39_wide <- df39 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df40_wide <- df40 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df41_wide <- df41 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df42_wide <- df42 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df43_wide <- df43 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

df44_wide <- df44 %>%
  select(Protein, SpeciesName, dN_dS) %>%
  tidyr::pivot_wider(names_from = SpeciesName, values_from = dN_dS)

library(dplyr)

for (i in 45:70) {
  df_name <- paste0("df", i)
  df_wide_name <- paste0("df", i, "_wide")
  
  assign(df_wide_name, get(df_name) %>%
           select(Protein, SpeciesName, dN_dS) %>%
           pivot_wider(names_from = SpeciesName, values_from = dN_dS))
}
###############################################################################
# 5) Combine the two wide data sets by stacking them row-wise
###############################################################################
selected_indices <- c(1:33, 35:62, 64:70)
df_names <- paste0("df", selected_indices, "_wide")
#34 and 63 should be removed
final_df <- dplyr::bind_rows(mget(df_names))
###############################################################################
# 6) final_df is now an R data set with:
#    - 2 rows (one for NP_004322.1 and one for P52815.2)
#    - 17 columns (one for each species, plus the Protein column)
###############################################################################
final_df
###############################################################################
# Suppose 'final_df' looks like this (with 2 rows: NP_004322.1, P52815.2):
#   Protein  Rhinolophus_ferrumequinum Desmodus_rotundus ... Pipistrellus_nathusii
# 1 NP_004322.1              0.038                0.075   ...          0.052
# 2 P52815.2                 0.097                0.052   ...          0.025
#
# We'll now add a 3rd row (LQ) with these species traits:
#
# Rhinolophus_ferrumequinum = 4.7
# Desmodus_rotundus         = 4.3
# Rousettus_aegyptiacus     = 2.6
# Pteropus_vampyrus         = 1.9
# Pteropus_giganteus        = 2.9
# Pteropus_alecto           = 1.6
# Pipistrellus_kuhlii       = 1.6
# Phyllostomus_hastatus     = 2.7
# Phyllostomus_discolor     = 2.0
# Myotis_myotis             = 5.5
# Myotis_lucifugus          = 6.0
# Myotis_brandtii           = 7.8
# Molossus_molossus         = 0.9
# Myotis_daubentonii        = 5.2
# Saccopteryx_bilineata     = 1.1
# Artibeus_jamaicensis      = 2.6
# Pipistrellus_nathusii     = 2.0
###############################################################################


# Create a single-row data frame with LQ for each species
LQ_row <- tibble(
  Protein = "LQ",
  Rhinolophus_ferrumequinum = 4.7,
  Desmodus_rotundus         = 4.3,
  Rousettus_aegyptiacus     = 2.6,
  Pteropus_vampyrus         = 1.9,
  Pteropus_giganteus        = 2.9,
  Pteropus_alecto           = 1.6,
  Pipistrellus_kuhlii       = 1.6,
  Phyllostomus_hastatus     = 2.7,
  Phyllostomus_discolor     = 2.0,
  Myotis_myotis             = 5.5,
  Myotis_lucifugus          = 6.0,
  Myotis_brandtii           = 7.8,
  Molossus_molossus         = 0.9,
  Myotis_daubentonii        = 5.2,
  Saccopteryx_bilineata     = 1.1,
  Artibeus_jamaicensis      = 2.6,
  Pipistrellus_nathusii     = 2.0
)

# Bind the LQ row onto the existing final_df
final_df_plus_LQ <- bind_rows(final_df, LQ_row)

# Take a look at the new data frame
final_df_plus_LQ

MaxLifespan_row <- tibble(
  Protein = "Maximum_Lifespan",
  Rhinolophus_ferrumequinum = 30.5,
  Desmodus_rotundus         = 29.2,
  Rousettus_aegyptiacus     = 22.9,
  Pteropus_vampyrus         = 20.9,
  Pteropus_giganteus        = 44,
  Pteropus_alecto           = 20.3,
  Pipistrellus_kuhlii       = 8,
  Phyllostomus_hastatus     = 18,
  Phyllostomus_discolor     = 9,
  Myotis_myotis             = 37.1,
  Myotis_lucifugus          = 34,
  Myotis_brandtii           = 41,
  Molossus_molossus         = 5.6,
  Myotis_daubentonii        = 28,
  Saccopteryx_bilineata     = 6,
  Artibeus_jamaicensis      = 19.2,
  Pipistrellus_nathusii     = 11
)

# 2) Create a row for Weight using your real values (placeholder below)
Weight_row <- tibble(
  Protein = "Weight",
  Rhinolophus_ferrumequinum = 22.9,
  Desmodus_rotundus         = 33.2,
  Rousettus_aegyptiacus     = 125,
  Pteropus_vampyrus         = 872,
  Pteropus_giganteus        = 1175,
  Pteropus_alecto           = 672,
  Pipistrellus_kuhlii       = 6,
  Phyllostomus_hastatus     = 81,
  Phyllostomus_discolor     = 43.4,
  Myotis_myotis             = 28.6,
  Myotis_lucifugus          = 10,
  Myotis_brandtii           = 7,
  Molossus_molossus         = 14,
  Myotis_daubentonii        = 8.5,
  Saccopteryx_bilineata     = 7.9,
  Artibeus_jamaicensis      = 42.4,
  Pipistrellus_nathusii     = 10
)

# 3) Append the new rows to the existing data frame
final_df_plus_all <- bind_rows(final_df_plus_LQ, MaxLifespan_row, Weight_row)

# 4) Inspect the new data set
final_df_plus_all

long_data <- final_df_plus_LQ %>%
  pivot_longer(
    cols = -Protein,            # Everything except the 'Protein' column
    names_to = "Species",       # The column names become a "Species" variable
    values_to = "value"         # The cell contents go into "value"
  )

# Step B: Pivot wider so that each Protein becomes a column
# We only do this if we specifically want each "Protein" or trait to be in a column. 
wide_data <- long_data %>%
  pivot_wider(
    names_from = Protein,   # e.g. "NP_004322.1", "P52815.2", "LQ", ...
    values_from = value
  )

# wide_data now has columns: 
# Species, NP_004322.1, P52815.2, LQ, (plus Maximum_Lifespan, Weight, etc. if present).
# Each row is a species. 
# Let's view it:
wide_data$Hibernation <- c(1, 0, 0, 0, 0, 0, 0, 0, 0,0,0,1,1,1,1,0,0)
wide_data

#VERSION WITHOUT MYOTIS BRANDTII | comment if does not work out
wide_data = wide_data[-12, ]
wide_data
library(ape)
# Load or create your phylogeny with matching tip labels
myPhylogeny <- read.tree("bat_tags_noen.nwk")

# (Optional) Inspect tip labels
myPhylogeny$tip.label
library(caper)
comp_dat <- comparative.data(
  phy = myPhylogeny,
  data = as.data.frame(wide_data),  # wide_data must be a data.frame (not just tibble)
  names.col = "Species",           # which column has species names?
  vcv = TRUE,                      # whether to store the variancecovariance matrix
  warn.dropped = TRUE             # whether to warn if some species are dropped
)
comp_dat$phy$tip.label
comp_dat$data$Species
'''model_pgls <- pgls(
  #formula = LQ ~ NP_004322.1 + P52815.2+NP_478054.2 + Q99836.1 + P47985.2+ Q9BY49.2 + Q9NQL2.1 +P15692.3 +P49023.3 + Q9UN79.3 + P09017.2 + P17023.4 +Q9Y6J9.1 + Q3B726.1 +  Q6ZMU5.2 + Q8N6M9.2 + Q9BT67.1 +Q16763.2 + WHV00477.1 +O14649.1 + P07332.3 + P28223.2 +Q02641.3 + P18074.1 + P47974.3 + P56704.2 + Q8N474.1 + Q06945.1 + Q13485.1 + Q9H9B1.4,  # Y ~ X1 + X2
  formula = LQ ~ NP_004322.1 + P52815.2+NP_478054.2 + Q99836.1 + P47985.2+ Q9BY49.2 + Q9NQL2.1 +P15692.3 +P49023.3 + Q9UN79.3 + P09017.2 + P17023.4 +Q9Y6J9.1 + Q3B726.1 +  Q6ZMU5.2 + Q8N6M9.2 + Q9BT67.1 +Q16763.2 + WHV00477.1 +O14649.1 + P07332.3 + P28223.2 +Q02641.3 + P18074.1 + P47974.3 + P56704.2 + Q8N474.1 + Q06945.1 + Q13485.1 + Q9H9B1.4 + P04141.1 + P49589.3 + Q8N1S5.3+ Q9H7L2.2 + NP_068767.3 + NP_597725.1 + NP_940978.2 + NP_001036009.1 + O14653.2 + O14734.1 + P20338.3 + Q9UBP6.1 + Q96HD9.1 + Q92536.3 ,  # Y ~ X1 + X2
  
  data = comp_dat
) '''
sigonly_pgls <- pgls(
  formula = LQ ~ P56704.2 + P09017.2 + Q3B726.1 + Q6ZMU5.2 + Q9BT67.1 + P15692.3 + Q9Y6J9.1 + Q99836.1 + NP_004322.1 + Q9H9B1.4,
  data = comp_dat
)

summary(sigonly_pgls)

sigonly1_pgls <- pgls(
  formula = LQ ~ P56704.2 + Q3B726.1 + Q9BT67.1 + P15692.3 + Q9Y6J9.1 + NP_004322.1 + Q9H9B1.4,
  data = comp_dat
)

summary(sigonly1_pgls)
library(lmtest)
bptest(sigonly1_pgls)
sigtru_pgls <- pgls(
  formula = LQ ~ P56704.2 + P09017.2 + Q99836.1 + NP_004322.1 + Q9H9B1.4 ,
  data = comp_dat
)
summary(sigtru_pgls)

head(wide_data)

# Remove non-numeric columns (like Species)
# Option 1: Create a data frame with all numeric columns
numeric_data <- wide_data[, sapply(wide_data, is.numeric)]

# Option 2: Exclude the target variable "LQ" if you only want the predictors
predictor_data <- numeric_data[, !names(numeric_data) %in% "LQ"]

# Compute the correlation matrix for the predictors
cor_mat <- cor(predictor_data, use = "pairwise.complete.obs")
#install.packages("corrplot")
library(corrplot)
# Visualize the correlation matrix as a heatmap with corrplot
corrplot(cor_mat, method = "color", type = "upper", order = "hclust", 
         tl.col = "black", tl.srt = 45)

threshold <- 0.75

# Find indices where the absolute correlation is greater than the threshold.
# We use 'which' with arr.ind = TRUE and then filter out duplicate pairs (only consider upper triangle)
high_cor_indices <- which(abs(cor_mat) > threshold, arr.ind = TRUE)
high_cor_indices <- high_cor_indices[high_cor_indices[,1] < high_cor_indices[,2], ]

# Create a data frame with the gene pairs and their correlation values
gene_pairs <- data.frame(
  Gene1 = rownames(cor_mat)[high_cor_indices[,1]],
  Gene2 = colnames(cor_mat)[high_cor_indices[,2]],
  Correlation = cor_mat[high_cor_indices]
)

# Print the result
print(gene_pairs)


#summary(model_pgls)
# LQ ~ NP_004322.1
model1 <- pgls(LQ ~ NP_004322.1, data = comp_dat)
summary(model1)
bptest(model1)

# LQ ~ P52815.2
model2 <- pgls(LQ ~ P52815.2, data = comp_dat)
summary(model2)
bptest(model2)

model3 <- pgls(LQ ~ NP_478054.2, data = comp_dat)
summary(model3)

model4 <- pgls(LQ ~ Q99836.1, data = comp_dat)
summary(model4)
bptest(model4)
model5 <- pgls(LQ ~ P47985.2, data = comp_dat)
summary(model5)

model6 <- pgls(LQ ~ Q9BY49.2, data = comp_dat)
summary(model6)

model7 <- pgls(LQ ~ Q9NQL2.1, data = comp_dat)
summary(model7)

model8 <- pgls(LQ ~ P15692.3, data = comp_dat)
summary(model8)
library(lmtest)
bptest(model8)


model9 <- pgls(LQ ~ P49023.3, data = comp_dat)
summary(model9)

model10 <- pgls(LQ ~ Q9UN79.3, data = comp_dat)
summary(model10)

model11 <- pgls(LQ ~ P09017.2, data = comp_dat)
summary(model11)

model12 <- pgls(LQ ~ P17023.4, data = comp_dat)
summary(model12)

model13 <- pgls(LQ ~ Q9Y6J9.1, data = comp_dat)
summary(model13)
bptest(model13)

model14 <- pgls(LQ ~ Q3B726.1, data = comp_dat)
summary(model14)
bptest(model14)
model15 <- pgls(LQ ~ Q6ZMU5.2, data = comp_dat)
summary(model15)
bptest(model15)
model16 <- pgls(LQ ~ Q8N6M9.2, data = comp_dat)
summary(model16)

model17 <- pgls(LQ ~ Q9BT67.1, data = comp_dat)
summary(model17)
bptest(model17)

model18 <- pgls(LQ ~ Q16763.2, data = comp_dat)
summary(model18)

model19 <- pgls(LQ ~ WHV00477.1, data = comp_dat)
summary(model19)

model20 <- pgls(LQ ~ O14649.1, data = comp_dat)
summary(model20)

model21 <- pgls(LQ ~ P07332.3, data = comp_dat)
summary(model21)

model22 <- pgls(LQ ~ P28223.2, data = comp_dat)
summary(model22)

model23 <- pgls(LQ ~ Q02641.3, data = comp_dat)
summary(model23)

model24 <- pgls(LQ ~ P18074.1, data = comp_dat)
summary(model24)

model25 <- pgls(LQ ~ P47974.3, data = comp_dat)
summary(model25)

model26 <- pgls(LQ ~ P56704.2, data = comp_dat)
summary(model26)
bptest(model26)

model27 <- pgls(LQ ~ Q8N474.1, data = comp_dat)
summary(model27)

model28 <- pgls(LQ ~ Q06945.1, data = comp_dat)
summary(model28)

model29 <- pgls(LQ ~ Q13485.1, data = comp_dat)
summary(model29)

model30 <- pgls(LQ ~ Q9H9B1.4, data = comp_dat)
summary(model30)
bptest(model30)

model31 <- pgls(LQ ~ P04141.1, data = comp_dat)
summary(model31)

model32 <- pgls(LQ ~ P49589.3, data = comp_dat)
summary(model32)

model33 <- pgls(LQ ~ Q8N1S5.3, data = comp_dat)
summary(model33)

#model34 <- pgls(LQ ~ Q9H7L2.2, data = comp_dat)
#summary(model34)

model35 <- pgls(LQ ~ NP_068767.3, data = comp_dat)
summary(model35)

model36 <- pgls(LQ ~ NP_597725.1, data = comp_dat)
summary(model36)

model37 <- pgls(LQ ~ NP_940978.2, data = comp_dat)
summary(model37)
bptest(model37)
model38 <- pgls(LQ ~ NP_001036009.1, data = comp_dat)
summary(model38)

model39 <- pgls(LQ ~ O14653.2, data = comp_dat)
summary(model39)
bptest(model39)

model40 <- pgls(LQ ~ O14734.1, data = comp_dat)
summary(model40)

model41 <- pgls(LQ ~ P20338.3, data = comp_dat)
summary(model41)

model42 <- pgls(LQ ~ Q9UBP6.1, data = comp_dat)
summary(model42)

model43 <- pgls(LQ ~ Q96HD9.1, data = comp_dat)
summary(model43)

model44 <- pgls(LQ ~ Q92536.3, data = comp_dat)
summary(model44)

modeln <- pgls(LQ ~ NP_000884.1, data = comp_dat)
summary(modeln)

modeln1 <- pgls(LQ ~ NP_002209.2, data = comp_dat)
summary(modeln1)
library(lmtest)
modelSIGWOBR <- pgls(LQ ~ NP_000884.1 + NP_068767.3 +Q96HD9.1 + Q9H9B1.4, data = comp_dat)
summary(modelSIGWOBR)
bptest(modelSIGWOBR)
modelsigonly <- pgls(LQ ~ P56704.2 + NP_001165095.1 + XP_005248023.1+ P09017.2 + P15692.3 + NP_149122.1 + NP_004322.1 + NP_000884.1 + NP_068767.3 + Q9H9B1.4 + NP_940978.2 + NP_001519.1 + NP_003367.4 , data = comp_dat)
summary(modelsigonly) 
modelecmonly <- pgls(LQ ~ NP_940978.2 +NP_002209.2 + NP_068767.3 + NP_000884.1 + P56704.2 + NP_001165095.1 + XP_005248023.1 , data = comp_dat)
summary(modelecmonly)
bptest(modelecmonly)
modelecmonly <- pgls(LQ ~  NP_940978.2 +NP_002209.2 + NP_068767.3 + NP_000884.1 , data = comp_dat)
summary(modelecmonly)
bptest(modelecmonly)
protein_ids <- c(
  "NP_000749.2", "NP_001519.1", "NP_002695.1", "NP_003003.3", "NP_003367.4",
  "NP_055253.2", "NP_055255.1", "NP_079484.1", "NP_148983.1", "NP_149122.1",
  "NP_149126.1", "NP_665905.2", "NP_775270.1", "NP_775296.1", "NP_775297.1",
  "NP_001020538.2", "NP_001020539.2", "NP_001165095.1",
  "NP_001257920.1", "NP_001257922.1", "XP_005248023.1", "XP_005263696.1",
  "NP_000884.1", "NP_002209.2", "NP_003108.2"
)

# Generate models dynamically
for (i in seq_along(protein_ids)) {
  model_name <- paste0("model", i + 44)  # starting from model45
  formula <- as.formula(paste0("LQ ~ `", protein_ids[i], "`"))
  assign(model_name, pgls(formula, data = comp_dat))
  print(model_name)
  print(summary(get(model_name)))
}

summary(model46)
bptest(model46)
summary(model49)
bptest(model49)
summary(model54)
bptest(model54)
summary(model62)
bptest(model62)
summary(model61)
bptest(model61)
summary(model65)
bptest(model65)
summary(model39) #membrin
bptest(model39)
summary(model43)
bptest(model43)

significant <- c(1, 8, 9, 11, 29, 30, 35, 37, 46, 49, 54, 62, 65, 67)
#without brandtii 30 (marg sig), 35, 67
marginally_significant <-c( 4, 13, 14, 15, 17, 39, 43, 61, 68 )
#without brandtii 43 (sig)
#install.packages('lmtest')
library(lmtest)
bptest(model35)
bptest(modeln)
bptest(modeln1)
#model <- lm(y ~ x1 + x2, data = your_data)
for (i in significant){
  model_name <- paste0('model',i)
  model_obj <- get(model_name)
  print(model_name)
  print(summary(model_obj))
  print(bptest(model_obj))
}
for (i in marginally_significant){
  model_name <- paste0('model',i)
  model_obj <- get(model_name)
  print(model_name)
  print(summary(model_obj))
  print(bptest(model_obj))
}

#install.packages('ggplot2')
library(ggplot2)
wide_data_df <- as.data.frame(wide_data)

# Make the comparative data backup just in case
comp_dat <- comparative.data(
  phy = myPhylogeny,
  data = wide_data_df,
  names.col = "Species",  
  vcv = TRUE,
  warn.dropped = TRUE
)
wide_data_df$pred_model_1 <- predict(model1, newdata = wide_data_df)
wide_data_df$pred_model_2 <- predict(model2, newdata = wide_data_df)
wide_data_df$pred_model_3 <- predict(model3, newdata = wide_data_df)
wide_data_df$pred_model_4 <- predict(model4, newdata = wide_data_df)
wide_data_df$pred_model_8 <- predict(model8, newdata = wide_data_df)
wide_data_df$pred_model_11 <- predict(model11, newdata = wide_data_df)
wide_data_df$pred_model_13 <- predict(model13, newdata = wide_data_df)
wide_data_df$pred_model_14 <- predict(model14, newdata = wide_data_df)
wide_data_df$pred_model_15 <- predict(model15, newdata = wide_data_df)
wide_data_df$pred_model_17 <- predict(model17, newdata = wide_data_df)
wide_data_df$pred_model_26 <- predict(model26, newdata = wide_data_df)
wide_data_df$pred_model_30 <- predict(model30, newdata = wide_data_df)
wide_data_df$pred_model_31 <- predict(model31, newdata = wide_data_df)
wide_data_df$pred_model_35 <- predict(model35, newdata = wide_data_df)
wide_data_df$pred_model_37 <- predict(model37, newdata = wide_data_df)
wide_data_df$pred_model_38 <- predict(model38, newdata = wide_data_df)
wide_data_df$pred_model_39 <- predict(model39, newdata = wide_data_df)
wide_data_df$pred_model_43 <- predict(model43, newdata = wide_data_df)
wide_data_df$pred_model_46 <- predict(model46, newdata = wide_data_df)
wide_data_df$pred_model_49 <- predict(model49, newdata = wide_data_df)
wide_data_df$pred_model_54 <- predict(model54, newdata = wide_data_df)
wide_data_df$pred_model_62 <- predict(model62, newdata = wide_data_df)
wide_data_df$pred_model_61 <- predict(model61, newdata = wide_data_df)
wide_data_df$pred_model_65 <- predict(model65, newdata = wide_data_df)
wide_data_df$pred_model_67 <- predict(model67, newdata = wide_data_df)
wide_data_df$pred_model_68 <- predict(model68, newdata = wide_data_df)
wide_data_df$pred_model_ecm <- predict(modelecmonly, newdata = wide_data_df)
wide_data_df$pred_model_sig <- predict(sigonly1_pgls, newdata = wide_data_df)

# leave one out cross val
library(caper)

# Extract species names
species_list <- wide_data_df$Species
n_species <- length(species_list)
n_predictors <- 7  # Number of predictors in the model

# Create a data frame to store the results
results_df <- data.frame(
  Species = character(n_species),
  Absolute_Error = numeric(n_species),
  Test_R2 = numeric(n_species),
  Train_R2 = numeric(n_species),
  Adjusted_Train_R2 = numeric(n_species),
  Log_Loss = numeric(n_species),
  stringsAsFactors = FALSE
)

# Run LOOCV
for (i in 1:n_species) {
  # Exclude one species for validation
  train_data <- wide_data_df[-i, ]
  test_data <- wide_data_df[i, , drop = FALSE]
  
  # Prepare the training data for PGLS
  comp_dat_train <- comparative.data(
    phy = myPhylogeny,
    data = train_data,
    names.col = "Species",
    vcv = TRUE,
    warn.dropped = TRUE
  )
  
  # Fit the PGLS model
  pgls_model <- pgls(
    formula = LQ ~ P56704.2 + Q3B726.1 + Q9BT67.1 + P15692.3 + Q9Y6J9.1 + NP_004322.1 + Q9H9B1.4,
    data = comp_dat_train
  )
  
  # Predict for the left-out species
  pred_test <- predict(pgls_model, newdata = test_data)
  
  # Predict for the training species
  pred_train <- predict(pgls_model, newdata = train_data)
  
  # Calculate MAE for the test species
  mae <- abs(test_data$LQ - pred_test)
  
  # Calculate R^2 for the test species
  actual_test <- test_data$LQ
  mean_actual_test <- mean(train_data$LQ)
  ss_total_test <- sum((actual_test - mean_actual_test)^2)
  ss_residual_test <- sum((actual_test - pred_test)^2)
  r_squared_test <- 1 - (ss_residual_test / ss_total_test)
  
  # Calculate R^2 for the training species
  actual_train <- train_data$LQ
  mean_actual_train <- mean(actual_train)
  ss_total_train <- sum((actual_train - mean_actual_train)^2)
  ss_residual_train <- sum((actual_train - pred_train)^2)
  r_squared_train <- 1 - (ss_residual_train / ss_total_train)
  
  # Calculate Adjusted R^2 for the training species
  n_train <- nrow(train_data)
  adj_r_squared_train <- 1 - ( (1 - r_squared_train) * (n_train - 1) / (n_train - n_predictors - 1) )
  
  # Calculate log-loss (assuming normal distribution)
  epsilon <- 1e-15  # small value to avoid log(0)
  log_loss <- -sum(log(pmax(epsilon, dnorm(actual_test, mean = pred_test, sd = sd(train_data$LQ)))))
  
  # Store the results
  results_df[i, ] <- c(
    as.character(test_data$Species),
    mae,
    r_squared_test,
    r_squared_train,
    adj_r_squared_train,
    log_loss
  )
}

# Display the results
library(knitr)
kable(results_df, format = "markdown")

kable(results1_df, format="markdown")
filtered_results_df <- results_df[
  !(results_df$Species %in% c("Myotis_brandtii", "Molossus_molossus")),
  c("Species", "Absolute_Error", "Adjusted_Train_R2")
]
rownames(filtered_results_df) <- NULL
kable(filtered_results_df, format="markdown")
filtered_results_df$Adjusted_Train_R2 <- as.numeric(filtered_results_df$Adjusted_Train_R2)
filtered_results_df$Absolute_Error <- as.numeric(filtered_results_df$Absolute_Error)
mean(filtered_results_df$Absolute_Error)
summary(filtered_results_df)

wide_data_df <- wide_data_df[order(wide_data_df$Q3B726.1), ]

ggplot(wide_data_df, aes(x = pred_model_ecm, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_ecm), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ ECM only",
    x = "Predicted LQ",
    y = "LQ"
  ) +
  theme_minimal() + coord_cartesian(clip = "off")
summary(modelecmonly)

ggplot(wide_data_df, aes(x = pred_model_sig, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_sig), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ sig only",
    x = "Predicted LQ",
    y = "LQ"
  ) +
  theme_minimal() + coord_cartesian(clip = "off")
summary(sigonly1_pgls)
bptest(sigonly1_pgls)
summary(modelSIGWOBR)

ggplot(wide_data_df, aes(x = Q3B726.1, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_14), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ Q3B726.1",
    x = "Q3B726.1 dN/dS",
    y = "LQ"
  ) +
  theme_minimal() + coord_cartesian(clip = "off")

wide_data_df <- wide_data_df[order(wide_data_df$NP_001519.1), ]
ggplot(wide_data_df, aes(x = NP_001519.1, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_46), color = "blue", size = 1) +
  labs(
    title = "PGLS: LQ ~ NP_001519.1",
    x = "NP_001519.1 dN/dS",
    y = "LQ"
  ) +
  theme_minimal()

wide_data_df <- wide_data_df[order(wide_data_df$NP_003367.4), ]
ggplot(wide_data_df, aes(x = NP_003367.4, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_49), color = "blue", size = 1) +
  labs(
    title = "PGLS: LQ ~ NP_003367.4",
    x = "NP_003367.4 dN/dS",
    y = "LQ"
  ) +
  theme_minimal()

wide_data_df <- wide_data_df[order(wide_data_df$NP_149122.1), ]
ggplot(wide_data_df, aes(x = NP_149122.1, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_54), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ NP_149122.1",
    x = "NP_149122.1 dN/dS",
    y = "LQ"
  ) +
  theme_minimal() + coord_cartesian(clip = "off")

wide_data_df <- wide_data_df[order(wide_data_df$NP_001020539.2), ]
ggplot(wide_data_df, aes(x = NP_001020539.2, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_61), color = "blue", size = 1) +
  labs(
    title = "PGLS: LQ ~ NP_001020539.2",
    x = "NP_001020539.2 dN/dS",
    y = "LQ"
  ) +
  theme_minimal()

wide_data_df <- wide_data_df[order(wide_data_df$NP_001165095.1), ]
ggplot(wide_data_df, aes(x = NP_001165095.1, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_62), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ NP_001165095.1",
    x = "NP_001165095.1 dN/dS",
    y = "LQ"
  ) +
  theme_minimal() + coord_cartesian(clip = "off")

summary(model65)

wide_data_df <- wide_data_df[order(wide_data_df$XP_005248023.1), ]
ggplot(wide_data_df, aes(x = XP_005248023.1, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_65), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ XP_005248023.1",
    x = "XP_005248023.1 dN/dS",
    y = "LQ"
  ) +
  theme_minimal() + coord_cartesian(clip = "off")

summary(model67)

wide_data_df <- wide_data_df[order(wide_data_df$NP_000884.1), ]
ggplot(wide_data_df, aes(x = NP_000884.1, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_67), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ NP_000884.1",
    x = "NP_000884.1 dN/dS",
    y = "LQ"
  ) +
  theme_minimal() + coord_cartesian(clip = "off")

summary(model68)

wide_data_df <- wide_data_df[order(wide_data_df$NP_002209.2), ]
ggplot(wide_data_df, aes(x = NP_002209.2, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_68), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ NP_002209.2",
    x = "NP_002209.2 dN/dS",
    y = "LQ"
  ) +
  theme_minimal() + coord_cartesian(clip = "off")


wide_data_df <- wide_data_df[order(wide_data_df$O14653.2), ]
ggplot(wide_data_df, aes(x = O14653.2, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_39), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ O14653.2",
    x = "O14653.2 dN/dS",
    y = "LQ"
  ) +
  theme_minimal()+ coord_cartesian(clip = "off")

wide_data_df <- wide_data_df[order(wide_data_df$Q96HD9.1), ]
ggplot(wide_data_df, aes(x = Q96HD9.1, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_43), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ Q96HD9.1",
    x = "Q96HD9.1 dN/dS",
    y = "LQ"
  ) +
  theme_minimal() + coord_cartesian(clip = "off")

summary(model43)

wide_data_df <- wide_data_df[order(wide_data_df$NP_940978.2), ]
ggplot(wide_data_df, aes(x = NP_940978.2, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_37), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ NP_940978.2",
    x = "NP_940978.2 dN/dS",
    y = "LQ"
  ) +
  theme_minimal() + coord_cartesian(clip = "off")

wide_data_df <- wide_data_df[order(wide_data_df$NP_001036009.1), ]
ggplot(wide_data_df, aes(x = NP_001036009.1, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_38), color = "blue", size = 1) +
  labs(
    title = "PGLS: LQ ~ NP_001036009.1",
    x = "NP_001036009.1 dN/dS",
    y = "LQ"
  ) +
  theme_minimal()

wide_data_df <- wide_data_df[order(wide_data_df$NP_068767.3), ]
ggplot(wide_data_df, aes(x = NP_068767.3, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_35), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ NP_068767.3",
    x = "NP_068767.3 dN/dS",
    y = "LQ"
  ) +
  theme_minimal() + coord_cartesian(clip = "off")
summary(model35)
wide_data_df <- wide_data_df[order(wide_data_df$P04141.1), ]
ggplot(wide_data_df, aes(x = P04141.1, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_31), color = "blue", size = 1) +
  labs(
    title = "PGLS: LQ ~ P04141.1",
    x = "P04141.1 dN/dS",
    y = "LQ"
  ) +
  theme_minimal()

wide_data_df <- wide_data_df[order(wide_data_df$Q9H9B1.4), ]
ggplot(wide_data_df, aes(x = Q9H9B1.4, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_30), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ Q9H9B1.4",
    x = "Q9H9B1.4 dN/dS",
    y = "LQ"
  ) +
  theme_minimal() + coord_cartesian(clip = "off")
summary(model30)


wide_data_df <- wide_data_df[order(wide_data_df$P56704.2), ]
ggplot(wide_data_df, aes(x = P56704.2, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_26), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ P56704.2",
    x = "P56704.2 dN/dS",
    y = "LQ"
  ) +
  theme_minimal() + coord_cartesian(clip = "off")

wide_data_df <- wide_data_df[order(wide_data_df$Q9BT67.1), ]
ggplot(wide_data_df, aes(x = Q9BT67.1, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_17), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ Q9BT67.1",
    x = "Q9BT67.1 dN/dS",
    y = "LQ"
  ) +
  theme_minimal() + coord_cartesian(clip = "off")

wide_data_df <- wide_data_df[order(wide_data_df$Q9Y6J9.1), ]
ggplot(wide_data_df, aes(x = Q9Y6J9.1, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_13), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ Q9Y6J9.1",
    x = "Q9Y6J9.1 dN/dS",
    y = "LQ"
  ) +
  theme_minimal() + coord_cartesian(clip = "off")

wide_data_df <- wide_data_df[order(wide_data_df$Q6ZMU5.2), ]
ggplot(wide_data_df, aes(x = Q6ZMU5.2, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_15), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ Q6ZMU5.2",
    x = "Q6ZMU5.2 dN/dS",
    y = "LQ"
  ) +
  theme_minimal() + coord_cartesian(clip = "off")

wide_data_df <- wide_data_df[order(wide_data_df$P09017.2), ]

ggplot(wide_data_df, aes(x = P09017.2, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_11), color = "blue", size = 1) +
  labs(
    title = "PGLS: LQ ~ P09017.2",
    x = "P09017.2 dN/dS",
    y = "LQ"
  ) +
  theme_minimal()

wide_data_df <- wide_data_df[order(wide_data_df$P15692.3), ]
ggplot(wide_data_df, aes(x = P15692.3, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_8), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ P15692.3",
    x = "P15692.3 dN/dS",
    y = "LQ"
  ) +
  theme_minimal()  + coord_cartesian(clip = "off")

wide_data_df <- wide_data_df[order(wide_data_df$Q99836.1), ]
ggplot(wide_data_df, aes(x = Q99836.1, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_model_4), color = "blue", size = 1) +
  geom_text(aes(label = Species), vjust = -0.5, hjust = 0.5, size =3) +
  scale_x_continuous(expand = expansion(mult = 0.1)) +
  labs(
    title = "PGLS: LQ ~ Q99836.1",
    x = "Q99836.1 dN/dS",
    y = "LQ"
  ) +
  theme_minimal() + coord_cartesian(clip = "off")

# Sort by the other protein
wide_data_df <- wide_data_df[order(wide_data_df$P52815.2), ]

ggplot(wide_data_df, aes(x = P52815.2, y = LQ)) +
  geom_point(color = "darkgreen") +
  geom_line(aes(y = pred_model_2), color = "blue", size = 1) +
  labs(
    title = "PGLS: LQ ~ P52815.2",
    x = "P52815.2 dN/dS",
    y = "LQ"
  ) +
  theme_minimal()

ggplot(wide_data_df, aes(x = NP_478054.2, y = LQ)) +
  geom_point(color = "darkgreen") +
  geom_line(aes(y = pred_model_3), color = "blue", size = 1) +
  labs(
    title = "PGLS: LQ ~ NP_478054.2",
    x = "NP_478054.2 dN/dS",
    y = "LQ"
  ) +
  theme_minimal()

wide_data_no_pv <- wide_data[wide_data$Species != "Pteropus_vampyrus", ]

# 2. Use that new data frame when running comparative.data
comp_dat <- comparative.data(
  phy        = myPhylogeny,
  data       = wide_data_no_pv,
  names.col  = "Species",
  vcv        = TRUE,
  warn.dropped = TRUE
)

# 3. Run your pgls model on comp_dat
model_pgls <- pgls(
  formula = LQ ~ NP_004322.1 + P52815.2 + NP_478054.2,
  data    = comp_dat
)

summary(model_pgls)

# If 'wide_data' is a tibble, convert to data.frame first:
wide_data_df1 <- as.data.frame(wide_data)

# Drop the row where Species == "Pteropus_vampyrus":
wide_data_no_pv <- subset(wide_data_df1, Species != "Pteropus_vampyrus")

# Now run comparative.data using the data.frame version (not tibble)
comp_dat <- comparative.data(
  phy       = myPhylogeny,
  data      = wide_data_no_pv,
  names.col = "Species",  # The column that has the species names
  vcv       = TRUE,
  warn.dropped = TRUE
)

model_pgls <- pgls(
  formula = LQ ~  NP_478054.2,
  data    = comp_dat
)

summary(model_pgls)

wide_data_no_pv$pred_modelNOPV <- predict(model_pgls, newdata = wide_data_no_pv)
wide_data_no_pv <- wide_data_no_pv[order(wide_data_no_pv1$NP_478054.2), ]

ggplot(wide_data_no_pv, aes(x = NP_478054.2, y = LQ)) +
  geom_point(color = "darkred") +
  geom_line(aes(y = pred_modelNOPV), color = "blue", size = 1) +
  labs(
    title = "PGLS: LQ ~ NP_478054.2",
    x = "NP_478054.2 dN/dS",
    y = "LQ"
  ) +
  theme_minimal()
