{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: ace_tools in c:\\anaconda\\lib\\site-packages (0.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install ace_tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pip in c:\\anaconda\\lib\\site-packages (24.0)\n",
      "Collecting pip\n",
      "  Downloading pip-24.3.1-py3-none-any.whl.metadata (3.7 kB)\n",
      "Requirement already satisfied: setuptools in c:\\anaconda\\lib\\site-packages (69.5.1)\n",
      "Collecting setuptools\n",
      "  Downloading setuptools-75.6.0-py3-none-any.whl.metadata (6.7 kB)\n",
      "Downloading pip-24.3.1-py3-none-any.whl (1.8 MB)\n",
      "   ---------------------------------------- 0.0/1.8 MB ? eta -:--:--\n",
      "   ---------------------------------------- 0.0/1.8 MB ? eta -:--:--\n",
      "    --------------------------------------- 0.0/1.8 MB 653.6 kB/s eta 0:00:03\n",
      "   -- ------------------------------------- 0.1/1.8 MB 1.1 MB/s eta 0:00:02\n",
      "   ---- ----------------------------------- 0.2/1.8 MB 1.4 MB/s eta 0:00:02\n",
      "   ------ --------------------------------- 0.3/1.8 MB 1.6 MB/s eta 0:00:01\n",
      "   -------- ------------------------------- 0.4/1.8 MB 1.7 MB/s eta 0:00:01\n",
      "   -------------- ------------------------- 0.7/1.8 MB 2.3 MB/s eta 0:00:01\n",
      "   ---------------------- ----------------- 1.0/1.8 MB 3.1 MB/s eta 0:00:01\n",
      "   --------------------------------- ------ 1.5/1.8 MB 4.0 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 1.8/1.8 MB 4.5 MB/s eta 0:00:00\n",
      "Downloading setuptools-75.6.0-py3-none-any.whl (1.2 MB)\n",
      "   ---------------------------------------- 0.0/1.2 MB ? eta -:--:--\n",
      "   ----------------- ---------------------- 0.5/1.2 MB 11.1 MB/s eta 0:00:01\n",
      "   ----------------- ---------------------- 0.5/1.2 MB 11.1 MB/s eta 0:00:01\n",
      "   ----------------- ---------------------- 0.5/1.2 MB 11.1 MB/s eta 0:00:01\n",
      "   ----------------- ---------------------- 0.5/1.2 MB 11.1 MB/s eta 0:00:01\n",
      "   ----------------- ---------------------- 0.5/1.2 MB 11.1 MB/s eta 0:00:01\n",
      "   ----------------- ---------------------- 0.5/1.2 MB 11.1 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 1.7 MB/s eta 0:00:01\n",
      "   ------------------ --------------------- 0.6/1.2 MB 406.6 kB/s eta 0:00:02\n",
      "   ------------------ --------------------- 0.6/1.2 MB 406.6 kB/s eta 0:00:02\n",
      "   ------------------ --------------------- 0.6/1.2 MB 406.6 kB/s eta 0:00:02\n",
      "   ------------------ --------------------- 0.6/1.2 MB 406.6 kB/s eta 0:00:02\n",
      "   ------------------ --------------------- 0.6/1.2 MB 406.6 kB/s eta 0:00:02\n",
      "   ------------------ --------------------- 0.6/1.2 MB 406.6 kB/s eta 0:00:02\n",
      "   ------------------ --------------------- 0.6/1.2 MB 406.6 kB/s eta 0:00:02\n",
      "   ------------------ --------------------- 0.6/1.2 MB 406.6 kB/s eta 0:00:02\n",
      "   ------------------ --------------------- 0.6/1.2 MB 406.6 kB/s eta 0:00:02\n",
      "   ------------------ --------------------- 0.6/1.2 MB 406.6 kB/s eta 0:00:02\n",
      "   ------------------ --------------------- 0.6/1.2 MB 406.6 kB/s eta 0:00:02\n",
      "   ------------------ --------------------- 0.6/1.2 MB 406.6 kB/s eta 0:00:02\n",
      "   ------------------ --------------------- 0.6/1.2 MB 406.6 kB/s eta 0:00:02\n",
      "   ------------------ --------------------- 0.6/1.2 MB 406.6 kB/s eta 0:00:02\n",
      "   ------------------ --------------------- 0.6/1.2 MB 406.6 kB/s eta 0:00:02\n",
      "   ------------------- -------------------- 0.6/1.2 MB 278.0 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 278.0 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 278.0 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 278.0 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 278.0 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 278.0 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 278.0 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 278.0 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 278.0 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 278.0 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 278.0 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 278.0 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 278.0 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 278.0 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 278.0 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 278.0 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 278.0 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 208.8 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 208.8 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 208.8 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 208.8 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 208.8 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 208.8 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 208.8 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 208.8 kB/s eta 0:00:03\n",
      "   ------------------- -------------------- 0.6/1.2 MB 208.8 kB/s eta 0:00:03\n",
      "   -------------------- ------------------- 0.6/1.2 MB 185.0 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 185.0 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 185.0 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 185.0 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 185.0 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 185.0 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 185.0 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 185.0 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 185.0 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 185.0 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 185.0 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 185.0 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 185.0 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 185.0 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   -------------------- ------------------- 0.6/1.2 MB 160.5 kB/s eta 0:00:04\n",
      "   --------------------- ------------------ 0.6/1.2 MB 131.5 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.6/1.2 MB 131.5 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.6/1.2 MB 131.5 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.6/1.2 MB 131.5 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.6/1.2 MB 131.5 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.6/1.2 MB 131.5 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.6/1.2 MB 131.5 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.6/1.2 MB 131.5 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.6/1.2 MB 131.5 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.6/1.2 MB 131.5 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.6/1.2 MB 131.5 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.6/1.2 MB 131.5 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.6/1.2 MB 131.5 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.6/1.2 MB 131.5 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.6/1.2 MB 131.5 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.7/1.2 MB 119.2 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.7/1.2 MB 119.2 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.7/1.2 MB 119.2 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.7/1.2 MB 119.2 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.7/1.2 MB 119.2 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.7/1.2 MB 119.2 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.7/1.2 MB 119.2 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.7/1.2 MB 119.2 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.7/1.2 MB 119.2 kB/s eta 0:00:05\n",
      "   --------------------- ------------------ 0.7/1.2 MB 119.2 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 112.9 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 112.9 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 112.9 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 112.9 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 112.9 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 112.9 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 112.9 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 112.9 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 112.9 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 107.1 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 107.1 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 107.1 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 107.1 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 107.1 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 107.1 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 107.1 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 107.1 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 107.1 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 107.1 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 107.1 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 107.1 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 107.1 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 107.1 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 107.1 kB/s eta 0:00:05\n",
      "   ---------------------- ----------------- 0.7/1.2 MB 107.1 kB/s eta 0:00:05\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 98.5 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 98.5 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 98.5 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 98.5 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 98.5 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 98.5 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 98.5 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ----------------------- ---------------- 0.7/1.2 MB 95.6 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------ --------------- 0.7/1.2 MB 84.3 kB/s eta 0:00:06\n",
      "   ------------------------- -------------- 0.8/1.2 MB 73.6 kB/s eta 0:00:07\n",
      "   ------------------------- -------------- 0.8/1.2 MB 73.6 kB/s eta 0:00:07\n",
      "   ------------------------- -------------- 0.8/1.2 MB 73.6 kB/s eta 0:00:07\n",
      "   ------------------------- -------------- 0.8/1.2 MB 73.6 kB/s eta 0:00:07\n",
      "   ------------------------- -------------- 0.8/1.2 MB 73.2 kB/s eta 0:00:07\n",
      "   ---------------------------------------  1.2/1.2 MB 114.7 kB/s eta 0:00:01\n",
      "   ---------------------------------------- 1.2/1.2 MB 115.1 kB/s eta 0:00:00\n",
      "Installing collected packages: setuptools, pip\n",
      "  Attempting uninstall: setuptools\n",
      "    Found existing installation: setuptools 69.5.1\n",
      "    Uninstalling setuptools-69.5.1:\n",
      "      Successfully uninstalled setuptools-69.5.1\n",
      "  Attempting uninstall: pip\n",
      "    Found existing installation: pip 24.0\n",
      "    Uninstalling pip-24.0:\n",
      "      Successfully uninstalled pip-24.0\n",
      "Successfully installed pip-24.3.1 setuptools-75.6.0\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install -U pip setuptools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Categories</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GHR</td>\n",
       "      <td>[INS/IGF-1 pathway dysregulation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GHRH</td>\n",
       "      <td>[INS/IGF-1 pathway dysregulation, intercellula...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SHC1</td>\n",
       "      <td>[INS/IGF-1 pathway dysregulation]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>POU1F1</td>\n",
       "      <td>[transcriptional alterations]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PROP1</td>\n",
       "      <td>[transcriptional alterations]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2400</th>\n",
       "      <td>PMF1</td>\n",
       "      <td>[transcriptional alterations]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2401</th>\n",
       "      <td>RP1</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2402</th>\n",
       "      <td>SLC41A1</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2403</th>\n",
       "      <td>CLTC-IT1</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2404</th>\n",
       "      <td>GDF1</td>\n",
       "      <td>[SIRT pathway dysregulation]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2405 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Gene                                         Categories\n",
       "0          GHR                  [INS/IGF-1 pathway dysregulation]\n",
       "1         GHRH  [INS/IGF-1 pathway dysregulation, intercellula...\n",
       "2         SHC1                  [INS/IGF-1 pathway dysregulation]\n",
       "3       POU1F1                      [transcriptional alterations]\n",
       "4        PROP1                      [transcriptional alterations]\n",
       "...        ...                                                ...\n",
       "2400      PMF1                      [transcriptional alterations]\n",
       "2401       RP1                                                 []\n",
       "2402   SLC41A1                                                 []\n",
       "2403  CLTC-IT1                                                 []\n",
       "2404      GDF1                       [SIRT pathway dysregulation]\n",
       "\n",
       "[2405 rows x 2 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the TSV file\n",
    "file_path = 'gene-aging-mechanisms.tsv'\n",
    "\n",
    "# Read the file as a DataFrame\n",
    "data = pd.read_csv(file_path, sep=\"\\t\", header=None, quoting=3, engine='python')\n",
    "\n",
    "# Rename columns for clarity\n",
    "data.columns = [\"Gene\", \"Categories\"]\n",
    "\n",
    "# Clean and split categories\n",
    "data[\"Gene\"] = data[\"Gene\"].str.strip('\"')  # Remove double quotes from gene names\n",
    "data[\"Categories\"] = data[\"Categories\"].str.strip(\"'\")  # Remove outer single quotes from categories\n",
    "data[\"Categories\"] = data[\"Categories\"].str.split(\"','\")  # Split multiple categories\n",
    "\n",
    "# Remove unnecessary symbols from category entries\n",
    "data[\"Categories\"] = data[\"Categories\"].apply(\n",
    "    lambda x: [cat.strip(\"'\\\"\") for cat in x] if isinstance(x, list) else x\n",
    ")\n",
    "\n",
    "data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method IndexOpsMixin.nunique of 0                       [INS/IGF-1 pathway dysregulation]\n",
       "1       [INS/IGF-1 pathway dysregulation, intercellula...\n",
       "2                       [INS/IGF-1 pathway dysregulation]\n",
       "3                           [transcriptional alterations]\n",
       "4                           [transcriptional alterations]\n",
       "                              ...                        \n",
       "2400                        [transcriptional alterations]\n",
       "2401                                                   []\n",
       "2402                                                   []\n",
       "2403                                                   []\n",
       "2404                         [SIRT pathway dysregulation]\n",
       "Name: Categories, Length: 2405, dtype: object>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['Categories'].nunique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                             Category  \\\n",
      "0                                                       \n",
      "1                          AMPK pathway dysregulation   \n",
      "2                     INS/IGF-1 pathway dysregulation   \n",
      "3                          SIRT pathway dysregulation   \n",
      "4                           TOR pathway dysregulation   \n",
      "5             accumulation of reactive oxygen species   \n",
      "6                      alterations in DNA methylation   \n",
      "7                alterations in histone modifications   \n",
      "8       changes in the extracellular matrix structure   \n",
      "9                                chromatin remodeling   \n",
      "10                 degradation of proteolytic systems   \n",
      "11                            disabled macroautophagy   \n",
      "12       impairment of proteins folding and stability   \n",
      "13  impairment of the mitochondrial integrity and ...   \n",
      "14             intercellular communication impairment   \n",
      "15                      mitochondrial DNA instability   \n",
      "16                            nuclear DNA instability   \n",
      "17                    nuclear architecture impairment   \n",
      "18                       senescent cells accumulation   \n",
      "19                               stem cell exhaustion   \n",
      "20                               sterile inflammation   \n",
      "21                                 telomere attrition   \n",
      "22                        transcriptional alterations   \n",
      "\n",
      "                                                Genes  \n",
      "0   [TERC, PTPN11, IL7R, NGF, NGFR, ERBB2, PDGFRB,...  \n",
      "1   [SIRT1, ZMPSTE24, ADIPOQ, AKAP13, PRKAG1, PRKA...  \n",
      "2   [GHR, GHRH, SHC1, IGF1R, KL, GH1, IGF1, IGF2, ...  \n",
      "3                                              [GDF1]  \n",
      "4   [ATM, HDAC3, AKT1, HIF1A, SIRT1, LEP, MTOR, ZM...  \n",
      "5   [TP53, TXN, PDGFB, BCL2, CAT, MSRA, SOD2, SOD1...  \n",
      "6   [MYC, FOS, PARP1, BRCA1, HELLS, BMI1, ZMPSTE24...  \n",
      "7   [ATM, HDAC3, BRCA1, CREBBP, BRCA2, RAD51, EP30...  \n",
      "8   [TGFB1, CTGF, ELN, EFEMP1, COL2A1, ADIPOQ, CAP...  \n",
      "9   [TP53, LMNA, MYC, TOP1, EP300, HELLS, RB1, SIR...  \n",
      "10  [PLAU, ERCC8, AKT1, RET, PARP1, BRCA1, VCP, TG...  \n",
      "11  [TP53, PIK3CB, VCP, FOXO1, HMGB1, PSEN1, CISD2...  \n",
      "12  [PEX5, HSP90AA1, HSPA9, HSPD1, HSPA8, CLU, TRA...  \n",
      "13  [TP53, TERT, AKT1, PEX5, PARP1, BCL2, HSP90AA1...  \n",
      "14  [GHRH, TP53, WRN, IGF1R, NRG1, INS, IL2, PDGFB...  \n",
      "15                   [TP53, PARP1, AKT3, LONP1, RRM1]  \n",
      "16  [TP53, TERT, ATM, ERCC2, ERCC8, WRN, LMNA, MYC...  \n",
      "17  [LMNA, PML, PRKCA, HMGB2, LMNB1, SUMO1, ZMPSTE...  \n",
      "18  [TP53, TERT, ATM, WRN, LMNA, IGF1R, HRAS, PRKC...  \n",
      "19  [TP53, ERCC2, STAT3, NRG1, BCL2, XRCC5, FOXO3,...  \n",
      "20  [KL, STAT3, INS, AKT1, FOS, NR3C1, NFKB1, TNF,...  \n",
      "21  [TERT, ATM, WRN, NBN, PARP1, RPA1, BLM, HSP90A...  \n",
      "22  [POU1F1, PROP1, TP53, ATM, ERCC2, TXN, E2F1, N...  \n"
     ]
    }
   ],
   "source": [
    "# Explode the 'Categories' column to create one row per category\n",
    "exploded_data = data.explode(\"Categories\")\n",
    "\n",
    "# Group by the 'Categories' column and collect the genes for each category\n",
    "grouped_by_category = exploded_data.groupby(\"Categories\")[\"Gene\"].apply(list).reset_index()\n",
    "\n",
    "# Rename columns for clarity\n",
    "grouped_by_category.columns = [\"Category\", \"Genes\"]\n",
    "\n",
    "# Display grouped data\n",
    "print(grouped_by_category)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                             Category  Gene Count\n",
      "0                                                            1147\n",
      "1                          AMPK pathway dysregulation          11\n",
      "2                     INS/IGF-1 pathway dysregulation          54\n",
      "3                          SIRT pathway dysregulation           1\n",
      "4                           TOR pathway dysregulation          31\n",
      "5             accumulation of reactive oxygen species         165\n",
      "6                      alterations in DNA methylation          20\n",
      "7                alterations in histone modifications          83\n",
      "8       changes in the extracellular matrix structure          59\n",
      "9                                chromatin remodeling          61\n",
      "10                 degradation of proteolytic systems         193\n",
      "11                            disabled macroautophagy          37\n",
      "12       impairment of proteins folding and stability          20\n",
      "13  impairment of the mitochondrial integrity and ...         110\n",
      "14             intercellular communication impairment         204\n",
      "15                      mitochondrial DNA instability           5\n",
      "16                            nuclear DNA instability         201\n",
      "17                    nuclear architecture impairment          25\n",
      "18                       senescent cells accumulation          48\n",
      "19                               stem cell exhaustion          49\n",
      "20                               sterile inflammation         131\n",
      "21                                 telomere attrition          35\n",
      "22                        transcriptional alterations         444\n"
     ]
    }
   ],
   "source": [
    "# Explode the 'Categories' column to create one row per category\n",
    "exploded_data = data.explode(\"Categories\")\n",
    "\n",
    "# Group by the 'Categories' column and count the number of genes in each category\n",
    "genes_per_category = exploded_data.groupby(\"Categories\")[\"Gene\"].count().reset_index()\n",
    "\n",
    "# Rename columns for clarity\n",
    "genes_per_category.columns = [\"Category\", \"Gene Count\"]\n",
    "\n",
    "# Display grouped and counted data\n",
    "print(genes_per_category)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          Gene  Number of Categories\n",
      "1          A2M                     2\n",
      "3         AASS                     2\n",
      "20        ACAN                     2\n",
      "21         ACE                     2\n",
      "41      ADAM10                     3\n",
      "...        ...                   ...\n",
      "2342     ZBTB1                     3\n",
      "2360     ZFP36                     2\n",
      "2368      ZHX2                     2\n",
      "2370     ZMIZ1                     2\n",
      "2372  ZMPSTE24                     8\n",
      "\n",
      "[451 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "# Explode the 'Categories' column to create one row per category\n",
    "exploded_data = data.explode(\"Categories\")\n",
    "\n",
    "# Group by 'Gene' and count the number of unique categories for each gene\n",
    "gene_category_counts = exploded_data.groupby(\"Gene\")[\"Categories\"].nunique().reset_index()\n",
    "\n",
    "# Filter genes that are present in multiple categories\n",
    "genes_in_multiple_categories = gene_category_counts[gene_category_counts[\"Categories\"] > 1]\n",
    "\n",
    "# Rename columns for clarity\n",
    "genes_in_multiple_categories.columns = [\"Gene\", \"Number of Categories\"]\n",
    "\n",
    "# Display the genes that belong to multiple categories\n",
    "print(genes_in_multiple_categories)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        Gene                     Category\n",
      "0                                        \n",
      "1      LIPT1                             \n",
      "2      LINS1                             \n",
      "3      LIMK2                             \n",
      "4     LIMCH1                             \n",
      "...      ...                          ...\n",
      "2400     TOX  transcriptional alterations\n",
      "2401  POU1F1  transcriptional alterations\n",
      "2402   GABPA  transcriptional alterations\n",
      "2403     FRK  transcriptional alterations\n",
      "2404    SAFB  transcriptional alterations\n",
      "\n",
      "[2405 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "# Explode the 'Categories' column to create one row per category\n",
    "exploded_data = data.explode(\"Categories\")\n",
    "\n",
    "# Count the number of genes in each category\n",
    "category_gene_counts = exploded_data.groupby(\"Categories\")[\"Gene\"].count().reset_index()\n",
    "category_gene_counts.columns = [\"Category\", \"Gene Count\"]\n",
    "\n",
    "# Merge the category counts back into the exploded data\n",
    "exploded_data = exploded_data.merge(category_gene_counts, left_on=\"Categories\", right_on=\"Category\")\n",
    "\n",
    "# Sort by gene name and category gene count\n",
    "exploded_data = exploded_data.sort_values(by=[\"Gene\", \"Gene Count\"])\n",
    "\n",
    "# Keep only the first occurrence of each gene (assigned to the smallest category)\n",
    "assigned_genes = exploded_data.drop_duplicates(subset=[\"Gene\"])\n",
    "\n",
    "# Create a cleaned dataset with unique gene-category assignments\n",
    "cleaned_data = assigned_genes[[\"Gene\", \"Category\"]].sort_values(by=\"Category\").reset_index(drop=True)\n",
    "\n",
    "# Display the cleaned dataset\n",
    "print(cleaned_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved genes for category '' to genes_by_category\\_genes.txt\n",
      "Saved genes for category 'AMPK pathway dysregulation' to genes_by_category\\AMPK_pathway_dysregulation_genes.txt\n",
      "Saved genes for category 'INS/IGF-1 pathway dysregulation' to genes_by_category\\INS_IGF-1_pathway_dysregulation_genes.txt\n",
      "Saved genes for category 'SIRT pathway dysregulation' to genes_by_category\\SIRT_pathway_dysregulation_genes.txt\n",
      "Saved genes for category 'TOR pathway dysregulation' to genes_by_category\\TOR_pathway_dysregulation_genes.txt\n",
      "Saved genes for category 'accumulation of reactive oxygen species' to genes_by_category\\accumulation_of_reactive_oxygen_species_genes.txt\n",
      "Saved genes for category 'alterations in DNA methylation' to genes_by_category\\alterations_in_DNA_methylation_genes.txt\n",
      "Saved genes for category 'alterations in histone modifications' to genes_by_category\\alterations_in_histone_modifications_genes.txt\n",
      "Saved genes for category 'changes in the extracellular matrix structure' to genes_by_category\\changes_in_the_extracellular_matrix_structure_genes.txt\n",
      "Saved genes for category 'chromatin remodeling' to genes_by_category\\chromatin_remodeling_genes.txt\n",
      "Saved genes for category 'degradation of proteolytic systems' to genes_by_category\\degradation_of_proteolytic_systems_genes.txt\n",
      "Saved genes for category 'disabled macroautophagy' to genes_by_category\\disabled_macroautophagy_genes.txt\n",
      "Saved genes for category 'impairment of proteins folding and stability' to genes_by_category\\impairment_of_proteins_folding_and_stability_genes.txt\n",
      "Saved genes for category 'impairment of the mitochondrial integrity and biogenesis' to genes_by_category\\impairment_of_the_mitochondrial_integrity_and_biogenesis_genes.txt\n",
      "Saved genes for category 'intercellular communication impairment' to genes_by_category\\intercellular_communication_impairment_genes.txt\n",
      "Saved genes for category 'mitochondrial DNA instability' to genes_by_category\\mitochondrial_DNA_instability_genes.txt\n",
      "Saved genes for category 'nuclear DNA instability' to genes_by_category\\nuclear_DNA_instability_genes.txt\n",
      "Saved genes for category 'nuclear architecture impairment' to genes_by_category\\nuclear_architecture_impairment_genes.txt\n",
      "Saved genes for category 'senescent cells accumulation' to genes_by_category\\senescent_cells_accumulation_genes.txt\n",
      "Saved genes for category 'stem cell exhaustion' to genes_by_category\\stem_cell_exhaustion_genes.txt\n",
      "Saved genes for category 'sterile inflammation' to genes_by_category\\sterile_inflammation_genes.txt\n",
      "Saved genes for category 'telomere attrition' to genes_by_category\\telomere_attrition_genes.txt\n",
      "Saved genes for category 'transcriptional alterations' to genes_by_category\\transcriptional_alterations_genes.txt\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Function to sanitize category names for file paths\n",
    "def sanitize_category_name(category):\n",
    "    # Replace problematic characters with underscores\n",
    "    return \"\".join(c if c.isalnum() or c in (' ', '-') else '_' for c in category).strip().replace(' ', '_')\n",
    "\n",
    "# Create a directory to store the files\n",
    "output_dir = \"genes_by_category\"\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# Save genes into separate files by category\n",
    "for category, group in cleaned_data.groupby(\"Category\"):\n",
    "    sanitized_category = sanitize_category_name(category)  # Sanitize the category name\n",
    "    file_name = f\"{sanitized_category}_genes.txt\"\n",
    "    file_path = os.path.join(output_dir, file_name)\n",
    "    \n",
    "    # Save only the gene names\n",
    "    group[\"Gene\"].to_csv(file_path, index=False, header=False)\n",
    "    print(f\"Saved genes for category '{category}' to {file_path}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes with updated categories saved to genes_with_categories.csv\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the data\n",
    "genes_file = \"_genes.txt\"\n",
    "criteria_file = \"gene-criteria1.tsv\"\n",
    "\n",
    "# Read the gene list\n",
    "with open(genes_file, \"r\") as file:\n",
    "    gene_list = [line.strip() for line in file if line.strip()]\n",
    "\n",
    "# Read the selection criteria file\n",
    "criteria_df = pd.read_csv(criteria_file, sep=\"\\t\", header=None, names=[\"Gene\", \"Category\"])\n",
    "\n",
    "# Map genes to categories\n",
    "categories = []\n",
    "for gene in gene_list:\n",
    "    matches = criteria_df[criteria_df[\"Gene\"] == gene]\n",
    "    if not matches.empty:\n",
    "        categories.append((gene, matches[\"Category\"].iloc[0]))\n",
    "    else:\n",
    "        categories.append((gene, \"Uncategorized\"))\n",
    "\n",
    "# Create a new DataFrame for the results\n",
    "result_df = pd.DataFrame(categories, columns=[\"Gene\", \"Category\"])\n",
    "\n",
    "# Save to a new file\n",
    "output_file = \"genes_with_categories.csv\"\n",
    "result_df.to_csv(output_file, index=False)\n",
    "print(f\"Genes with updated categories saved to {output_file}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "cleaned_data1 = pd.read_csv('genes_with_categories.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>\"\"</td>\n",
       "      <td>Uncategorized</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LIPT1</td>\n",
       "      <td>'Age-related changes in gene expression, methy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LINS1</td>\n",
       "      <td>'Age-related changes in gene expression, methy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LIMK2</td>\n",
       "      <td>'Age-related changes in gene expression, methy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LIMCH1</td>\n",
       "      <td>'Association of genetic variants and gene expr...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene                                           Category\n",
       "0      \"\"                                      Uncategorized\n",
       "1   LIPT1  'Age-related changes in gene expression, methy...\n",
       "2   LINS1  'Age-related changes in gene expression, methy...\n",
       "3   LIMK2  'Age-related changes in gene expression, methy...\n",
       "4  LIMCH1  'Association of genetic variants and gene expr..."
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cleaned_data1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved genes for category ''Age-related changes in gene expression, methylation or protein activity in humans'' to genes_by_category\\_Age-related_changes_in_gene_expression__methylation_or_protein_activity_in_humans__genes.txt\n",
      "Saved genes for category ''Age-related changes in gene expression, methylation or protein activity in non-mammals'' to genes_by_category\\_Age-related_changes_in_gene_expression__methylation_or_protein_activity_in_non-mammals__genes.txt\n",
      "Saved genes for category ''Age-related changes in gene expression, methylation or protein activity'' to genes_by_category\\_Age-related_changes_in_gene_expression__methylation_or_protein_activity__genes.txt\n",
      "Saved genes for category ''Association of genetic variants and gene expression levels with longevity'' to genes_by_category\\_Association_of_genetic_variants_and_gene_expression_levels_with_longevity__genes.txt\n",
      "Saved genes for category ''Association of the gene with accelerated aging in humans'' to genes_by_category\\_Association_of_the_gene_with_accelerated_aging_in_humans__genes.txt\n",
      "Saved genes for category ''Changes in gene activity enhance age-related deterioration'' to genes_by_category\\_Changes_in_gene_activity_enhance_age-related_deterioration__genes.txt\n",
      "Saved genes for category ''Changes in gene activity extend mammalian lifespan'' to genes_by_category\\_Changes_in_gene_activity_extend_mammalian_lifespan__genes.txt\n",
      "Saved genes for category ''Changes in gene activity extend non-mammalian lifespan'' to genes_by_category\\_Changes_in_gene_activity_extend_non-mammalian_lifespan__genes.txt\n",
      "Saved genes for category ''Changes in gene activity protect against age-related impairment'' to genes_by_category\\_Changes_in_gene_activity_protect_against_age-related_impairment__genes.txt\n",
      "Saved genes for category ''Changes in gene activity reduce mammalian lifespan'' to genes_by_category\\_Changes_in_gene_activity_reduce_mammalian_lifespan__genes.txt\n",
      "Saved genes for category ''Regulation of genes associated with aging'' to genes_by_category\\_Regulation_of_genes_associated_with_aging__genes.txt\n",
      "Saved genes for category 'Uncategorized' to genes_by_category\\Uncategorized_genes.txt\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Function to sanitize category names for file paths\n",
    "def sanitize_category_name(category):\n",
    "    # Replace problematic characters with underscores\n",
    "    return \"\".join(c if c.isalnum() or c in (' ', '-') else '_' for c in category).strip().replace(' ', '_')\n",
    "\n",
    "# Create a directory to store the files\n",
    "output_dir = \"genes_by_category\"\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# Save genes into separate files by category\n",
    "for category, group in cleaned_data1.groupby(\"Category\"):\n",
    "    sanitized_category = sanitize_category_name(category)  # Sanitize the category name\n",
    "    file_name = f\"{sanitized_category}_genes.txt\"\n",
    "    file_path = os.path.join(output_dir, file_name)\n",
    "    \n",
    "    # Save only the gene names\n",
    "    group[\"Gene\"].to_csv(file_path, index=False, header=False)\n",
    "    print(f\"Saved genes for category '{category}' to {file_path}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:7: SyntaxWarning: invalid escape sequence '\\i'\n",
      "<>:7: SyntaxWarning: invalid escape sequence '\\i'\n",
      "C:\\Users\\Леонид\\AppData\\Local\\Temp\\ipykernel_3300\\4142399844.py:7: SyntaxWarning: invalid escape sequence '\\i'\n",
      "  gene_file = \"genes_by_category\\impairment_of_the_mitochondrial_integrity_and_biogenesis_genes.txt\"\n",
      "c:\\Anaconda\\Lib\\site-packages\\Bio\\Entrez\\__init__.py:734: UserWarning: \n",
      "            Email address is not specified.\n",
      "\n",
      "            To make use of NCBI's E-utilities, NCBI requires you to specify your\n",
      "            email address with each request.  As an example, if your email address\n",
      "            is A.N.Other@example.com, you can specify it as follows:\n",
      "               from Bio import Entrez\n",
      "               Entrez.email = 'A.N.Other@example.com'\n",
      "            In case of excessive usage of the E-utilities, NCBI will attempt to contact\n",
      "            a user at the email address provided before blocking access to the\n",
      "            E-utilities.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching protein sequence for ALKBH7 (ID: 14150066)...\n",
      "Fetching protein sequence for NDUFV3 (ID: 20178326)...\n",
      "Fetching protein sequence for RAB3A (ID: 131801)...\n",
      "Fetching protein sequence for PMAIP1 (ID: 6919841)...\n",
      "Fetching protein sequence for SOD2 (ID: 1018191640)...\n",
      "Fetching protein sequence for SIRT5 (ID: 1776945293)...\n",
      "Fetching protein sequence for CEP89 (ID: 50083293)...\n",
      "Fetching protein sequence for SIRT4 (ID: 1880340295)...\n",
      "Fetching protein sequence for BAK1 (ID: 4502363)...\n",
      "Fetching protein sequence for IMMT (ID: 2181405828)...\n",
      "Fetching protein sequence for TIMM8A (ID: 6014944)...\n",
      "Fetching protein sequence for BAD (ID: 14670388)...\n",
      "Fetching protein sequence for HTRA2 (ID: 1012282651)...\n",
      "Fetching protein sequence for NOS3 (ID: 231571356)...\n",
      "Fetching protein sequence for MTFP1 (ID: 52783151)...\n",
      "Fetching protein sequence for DNM1L (ID: 1058916509)...\n",
      "Fetching protein sequence for SNCA (ID: 1768365570)...\n",
      "Fetching protein sequence for FEZ1 (ID: 13431526)...\n",
      "Fetching protein sequence for FASTK (ID: 13124202)...\n",
      "Fetching protein sequence for FANCG (ID: 4759336)...\n",
      "Fetching protein sequence for FLJ22555 (ID: 119590602)...\n",
      "Fetching protein sequence for ACAA2 (ID: 57015371)...\n",
      "Fetching protein sequence for TRAK2 (ID: 24418846)...\n",
      "Fetching protein sequence for SAMM50 (ID: 225543166)...\n",
      "Fetching protein sequence for BIK (ID: 4502411)...\n",
      "Fetching protein sequence for FXN (ID: 31742516)...\n",
      "Fetching protein sequence for OXA1L (ID: 2664703260)...\n",
      "Fetching protein sequence for GCLC (ID: 308199423)...\n",
      "Fetching protein sequence for GCLM (ID: 1346188)...\n",
      "Fetching protein sequence for BCS1L (ID: 1736318468)...\n",
      "Fetching protein sequence for TOMM40 (ID: 12230369)...\n",
      "Fetching protein sequence for TOMM20 (ID: 2498697)...\n",
      "Fetching protein sequence for HEBP2 (ID: 74753513)...\n",
      "Fetching protein sequence for BCL2L2 (ID: 1890276212)...\n",
      "Fetching protein sequence for CAMK2A (ID: 296434552)...\n",
      "Fetching protein sequence for RHOT1 (ID: 571026635)...\n",
      "Fetching protein sequence for AFG1L (ID: 74762631)...\n",
      "Fetching protein sequence for NDUFS2 (ID: 1788951023)...\n",
      "Fetching protein sequence for BNIP3L (ID: 4757860)...\n",
      "Fetching protein sequence for BAX (ID: 612149806)...\n",
      "Fetching protein sequence for COX7A2 (ID: 117121)...\n",
      "Fetching protein sequence for MICOS13 (ID: 1475409107)...\n",
      "Fetching protein sequence for MTX2 (ID: 13124345)...\n",
      "Fetching protein sequence for PPARGC1A (ID: 1237938322)...\n",
      "Fetching protein sequence for MTO1 (ID: 20981712)...\n",
      "Fetching protein sequence for CHCHD3 (ID: 1008265823)...\n",
      "Fetching protein sequence for QRSL1 (ID: 222831590)...\n",
      "Fetching protein sequence for PPIF (ID: 2462516334)...\n",
      "Fetching protein sequence for MRPL12 (ID: 20981709)...\n",
      "Fetching protein sequence for MRPL13 (ID: 22257023)...\n",
      "Fetching protein sequence for MRPL15 (ID: 92090636)...\n",
      "Fetching protein sequence for MRPL28 (ID: 85695426)...\n",
      "Fetching protein sequence for MRPL30 (ID: 74730583)...\n",
      "Fetching protein sequence for MRPL34 (ID: 20139691)...\n",
      "Fetching protein sequence for ARMCX3 (ID: 74753322)...\n",
      "Fetching protein sequence for SUPV3L1 (ID: 1021312380)...\n",
      "Fetching protein sequence for MRPL35 (ID: 124028627)...\n",
      "Fetching protein sequence for MRPL41 (ID: 74750734)...\n",
      "Fetching protein sequence for MRPL42 (ID: 24212391)...\n",
      "Fetching protein sequence for MRPL48 (ID: 118573683)...\n",
      "Fetching protein sequence for MRPL49 (ID: 6093524)...\n",
      "Fetching protein sequence for MRPS12 (ID: 16950589)...\n",
      "Fetching protein sequence for MRPS18B (ID: 7662645)...\n",
      "Fetching protein sequence for MRPS24 (ID: 74731578)...\n",
      "Fetching protein sequence for MRPS25 (ID: 13633894)...\n",
      "Fetching protein sequence for MRPS28 (ID: 300068923)...\n",
      "Fetching protein sequence for MRPS33 (ID: 7706349)...\n",
      "Fetching protein sequence for PRDX3 (ID: 695917253)...\n",
      "Fetching protein sequence for MRPS5 (ID: 1013395825)...\n",
      "Fetching protein sequence for CEBPA (ID: 566559994)...\n",
      "Fetching protein sequence for CERT1 (ID: 1821955878)...\n",
      "Fetching protein sequence for MRPL40 (ID: 21263795)...\n",
      "Fetching protein sequence for CLN8 (ID: 31083053)...\n",
      "Fetching protein sequence for MRPS16 (ID: 2287780773)...\n",
      "Fetching protein sequence for MAPT (ID: 1786986091)...\n",
      "Fetching protein sequence for MAN2A1 (ID: 2462602533)...\n",
      "Fetching protein sequence for COX10 (ID: 292495084)...\n",
      "Fetching protein sequence for POLRMT (ID: 50403691)...\n",
      "Fetching protein sequence for PUM2 (ID: 1210675821)...\n",
      "Protein sequences saved to: genes_by_category/opengenes1_proteins.fasta\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from http.client import IncompleteRead\n",
    "import time\n",
    "\n",
    "\n",
    "# Input file containing gene names\n",
    "gene_file = \"genes_by_category\\impairment_of_the_mitochondrial_integrity_and_biogenesis_genes.txt\"\n",
    "\n",
    "# Output FASTA file for large genes (protein sequences)\n",
    "output_fasta = \"genes_by_category/opengenes1_proteins.fasta\"\n",
    "\n",
    "# Function to fetch sequences robustly with retries\n",
    "def robust_fetch(fetch_function, retries=3, wait_time=5):\n",
    "    for attempt in range(retries):\n",
    "        try:\n",
    "            return fetch_function()\n",
    "        except IncompleteRead as e:\n",
    "            print(f\"IncompleteRead error: {e}. Retrying {attempt + 1}/{retries}...\")\n",
    "            time.sleep(wait_time)\n",
    "    raise Exception(\"Failed to fetch data after multiple retries\")\n",
    "\n",
    "# Fetch protein sequences from NCBI\n",
    "def fetch_protein_sequence(gene_name):\n",
    "    try:\n",
    "        # Search for the gene in the NCBI protein database (restricted to Homo sapiens)\n",
    "        query = f\"{gene_name}[Gene] AND Homo sapiens[Organism]\"\n",
    "        handle = Entrez.esearch(db=\"protein\", term=query, retmax=1)\n",
    "        record = Entrez.read(handle)\n",
    "        handle.close()\n",
    "\n",
    "        # Fetch the protein sequence by ID\n",
    "        if record[\"IdList\"]:\n",
    "            seq_id = record[\"IdList\"][0]\n",
    "            print(f\"Fetching protein sequence for {gene_name} (ID: {seq_id})...\")\n",
    "            handle = robust_fetch(\n",
    "                lambda: Entrez.efetch(\n",
    "                    db=\"protein\", id=seq_id, rettype=\"fasta\", retmode=\"text\"\n",
    "                )\n",
    "            )\n",
    "            sequence = handle.read()\n",
    "            handle.close()\n",
    "            return sequence\n",
    "        else:\n",
    "            print(f\"No protein sequence found for {gene_name}\")\n",
    "            return None\n",
    "    except Exception as e:\n",
    "        print(f\"Error fetching {gene_name}: {e}\")\n",
    "        return None\n",
    "\n",
    "# Read gene names and fetch protein sequences\n",
    "with open(gene_file, \"r\") as file:\n",
    "    genes = [line.strip() for line in file if line.strip()]\n",
    "\n",
    "with open(output_fasta, \"w\") as fasta_out:\n",
    "    for gene in genes:\n",
    "        protein_sequence = fetch_protein_sequence(gene)\n",
    "        if protein_sequence:\n",
    "            fasta_out.write(protein_sequence)\n",
    "\n",
    "print(f\"Protein sequences saved to: {output_fasta}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered file saved to realmergedprots_noen.faa\n"
     ]
    }
   ],
   "source": [
    "input_file = \"realmergedprots_filtered.faa\"\n",
    "output_file = \"realmergedprots_noen.faa\"\n",
    "\n",
    "with open(input_file, \"r\") as infile, open(output_file, \"w\") as outfile:\n",
    "    write = True\n",
    "    for line in infile:\n",
    "        if line.startswith(\">en\"):\n",
    "            write = False  # Skip writing this header line\n",
    "        elif line.startswith(\">\"):\n",
    "            write = True  # Reset for the next header\n",
    "        if write:\n",
    "            outfile.write(line)\n",
    "\n",
    "print(f\"Filtered file saved to {output_file}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "mmseqs/bin/mmseqs easy-search impairment_of_the_mitochondrial_integrity_and_biogenesis_genes.fasta realmergedprots_filtered.faa mitimpair.m8 -c 0.2 tmp --min-seq-id 0.3 --alt-ali 25"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "import Biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ace_tools==0.0Note: you may need to restart the kernel to use updated packages.\n",
      "\n",
      "aiohttp @ file:///C:/b/abs_8860tt1424/croot/aiohttp_1715108828392/work\n",
      "aiosignal @ file:///tmp/build/80754af9/aiosignal_1637843061372/work\n",
      "anaconda-anon-usage @ file:///C:/b/abs_c3w_h1zzjg/croot/anaconda-anon-usage_1710965204622/work\n",
      "anaconda-catalogs @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/anaconda-catalogs_1701816586117/work\n",
      "anaconda-client @ file:///C:/b/abs_34txutm0ue/croot/anaconda-client_1708640705294/work\n",
      "anaconda-cloud-auth @ file:///C:/b/abs_b02evi84gh/croot/anaconda-cloud-auth_1713991445770/work\n",
      "anaconda-navigator @ file:///C:/b/abs_d8d4a02c3t/croot/anaconda-navigator_1713464063970/work\n",
      "anaconda-project @ file:///C:/b/abs_95s0l9dwvd/croot/anaconda-project_1706049257687/work\n",
      "annotated-types @ file:///C:/b/abs_0dmaoyhhj3/croot/annotated-types_1709542968311/work\n",
      "anyio @ file:///C:/b/abs_847uobe7ea/croot/anyio_1706220224037/work\n",
      "appdirs==1.4.4\n",
      "archspec @ file:///croot/archspec_1709217642129/work\n",
      "argon2-cffi @ file:///opt/conda/conda-bld/argon2-cffi_1645000214183/work\n",
      "argon2-cffi-bindings @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/argon2-cffi-bindings_1699549801117/work\n",
      "asttokens @ file:///opt/conda/conda-bld/asttokens_1646925590279/work\n",
      "async-lru @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/async-lru_1701796859357/work\n",
      "attrs @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/attrs_1699473735254/work\n",
      "Babel @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/babel_1699475785740/work\n",
      "beautifulsoup4 @ file:///C:/b/abs_d5wytg_p0w/croot/beautifulsoup4-split_1718029833749/work\n",
      "biopython==1.84\n",
      "bleach @ file:///opt/conda/conda-bld/bleach_1641577558959/work\n",
      "bokeh @ file:///C:/b/abs_0aes0rbxnz/croot/bokeh_1718119072183/work\n",
      "boltons @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/boltons_1699480450092/work\n",
      "Bottleneck @ file:///C:/b/abs_f7un855idq/croot/bottleneck_1709069969633/work\n",
      "Brotli @ file:///C:/b/abs_3d36mno480/croot/brotli-split_1714483178642/work\n",
      "certifi @ file:///C:/b/abs_1fw_exq1si/croot/certifi_1725551736618/work/certifi\n",
      "cffi @ file:///C:/b/abs_78eb1_vq6z/croot/cffi_1714483206096/work\n",
      "chardet @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/chardet_1699498892802/work\n",
      "charset-normalizer @ file:///tmp/build/80754af9/charset-normalizer_1630003229654/work\n",
      "click @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/click_1699473800648/work\n",
      "cloudpickle @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/cloudpickle_1699480575079/work\n",
      "colorama @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/colorama_1699472650914/work\n",
      "comm @ file:///C:/b/abs_67a8058udb/croot/comm_1709322909844/work\n",
      "conda @ file:///C:/b/abs_69e_za3rd5/croot/conda_1722004627938/work\n",
      "conda-build @ file:///C:/b/abs_85st9w283e/croot/conda-build_1716991354861/work\n",
      "conda-content-trust @ file:///C:/b/abs_bdfatn_wzf/croot/conda-content-trust_1714483201909/work\n",
      "conda-libmamba-solver @ file:///croot/conda-libmamba-solver_1706733287605/work/src\n",
      "conda-pack @ file:///C:/b/abs_0cu91sbau5/croot/conda-pack_1710258097566/work\n",
      "conda-package-handling @ file:///C:/b/abs_831_pypuuz/croot/conda-package-handling_1718138287790/work\n",
      "conda-repo-cli @ file:///C:/b/abs_7dlvg5f_6q/croot/conda-repo-cli_1709246667780/work\n",
      "conda-token @ file:///croot/conda-token_1718995751285/work\n",
      "conda_index @ file:///C:/b/abs_45v3w98im_/croot/conda-index_1719338245731/work\n",
      "conda_package_streaming @ file:///C:/b/abs_fe6zb0_j8_/croot/conda-package-streaming_1718136098642/work\n",
      "contourpy @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/contourpy_1701817388761/work\n",
      "cryptography @ file:///C:/b/abs_97lgskwerx/croot/cryptography_1714664736250/work\n",
      "cycler==0.12.1\n",
      "cytoolz @ file:///C:/b/abs_d43s8lnb60/croot/cytoolz_1701723636699/work\n",
      "dask @ file:///C:/b/abs_b0it6fcs79/croot/dask-core_1715838625158/work\n",
      "dask-expr @ file:///C:/b/abs_cf38ehsxus/croot/dask-expr_1715846496633/work\n",
      "debugpy @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/debugpy_1699554994633/work\n",
      "decorator @ file:///opt/conda/conda-bld/decorator_1643638310831/work\n",
      "defusedxml @ file:///tmp/build/80754af9/defusedxml_1615228127516/work\n",
      "distributed @ file:///C:/b/abs_27jqurd3gk/croot/distributed_1715844512065/work\n",
      "distro @ file:///C:/b/abs_71xr36ua5r/croot/distro_1714488282676/work\n",
      "entrypoints @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/entrypoints_1699484295227/work\n",
      "et_xmlfile==2.0.0\n",
      "ete3==3.1.3\n",
      "executing @ file:///opt/conda/conda-bld/executing_1646925071911/work\n",
      "fastjsonschema @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/python-fastjsonschema_1699475134300/work\n",
      "filelock @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/filelock_1701807523603/work\n",
      "fonttools==4.54.1\n",
      "frozendict @ file:///C:/b/abs_2alamqss6p/croot/frozendict_1713194885124/work\n",
      "frozenlist @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/frozenlist_1699557027827/work\n",
      "fsspec @ file:///C:/b/abs_d5jzmndkba/croot/fsspec_1714461594761/work\n",
      "h11 @ file:///C:/b/abs_1czwoyexjf/croot/h11_1706652332846/work\n",
      "HeapDict @ file:///Users/ktietz/demo/mc3/conda-bld/heapdict_1630598515714/work\n",
      "httpcore @ file:///C:/b/abs_55n7g233bw/croot/httpcore_1706728507241/work\n",
      "httpx @ file:///C:/b/abs_43e135shby/croot/httpx_1723474830126/work\n",
      "idna @ file:///C:/b/abs_aad84bnnw5/croot/idna_1714398896795/work\n",
      "intake @ file:///C:/b/abs_15sg46u7x7/croot/intake_1717513827138/work\n",
      "ipykernel @ file:///C:/b/abs_c2u94kxcy6/croot/ipykernel_1705933907920/work\n",
      "ipython @ file:///C:/b/abs_53it5seiim/croot/ipython_1726064251844/work\n",
      "jaraco.classes @ file:///tmp/build/80754af9/jaraco.classes_1620983179379/work\n",
      "jedi @ file:///C:/b/abs_1b8kmj7rrm/croot/jedi_1721058359741/work\n",
      "Jinja2 @ file:///C:/b/abs_92fccttino/croot/jinja2_1716993447201/work\n",
      "joblib==1.4.2\n",
      "json5 @ file:///tmp/build/80754af9/json5_1624432770122/work\n",
      "jsonpatch @ file:///C:/b/abs_4fdm88t7zi/croot/jsonpatch_1714483974578/work\n",
      "jsonpointer==2.1\n",
      "jsonschema @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/jsonschema_1701796333530/work\n",
      "jsonschema-specifications @ file:///C:/Users/dev-admin/py312/jsonschema-specifications_1706803038066/work\n",
      "jupyter-events @ file:///C:/b/abs_c2m9s5b5m5/croot/jupyter_events_1718738115254/work\n",
      "jupyter-lsp @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/jupyter-lsp-meta_1701806528286/work\n",
      "jupyter_client @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/jupyter_client_1701796382758/work\n",
      "jupyter_core @ file:///C:/b/abs_beftpbuevw/croot/jupyter_core_1718818307097/work\n",
      "jupyter_server @ file:///C:/b/abs_9a333nh6yu/croot/jupyter_server_1718827092223/work\n",
      "jupyter_server_terminals @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/jupyter_server_terminals_1701798041635/work\n",
      "jupyterlab @ file:///C:/b/abs_88qjk9lo2r/croot/jupyterlab_1725895228333/work\n",
      "jupyterlab-pygments @ file:///tmp/build/80754af9/jupyterlab_pygments_1601490720602/work\n",
      "jupyterlab_server @ file:///C:/b/abs_fdi5r_tpjc/croot/jupyterlab_server_1725865372811/work\n",
      "keyring @ file:///C:/b/abs_78uoj9sw00/croot/keyring_1709632550180/work\n",
      "kiwisolver==1.4.7\n",
      "libarchive-c @ file:///tmp/build/80754af9/python-libarchive-c_1617780486945/work\n",
      "libmambapy @ file:///C:/b/abs_89suv9zwkz/croot/mamba-split_1714484512428/work/libmambapy\n",
      "lmdb @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/python-lmdb_1699495888630/work\n",
      "locket @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/locket_1699476508014/work\n",
      "lz4 @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/lz4_1699559532268/work\n",
      "MarkupSafe @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/markupsafe_1707425732791/work\n",
      "matplotlib==3.9.2\n",
      "matplotlib-inline @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/matplotlib-inline_1699484796387/work\n",
      "menuinst @ file:///C:/b/abs_e7zsrwvoy0/croot/menuinst_1718132546150/work\n",
      "mistune @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/mistune_1699559923704/work\n",
      "mkl-fft @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/mkl_fft_1699473528480/work\n",
      "mkl-random @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/mkl_random_1699473588250/work\n",
      "mkl-service==2.4.0\n",
      "more-itertools @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/more-itertools_1701811589421/work\n",
      "mpmath==1.3.0\n",
      "msgpack @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/msgpack-python_1699473924872/work\n",
      "multidict @ file:///C:/b/abs_44ido987fv/croot/multidict_1701097803486/work\n",
      "mysql-connector-python==9.1.0\n",
      "navigator-updater @ file:///C:/b/abs_22_3uaq2ey/croot/navigator-updater_1718030405023/work\n",
      "nbclient @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/nbclient_1701796752236/work\n",
      "nbconvert @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/nbconvert_1701806279884/work\n",
      "nbformat @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/nbformat_1701796704118/work\n",
      "nest-asyncio @ file:///C:/b/abs_65d6lblmoi/croot/nest-asyncio_1708532721305/work\n",
      "networkx==3.3\n",
      "notebook @ file:///C:/b/abs_03qbdiu7cg/croot/notebook_1725954777879/work\n",
      "notebook_shim @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/notebook-shim_1701806592322/work\n",
      "numexpr @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/numexpr_1699503421264/work\n",
      "numpy @ file:///C:/b/abs_c1ywpu18ar/croot/numpy_and_numpy_base_1708638681471/work/dist/numpy-1.26.4-cp312-cp312-win_amd64.whl#sha256=becc06674317799ad0165a939a7613809d0bee9bd328a1e4308c57c39cacf08c\n",
      "openpyxl==3.1.5\n",
      "overrides @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/overrides_1701806336503/work\n",
      "packaging @ file:///C:/b/abs_cc1h2xfosn/croot/packaging_1710807447479/work\n",
      "pandas @ file:///C:/b/abs_9aotnvvz16/croot/pandas_1718308978393/work/dist/pandas-2.2.2-cp312-cp312-win_amd64.whl#sha256=93959056e02e9855025011adb18394296a58d49e72b9342733b7693a5267c790\n",
      "pandocfilters @ file:///opt/conda/conda-bld/pandocfilters_1643405455980/work\n",
      "parso @ file:///opt/conda/conda-bld/parso_1641458642106/work\n",
      "partd @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/partd_1699482661603/work\n",
      "pillow @ file:///C:/b/abs_95ls_qh4c2/croot/pillow_1714398870615/work\n",
      "pkce @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/pkce_1699482718905/work\n",
      "pkginfo @ file:///C:/b/abs_bciypal17m/croot/pkginfo_1715696027839/work\n",
      "platformdirs @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/platformdirs_1701797392447/work\n",
      "pluggy @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/pluggy_1699472504117/work\n",
      "ply @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/ply_1699473999871/work\n",
      "prometheus-client @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/prometheus_client_1699495496669/work\n",
      "prompt-toolkit @ file:///C:/b/abs_68uwr58ed1/croot/prompt-toolkit_1704404394082/work\n",
      "psutil @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/psutil_1699482842340/work\n",
      "pure-eval @ file:///opt/conda/conda-bld/pure_eval_1646925070566/work\n",
      "pyarrow @ file:///C:/b/abs_93i_y2dub4/croot/pyarrow_1707330894046/work/python\n",
      "pycosat @ file:///C:/b/abs_5csdern___/croot/pycosat_1714513102923/work\n",
      "pycparser @ file:///tmp/build/80754af9/pycparser_1636541352034/work\n",
      "pydantic @ file:///C:/b/abs_f6ffn86wjx/croot/pydantic_1709578044014/work\n",
      "pydantic_core @ file:///C:/b/abs_05smrkbwjm/croot/pydantic-core_1709574000211/work\n",
      "Pygments @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/pygments_1699474141968/work\n",
      "PyJWT @ file:///C:/b/abs_04qhgo75wz/croot/pyjwt_1715095119685/work\n",
      "pyparsing==3.2.0\n",
      "PyQt5==5.15.10\n",
      "PyQt5-sip @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/pyqt-split_1699478669290/work/pyqt_sip\n",
      "PyQtWebEngine==5.15.6\n",
      "PySocks @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/pysocks_1699473336188/work\n",
      "python-dateutil @ file:///C:/b/abs_3au_koqnbs/croot/python-dateutil_1716495777160/work\n",
      "python-dotenv @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/python-dotenv_1699475097728/work\n",
      "python-json-logger @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/python-json-logger_1699543626759/work\n",
      "python-snappy @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/python-snappy_1699564939880/work\n",
      "pytz @ file:///C:/b/abs_6ap4tsz1ox/croot/pytz_1713974360290/work\n",
      "pywin32==305.1\n",
      "pywin32-ctypes @ file:///C:/b/abs_2cfx5l4nvi/croot/pywin32-ctypes_1709340246092/work\n",
      "pywinpty @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/pywinpty_1699496010557/work/target/wheels/pywinpty-2.0.10-cp312-none-win_amd64.whl#sha256=227f3f94f568f63ab3418c02875c4b429a0a3638e1a2917a22bb8e4ba1762179\n",
      "PyYAML @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/pyyaml_1699479991712/work\n",
      "pyzmq @ file:///C:/b/abs_89aq69t0up/croot/pyzmq_1705605705281/work\n",
      "QtPy @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/qtpy_1701807198514/work\n",
      "referencing @ file:///C:/Users/dev-admin/py312/referencing_1706802962559/work\n",
      "requests @ file:///C:/b/abs_9c5n24p1y7/croot/requests_1716902867639/work\n",
      "requests-toolbelt @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/requests-toolbelt_1699475466775/work\n",
      "rfc3339-validator @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/rfc3339-validator_1699543924991/work\n",
      "rfc3986-validator @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/rfc3986-validator_1699543955651/work\n",
      "rpds-py @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/rpds-py_1699565648947/work\n",
      "ruamel-yaml-conda @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/ruamel_yaml_1699543990032/work\n",
      "ruamel.yaml @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/ruamel.yaml_1699483184324/work\n",
      "scikit-learn==1.5.2\n",
      "scipy==1.14.1\n",
      "semver @ file:///C:/b/abs_4bkbn3v6jp/croot/semver_1709243682483/work\n",
      "Send2Trash @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/send2trash_1701806400767/work\n",
      "setuptools==75.6.0\n",
      "sip @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/sip_1699475590677/work\n",
      "six @ file:///tmp/build/80754af9/six_1644875935023/work\n",
      "sniffio @ file:///C:/b/abs_3akdewudo_/croot/sniffio_1705431337396/work\n",
      "sortedcontainers @ file:///tmp/build/80754af9/sortedcontainers_1623949099177/work\n",
      "soupsieve @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/soupsieve_1699496611169/work\n",
      "stack-data @ file:///opt/conda/conda-bld/stack_data_1646927590127/work\n",
      "sympy==1.13.3\n",
      "tblib @ file:///Users/ktietz/demo/mc3/conda-bld/tblib_1629402031467/work\n",
      "terminado @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/terminado_1699545066607/work\n",
      "threadpoolctl==3.5.0\n",
      "tinycss2 @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/tinycss2_1699545185754/work\n",
      "toolz @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/toolz_1699474617784/work\n",
      "torch==2.4.1+cu124\n",
      "torchaudio==2.4.1+cu124\n",
      "torchvision==0.19.1+cu124\n",
      "tornado @ file:///C:/b/abs_7bua0304mj/croot/tornado_1718740122405/work\n",
      "tqdm @ file:///C:/b/abs_a7h7n45ccq/croot/tqdm_1716395987133/work\n",
      "traitlets @ file:///C:/b/abs_bfsnoxl4pq/croot/traitlets_1718227069245/work\n",
      "truststore @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/truststore_1701881385424/work\n",
      "typing_extensions @ file:///C:/b/abs_0as9mdbkfl/croot/typing_extensions_1715268906610/work\n",
      "tzdata @ file:///croot/python-tzdata_1690578112552/work\n",
      "ujson @ file:///C:/b/abs_c93a278xy4/croot/ujson_1717598062082/work\n",
      "urllib3 @ file:///C:/b/abs_a7hvzm4y95/croot/urllib3_1718912661242/work\n",
      "wcwidth @ file:///Users/ktietz/demo/mc3/conda-bld/wcwidth_1629357192024/work\n",
      "webencodings @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/webencodings_1699497069416/work\n",
      "websocket-client @ file:///C:/b/abs_5dmnxxoci9/croot/websocket-client_1715878351319/work\n",
      "wheel==0.43.0\n",
      "win-inet-pton @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/win_inet_pton_1699472992992/work\n",
      "xyzservices @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/xyzservices_1699548906163/work\n",
      "yarl @ file:///C:/b/abs_8bxwdyhjvp/croot/yarl_1701105248152/work\n",
      "zict @ file:///C:/Users/dev-admin/perseverance-python-buildout/croot/zict_1699548979005/work\n",
      "zstandard @ file:///C:/b/abs_e46tf0_91g/croot/zstandard_1714678303615/work\n"
     ]
    }
   ],
   "source": [
    "%pip freeze"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Category: accumulation_of_reactive_oxygen_species_genes, Number of genes: 126\n",
      "Category: alterations_in_DNA_methylation_genes, Number of genes: 18\n",
      "Category: alterations_in_histone_modifications_genes, Number of genes: 45\n",
      "Category: AMPK pathway dysregulation_genes, Number of genes: 11\n",
      "Category: AMPK_pathway_dysregulation_genes, Number of genes: 11\n",
      "Category: changes_in_the_extracellular_matrix_structure_genes, Number of genes: 57\n",
      "Category: chromatin_remodeling_genes, Number of genes: 39\n",
      "Category: degradation_of_proteolytic_systems_genes, Number of genes: 114\n",
      "Category: disabled_macroautophagy_genes, Number of genes: 34\n",
      "Category: impairment_of_proteins_folding_and_stability_genes, Number of genes: 20\n",
      "Category: impairment_of_the_mitochondrial_integrity_and_biogenesis_genes, Number of genes: 79\n",
      "Category: INS_IGF-1_pathway_dysregulation_genes, Number of genes: 45\n",
      "Category: intercellular_communication_impairment_genes, Number of genes: 118\n",
      "Category: mitochondrial_DNA_instability_genes, Number of genes: 5\n",
      "Category: nuclear_architecture_impairment_genes, Number of genes: 24\n",
      "Category: nuclear_DNA_instability_genes, Number of genes: 82\n",
      "Category: senescent_cells_accumulation_genes, Number of genes: 33\n",
      "Category: SIRT_pathway_dysregulation_genes, Number of genes: 1\n",
      "Category: stem_cell_exhaustion_genes, Number of genes: 38\n",
      "Category: sterile_inflammation_genes, Number of genes: 99\n",
      "Category: telomere_attrition_genes, Number of genes: 30\n",
      "Category: TOR_pathway_dysregulation_genes, Number of genes: 26\n",
      "Category: transcriptional_alterations_genes, Number of genes: 214\n",
      "Category: _Age-related_changes_in_gene_expression__methylation_or_protein_activity_in_humans__genes, Number of genes: 932\n",
      "Category: _Age-related_changes_in_gene_expression__methylation_or_protein_activity_in_non-mammals__genes, Number of genes: 2\n",
      "Category: _Age-related_changes_in_gene_expression__methylation_or_protein_activity__genes, Number of genes: 54\n",
      "Category: _Association_of_genetic_variants_and_gene_expression_levels_with_longevity__genes, Number of genes: 112\n",
      "Category: _Association_of_the_gene_with_accelerated_aging_in_humans__genes, Number of genes: 2\n",
      "Category: _Changes_in_gene_activity_enhance_age-related_deterioration__genes, Number of genes: 1\n",
      "Category: _Changes_in_gene_activity_extend_mammalian_lifespan__genes, Number of genes: 12\n",
      "Category: _Changes_in_gene_activity_extend_non-mammalian_lifespan__genes, Number of genes: 8\n",
      "Category: _Changes_in_gene_activity_protect_against_age-related_impairment__genes, Number of genes: 3\n",
      "Category: _Changes_in_gene_activity_reduce_mammalian_lifespan__genes, Number of genes: 11\n",
      "Category: _Regulation_of_genes_associated_with_aging__genes, Number of genes: 8\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "def count_genes_in_categories(folder_path='genes_by_category'):\n",
    "    \"\"\"\n",
    "    Counts and displays the number of genes (lines) in each .txt file\n",
    "    in the given folder. Assumes each line in the file represents one gene.\n",
    "    \"\"\"\n",
    "    for filename in os.listdir(folder_path):\n",
    "        if filename.endswith(\".txt\"):\n",
    "            # Derive category name from the file name (strip '.txt')\n",
    "            category_name = os.path.splitext(filename)[0]\n",
    "            \n",
    "            # Construct full path to the file\n",
    "            file_path = os.path.join(folder_path, filename)\n",
    "            \n",
    "            # Read all lines and count them\n",
    "            with open(file_path, 'r') as f:\n",
    "                lines = [line.strip() for line in f if line.strip()]  # strip out empty lines\n",
    "            num_genes = len(lines)\n",
    "            \n",
    "            print(f\"Category: {category_name}, Number of genes: {num_genes}\")\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    count_genes_in_categories()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                             category  num_genes\n",
      "0       accumulation_of_reactive_oxygen_species_genes        126\n",
      "1                alterations_in_DNA_methylation_genes         18\n",
      "2          alterations_in_histone_modifications_genes         45\n",
      "3                    AMPK pathway dysregulation_genes         11\n",
      "4                    AMPK_pathway_dysregulation_genes         11\n",
      "5   changes_in_the_extracellular_matrix_structure_...         57\n",
      "6                          chromatin_remodeling_genes         39\n",
      "7            degradation_of_proteolytic_systems_genes        114\n",
      "8                       disabled_macroautophagy_genes         34\n",
      "9   impairment_of_proteins_folding_and_stability_g...         20\n",
      "10  impairment_of_the_mitochondrial_integrity_and_...         79\n",
      "11              INS_IGF-1_pathway_dysregulation_genes         45\n",
      "12       intercellular_communication_impairment_genes        118\n",
      "13                mitochondrial_DNA_instability_genes          5\n",
      "14              nuclear_architecture_impairment_genes         24\n",
      "15                      nuclear_DNA_instability_genes         82\n",
      "16                 senescent_cells_accumulation_genes         33\n",
      "17                   SIRT_pathway_dysregulation_genes          1\n",
      "18                         stem_cell_exhaustion_genes         38\n",
      "19                         sterile_inflammation_genes         99\n",
      "20                           telomere_attrition_genes         30\n",
      "21                    TOR_pathway_dysregulation_genes         26\n",
      "22                  transcriptional_alterations_genes        214\n",
      "23  _Age-related_changes_in_gene_expression__methy...        932\n",
      "24  _Age-related_changes_in_gene_expression__methy...          2\n",
      "25  _Age-related_changes_in_gene_expression__methy...         54\n",
      "26  _Association_of_genetic_variants_and_gene_expr...        112\n",
      "27  _Association_of_the_gene_with_accelerated_agin...          2\n",
      "28  _Changes_in_gene_activity_enhance_age-related_...          1\n",
      "29  _Changes_in_gene_activity_extend_mammalian_lif...         12\n",
      "30  _Changes_in_gene_activity_extend_non-mammalian...          8\n",
      "31  _Changes_in_gene_activity_protect_against_age-...          3\n",
      "32  _Changes_in_gene_activity_reduce_mammalian_lif...         11\n",
      "33  _Regulation_of_genes_associated_with_aging__genes          8\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "def generate_genes_df(folder_path='genes_by_category'):\n",
    "    \"\"\"\n",
    "    Generates a DataFrame with two columns: 'category' and 'num_genes'.\n",
    "    Each row corresponds to a .txt file in the folder, \n",
    "    where 'category' is the file name (without .txt) \n",
    "    and 'num_genes' is the count of non-empty lines (genes).\n",
    "    \"\"\"\n",
    "    data = []\n",
    "    \n",
    "    # Loop over all files in the folder\n",
    "    for filename in os.listdir(folder_path):\n",
    "        if filename.endswith(\".txt\"):\n",
    "            category_name = os.path.splitext(filename)[0]  # Remove the '.txt' extension\n",
    "            file_path = os.path.join(folder_path, filename)\n",
    "            \n",
    "            # Count the number of non-empty lines (genes)\n",
    "            with open(file_path, 'r') as f:\n",
    "                lines = [line.strip() for line in f if line.strip()]  \n",
    "            num_genes = len(lines)\n",
    "            \n",
    "            # Append to our data list\n",
    "            data.append((category_name, num_genes))\n",
    "    \n",
    "    # Create a DataFrame from our data\n",
    "    df = pd.DataFrame(data, columns=['category', 'num_genes'])\n",
    "    return df\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    genes_df = generate_genes_df()\n",
    "    print(genes_df)\n",
    "    # Optionally, to save this DataFrame to a CSV file:\n",
    "    # genes_df.to_csv(\"genes_by_category_summary.csv\", index=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
