{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found files: ['secrets\\\\ajsec.tab', 'secrets\\\\drsec.tab', 'secrets\\\\efsec.tab', 'secrets\\\\ensec.tab', 'secrets\\\\mbsec.tab', 'secrets\\\\mdsec.tab', 'secrets\\\\mlsec.tab', 'secrets\\\\mmsec.tab', 'secrets\\\\mosec.tab', 'secrets\\\\pasec.tab', 'secrets\\\\pdsec.tab', 'secrets\\\\pgsec.tab', 'secrets\\\\phsec.tab', 'secrets\\\\pksec.tab', 'secrets\\\\pnsec.tab', 'secrets\\\\pvsec.tab', 'secrets\\\\rasec.tab', 'secrets\\\\rfsec.tab', 'secrets\\\\sbsec.tab']\n",
      "DataFrames loaded: ['ajdf', 'drdf', 'efdf', 'endf', 'mbdf', 'mddf', 'mldf', 'mmdf', 'modf', 'padf', 'pddf', 'pgdf', 'phdf', 'pkdf', 'pndf', 'pvdf', 'radf', 'rfdf', 'sbdf']\n",
      "ajdf head:\n",
      "           qseqid            sseqid  pident  length  mismatch  gapopen  \\\n",
      "0     NP_005802.1  ajXP_037000028.1   0.987     407         5        0   \n",
      "1  XP_006719257.1  ajXP_037000028.1   0.987     407         5        0   \n",
      "2     NP_004855.2  ajXP_036997943.2   0.618     309       117        0   \n",
      "3     NP_057288.1  ajXP_037014354.1   0.790     429        90        0   \n",
      "4     NP_065685.1  ajXP_037023745.1   0.841     365        58        0   \n",
      "\n",
      "   qstart  qend  sstart  send         evalue  bitscore  \n",
      "0       1   407       1   405  6.545000e-280       852  \n",
      "1       1   407       1   405  6.545000e-280       852  \n",
      "2       1   308       1   309  4.820000e-120       385  \n",
      "3       1   429       1   429  1.213000e-231       714  \n",
      "4       1   364       1   365  1.571000e-213       658  \n",
      "drdf head:\n",
      "        qseqid            sseqid  pident  length  mismatch  gapopen  qstart  \\\n",
      "0  NP_000790.2  drXP_053772465.1   0.847     189        29        0       5   \n",
      "1  NP_000630.1  drXP_024407626.1   0.836     281        46        0       1   \n",
      "2  NP_000791.1  drXP_045039822.1   0.929     155        11        0       1   \n",
      "3  NP_000791.1  drXP_045039823.1   0.929     155        11        0       1   \n",
      "4  NP_000791.1  drXP_045039824.1   0.929     155        11        0       1   \n",
      "\n",
      "   qend  sstart  send         evalue  bitscore  \n",
      "0   193      11   199   1.095000e-98       317  \n",
      "1   281       1   281  8.088000e-163       507  \n",
      "2   155       1   155   1.092000e-96       309  \n",
      "3   155       1   155   1.092000e-96       309  \n",
      "4   155       1   155   1.092000e-96       309  \n",
      "efdf head:\n",
      "           qseqid            sseqid  pident  length  mismatch  gapopen  \\\n",
      "0     NP_006841.1  efXP_027985020.2   0.718     181        51        0   \n",
      "1     NP_006841.1  efXP_027985024.2   0.718     181        51        0   \n",
      "2     NP_758525.1  efXP_027991056.1   0.755     163        39        0   \n",
      "3     NP_060872.1  efXP_008146987.1   0.843     166        26        0   \n",
      "4  NP_001166242.1  efXP_054566753.1   0.692     208        61        0   \n",
      "\n",
      "   qstart  qend  sstart  send        evalue  bitscore  \n",
      "0      26   206       3   183  1.120000e-82       271  \n",
      "1      26   206       3   183  1.120000e-82       271  \n",
      "2       1   160       1   163  3.826000e-81       264  \n",
      "3       4   169       4   169  1.546000e-87       283  \n",
      "4       1   208      30   228  1.143000e-90       296  \n",
      "endf head:\n",
      "           qseqid          sseqid  pident  length  mismatch  gapopen  qstart  \\\n",
      "0     NP_071920.1  enKAK1343172.1   0.688     634       190        0      67   \n",
      "1  NP_001009931.1  enKAK1329028.1   0.524     295       138        0       1   \n",
      "2     NP_002167.1  enKAK1334567.1   0.535     185        85        0       3   \n",
      "3  NP_001121070.1  enKAK1335326.1   0.831     184        30        0      57   \n",
      "4     NP_062427.1  enKAK1329107.1   0.900     101        10        0       1   \n",
      "\n",
      "   qend  sstart  send         evalue  bitscore  \n",
      "0   700       1   610  4.897000e-310       954  \n",
      "1   290       1   295   4.382000e-72       271  \n",
      "2   187    1208  1391   2.064000e-57       196  \n",
      "3   236       1   184   4.534000e-98       316  \n",
      "4   101       1   101   8.772000e-59       195  \n",
      "mbdf head:\n",
      "           qseqid            sseqid  pident  length  mismatch  gapopen  \\\n",
      "0     NP_002954.2  mbXP_005879398.1   0.673     101        33        0   \n",
      "1  XP_005245488.1  mbXP_005879398.1   0.673     101        33        0   \n",
      "2     NP_757351.1  mbXP_005860170.1   0.630     541       190        0   \n",
      "3     NP_000749.2  mbXP_014394659.1   0.692     104        32        0   \n",
      "4     NP_056322.2  mbXP_014399779.1   0.820     344        62        0   \n",
      "\n",
      "   qstart  qend  sstart  send         evalue  bitscore  \n",
      "0       1   101       1   101   1.759000e-45       158  \n",
      "1       1   101       1   101   1.759000e-45       158  \n",
      "2      14   554       8   521  2.863000e-208       653  \n",
      "3      41   144      82   185   6.363000e-44       156  \n",
      "4      25   368       7   349  7.199000e-206       646  \n",
      "mddf head:\n",
      "           qseqid            sseqid  pident  length  mismatch  gapopen  \\\n",
      "0  NP_001138406.1  mdXP_059554725.1   0.954     155         7        0   \n",
      "1  NP_001138406.1  mdXP_059554726.1   0.954     155         7        0   \n",
      "2  NP_001138406.1  mdXP_059554727.1   0.954     155         7        0   \n",
      "3  NP_001138406.1  mdXP_059554728.1   0.954     155         7        0   \n",
      "4  NP_001138406.1  mdXP_059554729.1   0.954     155         7        0   \n",
      "\n",
      "   qstart  qend  sstart  send         evalue  bitscore  \n",
      "0       1   155       1   155  9.263000e-100       318  \n",
      "1       1   155       1   155  9.263000e-100       318  \n",
      "2       1   155       1   155  9.263000e-100       318  \n",
      "3       1   155       1   155  9.263000e-100       318  \n",
      "4       1   155       1   155  9.263000e-100       318  \n",
      "mldf head:\n",
      "           qseqid            sseqid  pident  length  mismatch  gapopen  \\\n",
      "0  NP_001116437.1  mlXP_023615055.1   0.683     783       247        0   \n",
      "1     NP_006262.1  mlXP_014316682.1   0.978      94         2        0   \n",
      "2     NP_006262.1  mlXP_023614987.1   0.978      94         2        0   \n",
      "3     NP_002957.1  mlXP_014316846.1   0.979      97         2        0   \n",
      "4     NP_002957.1  mlXP_014316847.1   0.979      97         2        0   \n",
      "\n",
      "   qstart  qend  sstart  send        evalue  bitscore  \n",
      "0       1   783       1   779  0.000000e+00      1120  \n",
      "1       1    94       1    94  1.893000e-57       192  \n",
      "2       1    94       1    94  1.893000e-57       192  \n",
      "3       1    97       1    97  6.593000e-62       205  \n",
      "4       1    97       1    97  6.593000e-62       205  \n",
      "mmdf head:\n",
      "           qseqid            sseqid  pident  length  mismatch  gapopen  \\\n",
      "0  XP_005273733.1  mmXP_036169643.1   0.965     205         7        0   \n",
      "1     NP_005108.1  mmXP_036179848.1   0.748     211        53        0   \n",
      "2     NP_001997.5  mmXP_036168677.1   0.948     155         8        0   \n",
      "3     NP_062825.1  mmXP_036155956.1   0.952     211        10        0   \n",
      "4     NP_061986.1  mmXP_036163162.1   0.819     198        35        0   \n",
      "\n",
      "   qstart  qend  sstart  send         evalue  bitscore  \n",
      "0       1   205       1   205  1.144000e-131       413  \n",
      "1       6   216       9   219  4.926000e-104       334  \n",
      "2     134   288       1   155   2.051000e-96       316  \n",
      "3       1   211       1   210  2.316000e-135       424  \n",
      "4      12   209      12   204   1.268000e-98       318  \n",
      "modf head:\n",
      "           qseqid            sseqid  pident  length  mismatch  gapopen  \\\n",
      "0     NP_000538.3  moXP_036119159.1   0.703     873       256        0   \n",
      "1     NP_000568.1  moXP_036118723.1   0.822     158        28        0   \n",
      "2     NP_776214.1  moXP_036118722.1   0.801     176        35        0   \n",
      "3     NP_003384.2  moXP_036121579.1   0.977     351         8        0   \n",
      "4  NP_001136016.1  moXP_036136338.1   0.850     200        30        0   \n",
      "\n",
      "   qstart  qend  sstart  send         evalue  bitscore  \n",
      "0       1   873       1   864   0.000000e+00      1292  \n",
      "1       1   158       1   158   5.062000e-88       284  \n",
      "2       1   176       1   176   2.718000e-95       306  \n",
      "3       1   351       3   353  1.124000e-243       744  \n",
      "4       1   200       1   199  1.840000e-102       328  \n",
      "padf head:\n",
      "        qseqid            sseqid  pident  length  mismatch  gapopen  qstart  \\\n",
      "0  NP_004741.1  paXP_024900115.1   0.896     416        41        0       5   \n",
      "1  NP_005614.2  paXP_006925024.1   0.737      99        26        0       1   \n",
      "2  NP_003732.2  paXP_024898109.1   0.895     873        91        0      85   \n",
      "3  NP_002987.1  paXP_024900411.1   0.617     406       152        0       1   \n",
      "4  NP_001556.2  paXP_006907700.2   0.765      98        23        0       1   \n",
      "\n",
      "   qend  sstart  send         evalue  bitscore  \n",
      "0   404      17   432  1.679000e-256       785  \n",
      "1    99       1    99   3.233000e-46       160  \n",
      "2   955      77   949   0.000000e+00      1673  \n",
      "3   397       1   406  4.302000e-153       485  \n",
      "4    98       1    98   3.081000e-46       160  \n",
      "pddf head:\n",
      "           qseqid            sseqid  pident  length  mismatch  gapopen  \\\n",
      "0  NP_001019382.1  pdXP_028358668.1   0.877      98        12        0   \n",
      "1  NP_001019382.1  pdXP_035870926.1   0.877      98        12        0   \n",
      "2  NP_001019382.1  pdXP_035870927.1   0.877      98        12        0   \n",
      "3  NP_001019382.1  pdXP_028358666.2   0.877      98        12        0   \n",
      "4  NP_001019382.1  pdXP_028358667.1   0.877      98        12        0   \n",
      "\n",
      "   qstart  qend  sstart  send        evalue  bitscore  \n",
      "0       1    98       1    98  1.068000e-51       176  \n",
      "1       1    98       1    98  1.068000e-51       176  \n",
      "2       1    98       1    98  1.068000e-51       176  \n",
      "3       1    98       1    98  1.068000e-51       176  \n",
      "4       1    98       1    98  1.068000e-51       176  \n",
      "pgdf head:\n",
      "           qseqid            sseqid  pident  length  mismatch  gapopen  \\\n",
      "0     NP_996890.1  pgXP_039741866.1   0.869      92        12        0   \n",
      "1     NP_002976.2  pgXP_039741870.1   0.813      91        17        0   \n",
      "2     NP_005469.2  pgXP_039694465.1   0.688     135        42        0   \n",
      "3     NP_005469.2  pgXP_039694466.1   0.688     135        42        0   \n",
      "4  NP_001137280.1  pgXP_039722011.1   0.979     247         5        0   \n",
      "\n",
      "   qstart  qend  sstart  send         evalue  bitscore  \n",
      "0       1    92       1    92   1.944000e-51       175  \n",
      "1       1    91       1    91   2.510000e-48       166  \n",
      "2       1   135      24   158   1.015000e-54       187  \n",
      "3       1   135      24   158   1.015000e-54       187  \n",
      "4       1   247       1   247  3.314000e-164       509  \n",
      "phdf head:\n",
      "           qseqid            sseqid  pident  length  mismatch  gapopen  \\\n",
      "0     NP_542184.1  phXP_045689317.1   0.975     201         5        0   \n",
      "1     NP_597716.1  phXP_045706266.1   0.869     407        53        0   \n",
      "2     NP_597716.1  phXP_045706267.1   0.869     407        53        0   \n",
      "3  XP_005266705.1  phXP_045706266.1   0.869     407        53        0   \n",
      "4  XP_005266705.1  phXP_045706267.1   0.869     407        53        0   \n",
      "\n",
      "   qstart  qend  sstart  send         evalue  bitscore  \n",
      "0       1   201       1   201  5.952000e-125       393  \n",
      "1       1   406       1   407  1.687000e-259       793  \n",
      "2       1   406       1   407  1.687000e-259       793  \n",
      "3       1   406       1   407  1.687000e-259       793  \n",
      "4       1   406       1   407  1.687000e-259       793  \n",
      "pkdf head:\n",
      "           qseqid            sseqid  pident  length  mismatch  gapopen  \\\n",
      "0  XP_006710726.1  pkXP_036308781.1   0.867     241        31        0   \n",
      "1     NP_003385.2  pkXP_036314339.1   0.969     389        12        0   \n",
      "2     NP_003385.2  pkXP_036314340.1   0.969     389        12        0   \n",
      "3     NP_003229.1  pkXP_036309385.1   0.966     414        14        0   \n",
      "4     NP_001556.2  pkXP_036269155.1   0.639      97        35        0   \n",
      "\n",
      "   qstart  qend  sstart  send         evalue  bitscore  \n",
      "0       1   241       1   235  7.518000e-137       430  \n",
      "1       1   389       1   389  8.958000e-270       822  \n",
      "2       1   389       1   389  8.958000e-270       822  \n",
      "3       1   414       1   414  1.383000e-277       846  \n",
      "4       1    97       1    97   2.329000e-38       138  \n",
      "pndf head:\n",
      "           qseqid          sseqid  pident  length  mismatch  gapopen  qstart  \\\n",
      "0     NP_005109.2  pnCAK6433660.1   0.766     256        59        0       1   \n",
      "1     NP_003317.1  pnCAK6439553.1   0.657     181        62        0       1   \n",
      "2     NP_001235.1  pnCAK6433662.1   0.777     234        52        0       1   \n",
      "3  XP_006722994.1  pnCAK6445571.1   0.610     252        93        0       1   \n",
      "4  NP_001193673.1  pnCAK6438013.1   0.454     731       386        0       3   \n",
      "\n",
      "   qend  sstart  send         evalue  bitscore  \n",
      "0   251       1   256  6.115000e-129       406  \n",
      "1   181       1   181   2.100000e-73       242  \n",
      "2   234       1   234  4.886000e-123       388  \n",
      "3   252       1   239   2.066000e-81       269  \n",
      "4   733      28   735  3.517000e-197       638  \n",
      "pvdf head:\n",
      "        qseqid            sseqid  pident  length  mismatch  gapopen  qstart  \\\n",
      "0  NP_055310.1  pvXP_011367843.1   0.823     460        81        0       1   \n",
      "1  NP_647475.1  pvXP_011381586.1   0.776     412        91        0       1   \n",
      "2  NP_835228.2  pvXP_023391896.1   0.922     388        30        0       1   \n",
      "3  NP_835228.2  pvXP_023391897.1   0.922     388        30        0       1   \n",
      "4  NP_835228.2  pvXP_011367722.1   0.922     388        30        0       1   \n",
      "\n",
      "   qend  sstart  send         evalue  bitscore  \n",
      "0   460       1   459  1.200000e-262       805  \n",
      "1   406       1   412  7.241000e-215       664  \n",
      "2   388       1   388  6.382000e-251       767  \n",
      "3   388       1   388  6.382000e-251       767  \n",
      "4   388       1   388  6.382000e-251       767  \n",
      "radf head:\n",
      "        qseqid            sseqid  pident  length  mismatch  gapopen  qstart  \\\n",
      "0  NP_005439.2  raXP_016008878.1   0.729     394       106        0       1   \n",
      "1  NP_001192.2  raXP_016011748.2   0.882     476        56        0       1   \n",
      "2  NP_001193.2  raXP_015997634.1   0.975     409        10        0       1   \n",
      "3  NP_001193.2  raXP_036073791.1   0.975     409        10        0       1   \n",
      "4  NP_570911.2  raXP_015997634.1   0.975     409        10        0       1   \n",
      "\n",
      "   qend  sstart  send         evalue  bitscore  \n",
      "0   392       1   394  4.800000e-191       595  \n",
      "1   472       1   476  6.356000e-281       859  \n",
      "2   408       1   409  5.157000e-277       844  \n",
      "3   408       1   409  5.157000e-277       844  \n",
      "4   408       1   409  5.157000e-277       844  \n",
      "rfdf head:\n",
      "           qseqid            sseqid  pident  length  mismatch  gapopen  \\\n",
      "0     NP_996890.1  rfXP_032947688.1   0.880      92        11        0   \n",
      "1     NP_002976.2  rfXP_032946746.1   0.835      91        15        0   \n",
      "2  NP_001020538.2  rfXP_032952374.1   0.942     209        12        0   \n",
      "3  NP_001020539.2  rfXP_032952381.1   0.937     191        12        0   \n",
      "4  NP_001020541.2  rfXP_032952390.1   0.918     147        12        0   \n",
      "\n",
      "   qstart  qend  sstart  send         evalue  bitscore  \n",
      "0       1    92       1    92   8.320000e-52       176  \n",
      "1       1    91       1    91   2.163000e-49       169  \n",
      "2     181   389       1   208  3.932000e-135       433  \n",
      "3     181   371       1   190  2.313000e-123       398  \n",
      "4     181   327       1   146   2.932000e-89       297  \n",
      "sbdf head:\n",
      "           qseqid            sseqid  pident  length  mismatch  gapopen  \\\n",
      "0  NP_001007538.1  sbXP_066110943.1   0.849     332        50        0   \n",
      "1  NP_001007538.1  sbXP_066110944.1   0.849     332        50        0   \n",
      "2  NP_001007538.1  sbXP_066110945.1   0.849     332        50        0   \n",
      "3  NP_001007538.1  sbXP_066110946.1   0.849     332        50        0   \n",
      "4  NP_001007538.1  sbXP_066110947.1   0.849     332        50        0   \n",
      "\n",
      "   qstart  qend  sstart  send         evalue  bitscore  \n",
      "0       1   332       1   332  8.822000e-205       631  \n",
      "1       1   332       1   332  8.822000e-205       631  \n",
      "2       1   332       1   332  8.822000e-205       631  \n",
      "3       1   332       1   332  8.822000e-205       631  \n",
      "4       1   332       1   332  8.822000e-205       631  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import os\n",
    "\n",
    "# Define columns\n",
    "columns = [\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\",  \n",
    "           \"qstart\", \"qend\", \"sstart\", \"send\", \"evalue\", \"bitscore\"]\n",
    "\n",
    "# Dictionary to store DataFrames\n",
    "dfs = {}\n",
    "\n",
    "# Print to debug which files are being processed\n",
    "filepaths = glob.glob(\"secrets/*.tab\")\n",
    "print(\"Found files:\", filepaths)  # Debug statement\n",
    "\n",
    "# Load each .tab file from the 'rbhresults' directory and store it in the dictionary\n",
    "for filepath in filepaths:\n",
    "    # Extract the prefix from the filename (e.g., \"aj\" from \"rbhresults/ajrbh.tab\")\n",
    "    prefix = os.path.basename(filepath)[:2]\n",
    "    # Load the file into a DataFrame\n",
    "    df = pd.read_csv(filepath, sep='\\t', names=columns)\n",
    "    # Store the DataFrame in the dictionary with the prefix as the key\n",
    "    dfs[f\"{prefix}df\"] = df\n",
    "\n",
    "# Print out the keys of the dictionary to verify all DataFrames were loaded\n",
    "print(\"DataFrames loaded:\", list(dfs.keys()))  # Debug statement\n",
    "\n",
    "# Example: Print the head of each DataFrame to verify\n",
    "for name, dataframe in dfs.items():\n",
    "    print(f\"{name} head:\")\n",
    "    print(dataframe.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define DataFrames in the global namespace using the prefix as variable names\n",
    "for prefix, df in dfs.items():\n",
    "    globals()[prefix] = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All bat protein sequences written to rbh_all_bats_secrets.fasta\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "\n",
    "# File paths\n",
    "bat_fasta_file = \"realmergedprots.faa\"  # This is the merged bat proteome file\n",
    "output_file = \"rbh_all_bats_secrets.fasta\"\n",
    "\n",
    "# Load bat sequences with duplicates handled\n",
    "def load_fasta_with_duplicates(fasta_file):\n",
    "    sequences = {}\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        if record.id not in sequences:\n",
    "            sequences[record.id] = record\n",
    "        else:\n",
    "            print(f\"Duplicate ID found: {record.id}, ignoring this entry.\")\n",
    "    return sequences\n",
    "\n",
    "# Load the bat sequences\n",
    "bat_sequences = load_fasta_with_duplicates(bat_fasta_file)\n",
    "\n",
    "# List of DataFrames (replace 'dfs' with your dictionary or variable holding the DataFrames)\n",
    "dataframe_list = [ajdf, drdf, efdf, endf, mbdf, mddf, mldf, mmdf, modf, padf, pddf, pgdf, phdf, pkdf, pndf, pvdf, radf, rfdf, sbdf]\n",
    "\n",
    "# Open the output FASTA file\n",
    "with open(output_file, \"w\") as output_fasta:\n",
    "    # Loop over each DataFrame\n",
    "    for df in dataframe_list:\n",
    "        # Process each sseqid in the DataFrame\n",
    "        for _, row in df.iterrows():\n",
    "            sseqid_full = row[\"sseqid\"]  # Keep the full ID with prefix\n",
    "            \n",
    "            # Check if the sseqid_full is in the bat sequences\n",
    "            if sseqid_full in bat_sequences:\n",
    "                sseq = bat_sequences[sseqid_full].seq\n",
    "                sdesc = bat_sequences[sseqid_full].description\n",
    "                \n",
    "                # Determine the ID source based on the prefix if needed\n",
    "                prefix = sseqid_full[:2]\n",
    "                if prefix == \"pn\" or prefix == 'en':\n",
    "                    id_source = \"GenBank\"\n",
    "                else:\n",
    "                    id_source = \"RefSeq\"\n",
    "                \n",
    "                # Write the bat protein sequence to the output file\n",
    "                output_fasta.write(f\">{sseqid_full} {sdesc} [{id_source}]\\n{sseq}\\n\")\n",
    "            else:\n",
    "                print(f\"SseqID {sseqid_full} not found in bat proteome.\")\n",
    "\n",
    "print(f\"All bat protein sequences written to {output_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     representative                                             member\n",
      "0  ajXP_036981969.1  [ajXP_036981969.1, ajXP_053514474.1, ajXP_0369...\n",
      "1  ajXP_036982243.2                                 [ajXP_036982243.2]\n",
      "2  ajXP_036982257.2               [ajXP_036982257.2, ajXP_036982248.2]\n",
      "3  ajXP_036982516.2                                 [ajXP_036982516.2]\n",
      "4  ajXP_036982712.2                                 [ajXP_036982712.2]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import os\n",
    "rbh_df = pd.read_csv(\"rbhbatsecretsclustershighid.tsv\", sep=\"\\t\", header=None, names=[\"representative\", \"member\"])\n",
    "\n",
    "aggregated_df = rbh_df.groupby(\"representative\")[\"member\"].apply(list).reset_index()\n",
    "\n",
    "print(aggregated_df.head())\n",
    "\n",
    "aggregated_df.to_csv(\"aggregated_SECRETS_clusters.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Silhouette Score based on pident: 0.10992582380010277\n",
      "Silhouette Score based on evalue: 0.07308145066046258\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.metrics import silhouette_score\n",
    "\n",
    "# Load the alignment data\n",
    "alignment_df = pd.read_csv(\"secrets_cluster_aligns.tsv\", sep=\"\\t\", \n",
    "                           names=[\"query\", \"target\", \"evalue\", \"pident\", \"alnlen\", \"qstart\", \"qend\", \"tstart\", \"tend\"])\n",
    "\n",
    "# Load clustering information with representative and members\n",
    "aggregated_df = pd.read_csv(\"aggregated_SECRETS_clusters.tsv\", sep=\"\\t\")\n",
    "\n",
    "# Filter clusters with more than 1 unique protein and at least 2 different species\n",
    "filtered_clusters = []\n",
    "for _, row in aggregated_df.iterrows():\n",
    "    members = row[\"member\"].strip(\"[]\").replace(\"'\", \"\").split(\", \")\n",
    "    unique_prefixes = set(m[:2] for m in members)  # Get unique species prefixes\n",
    "\n",
    "    if len(members) > 1 and len(unique_prefixes) > 1:\n",
    "        filtered_clusters.append({\"representative\": row[\"representative\"], \"members\": members})\n",
    "\n",
    "# Create a combined list of all proteins across clusters\n",
    "all_proteins = {protein for cluster in filtered_clusters for protein in cluster[\"members\"]}\n",
    "protein_index = {protein: idx for idx, protein in enumerate(all_proteins)}\n",
    "num_proteins = len(all_proteins)\n",
    "\n",
    "# Initialize distance matrices for pident and evalue\n",
    "pident_matrix = np.zeros((num_proteins, num_proteins))\n",
    "evalue_matrix = np.zeros((num_proteins, num_proteins))\n",
    "\n",
    "# Populate matrices based on pairwise alignments\n",
    "for _, row in alignment_df.iterrows():\n",
    "    if row[\"query\"] in protein_index and row[\"target\"] in protein_index:\n",
    "        i, j = protein_index[row[\"query\"]], protein_index[row[\"target\"]]\n",
    "        pident_matrix[i, j] = row[\"pident\"] / 100  # Convert to fraction\n",
    "        pident_matrix[j, i] = row[\"pident\"] / 100\n",
    "        evalue_matrix[i, j] = -np.log(row[\"evalue\"] + 1e-300)\n",
    "        evalue_matrix[j, i] = -np.log(row[\"evalue\"] + 1e-300)\n",
    "\n",
    "# Convert similarity to distance for silhouette calculation\n",
    "pident_distance = 1 - pident_matrix\n",
    "evalue_distance = 1 - (evalue_matrix / np.max(evalue_matrix))\n",
    "\n",
    "# Create labels based on cluster membership\n",
    "labels = np.full(num_proteins, -1)  # Initialize with -1 for unassigned\n",
    "for cluster_id, cluster in enumerate(filtered_clusters):\n",
    "    for member in cluster[\"members\"]:\n",
    "        if member in protein_index:\n",
    "            labels[protein_index[member]] = cluster_id  # Assign cluster ID as label\n",
    "\n",
    "# Filter out unassigned proteins\n",
    "valid_indices = labels != -1\n",
    "# Convert similarity to distance for silhouette calculation\n",
    "pident_distance = 1 - pident_matrix\n",
    "evalue_distance = 1 - (evalue_matrix / np.max(evalue_matrix))  # Normalize evalue distances\n",
    "\n",
    "# Set diagonal to zero for both distance matrices\n",
    "np.fill_diagonal(pident_distance, 0)\n",
    "np.fill_diagonal(evalue_distance, 0)\n",
    "\n",
    "# Assign labels for silhouette calculation\n",
    "labels = labels[valid_indices]  # Ensure labels are only for valid indices\n",
    "\n",
    "# Calculate silhouette scores for both pident and evalue distances\n",
    "pident_silhouette_score = silhouette_score(pident_distance, labels, metric=\"precomputed\")\n",
    "evalue_silhouette_score = silhouette_score(evalue_distance, labels, metric=\"precomputed\")\n",
    "\n",
    "# Display results\n",
    "print(\"Silhouette Score based on pident:\", pident_silhouette_score)\n",
    "print(\"Silhouette Score based on evalue:\", evalue_silhouette_score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pident_distance min/max: 0.0 1.0\n",
      "evalue_distance min/max: 0.0 1.0\n",
      "Silhouette Score based on pident: 0.10992582380010277\n",
      "Silhouette Score based on evalue: 0.07308145066046258\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.metrics import silhouette_score\n",
    "\n",
    "# Load the alignment data\n",
    "alignment_df = pd.read_csv(\"secrets_cluster_aligns.tsv\", sep=\"\\t\", \n",
    "                           names=[\"query\", \"target\", \"evalue\", \"pident\", \"alnlen\", \"qstart\", \"qend\", \"tstart\", \"tend\"])\n",
    "\n",
    "# Load clustering information with representative and members\n",
    "aggregated_df = pd.read_csv(\"aggregated_SECRETS_clusters.tsv\", sep=\"\\t\")\n",
    "\n",
    "# Filter clusters with more than 1 unique protein and at least 2 different species\n",
    "filtered_clusters = []\n",
    "for _, row in aggregated_df.iterrows():\n",
    "    members = row[\"member\"].strip(\"[]\").replace(\"'\", \"\").split(\", \")\n",
    "    unique_prefixes = set(m[:2] for m in members)  # Get unique species prefixes\n",
    "\n",
    "    if len(members) > 1 and len(unique_prefixes) > 1:\n",
    "        filtered_clusters.append({\"representative\": row[\"representative\"], \"members\": members})\n",
    "\n",
    "# Create a combined list of all proteins across clusters\n",
    "all_proteins = {protein for cluster in filtered_clusters for protein in cluster[\"members\"]}\n",
    "protein_index = {protein: idx for idx, protein in enumerate(all_proteins)}\n",
    "num_proteins = len(all_proteins)\n",
    "\n",
    "# Initialize distance matrices for pident and evalue\n",
    "pident_matrix = np.zeros((num_proteins, num_proteins))\n",
    "evalue_matrix = np.zeros((num_proteins, num_proteins))\n",
    "\n",
    "# Populate matrices based on pairwise alignments\n",
    "for _, row in alignment_df.iterrows():\n",
    "    if row[\"query\"] in protein_index and row[\"target\"] in protein_index:\n",
    "        i, j = protein_index[row[\"query\"]], protein_index[row[\"target\"]]\n",
    "        pident_matrix[i, j] = row[\"pident\"] / 100  # Convert to fraction\n",
    "        pident_matrix[j, i] = row[\"pident\"] / 100\n",
    "        evalue_matrix[i, j] = -np.log(row[\"evalue\"] + 1e-300)\n",
    "        evalue_matrix[j, i] = -np.log(row[\"evalue\"] + 1e-300)\n",
    "\n",
    "# Symmetry checks\n",
    "assert np.allclose(pident_matrix, pident_matrix.T), \"pident_matrix is not symmetric\"\n",
    "assert np.allclose(evalue_matrix, evalue_matrix.T), \"evalue_matrix is not symmetric\"\n",
    "\n",
    "# Convert similarity to distance for silhouette calculation\n",
    "pident_distance = 1 - pident_matrix\n",
    "evalue_distance = 1 - (evalue_matrix / np.max(evalue_matrix))\n",
    "\n",
    "# Ensure the diagonal is zero\n",
    "np.fill_diagonal(pident_distance, 0)\n",
    "np.fill_diagonal(evalue_distance, 0)\n",
    "\n",
    "# Check distance matrix ranges\n",
    "print(\"pident_distance min/max:\", pident_distance.min(), pident_distance.max())\n",
    "print(\"evalue_distance min/max:\", evalue_distance.min(), evalue_distance.max())\n",
    "\n",
    "# Create labels based on cluster membership\n",
    "labels = np.full(num_proteins, -1)  # Initialize with -1 for unassigned\n",
    "for cluster_id, cluster in enumerate(filtered_clusters):\n",
    "    for member in cluster[\"members\"]:\n",
    "        if member in protein_index:\n",
    "            labels[protein_index[member]] = cluster_id  # Assign cluster ID as label\n",
    "\n",
    "# Filter out unassigned proteins only once\n",
    "valid_indices = labels != -1\n",
    "pident_distance = pident_distance[valid_indices][:, valid_indices]\n",
    "evalue_distance = evalue_distance[valid_indices][:, valid_indices]\n",
    "labels = labels[valid_indices]\n",
    "\n",
    "# Re-check diagonal for zero values\n",
    "np.fill_diagonal(pident_distance, 0)\n",
    "np.fill_diagonal(evalue_distance, 0)\n",
    "\n",
    "# Calculate silhouette scores for both pident and evalue distances\n",
    "pident_silhouette_score = silhouette_score(pident_distance, labels, metric=\"precomputed\")\n",
    "evalue_silhouette_score = silhouette_score(evalue_distance, labels, metric=\"precomputed\")\n",
    "\n",
    "# Display results\n",
    "print(\"Silhouette Score based on pident:\", pident_silhouette_score)\n",
    "print(\"Silhouette Score based on evalue:\", evalue_silhouette_score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overall Silhouette Score based on pident: 0.10992582380010277\n",
      "Overall Silhouette Score based on evalue: 0.07308145066046258\n",
      "\n",
      "Average Silhouette Scores for Each Cluster:\n",
      "     cluster  pident_silhouette  evalue_silhouette\n",
      "0          0           0.200956           0.079232\n",
      "1          1           0.000000           0.000000\n",
      "2          2           0.000000           0.000000\n",
      "3          3           0.000000           0.000000\n",
      "4          4           0.000000           0.000000\n",
      "..       ...                ...                ...\n",
      "892      892           0.078613           0.066680\n",
      "893      893           0.000000           0.000000\n",
      "894      894           0.000000           0.000000\n",
      "895      895           0.038712           0.036874\n",
      "896      896           0.000000           0.000000\n",
      "\n",
      "[897 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.metrics import silhouette_score, silhouette_samples\n",
    "\n",
    "# Load the alignment data\n",
    "alignment_df = pd.read_csv(\"secrets_cluster_aligns.tsv\", sep=\"\\t\", \n",
    "                           names=[\"query\", \"target\", \"evalue\", \"pident\", \"alnlen\", \"qstart\", \"qend\", \"tstart\", \"tend\"])\n",
    "\n",
    "# Load clustering information with representative and members\n",
    "aggregated_df = pd.read_csv(\"aggregated_SECRETS_clusters.tsv\", sep=\"\\t\")\n",
    "\n",
    "# Filter clusters with more than 1 unique protein and at least 2 different species\n",
    "filtered_clusters = []\n",
    "for _, row in aggregated_df.iterrows():\n",
    "    members = row[\"member\"].strip(\"[]\").replace(\"'\", \"\").split(\", \")\n",
    "    unique_prefixes = set(m[:2] for m in members)  # Get unique species prefixes\n",
    "\n",
    "    if len(members) > 1 and len(unique_prefixes) > 1:\n",
    "        filtered_clusters.append({\"representative\": row[\"representative\"], \"members\": members})\n",
    "\n",
    "# Create a combined list of all proteins across clusters\n",
    "all_proteins = {protein for cluster in filtered_clusters for protein in cluster[\"members\"]}\n",
    "protein_index = {protein: idx for idx, protein in enumerate(all_proteins)}\n",
    "num_proteins = len(all_proteins)\n",
    "\n",
    "# Initialize distance matrices for pident and evalue\n",
    "pident_matrix = np.zeros((num_proteins, num_proteins))\n",
    "evalue_matrix = np.zeros((num_proteins, num_proteins))\n",
    "\n",
    "# Populate matrices based on pairwise alignments\n",
    "for _, row in alignment_df.iterrows():\n",
    "    if row[\"query\"] in protein_index and row[\"target\"] in protein_index:\n",
    "        i, j = protein_index[row[\"query\"]], protein_index[row[\"target\"]]\n",
    "        pident_matrix[i, j] = row[\"pident\"] / 100  # Convert to fraction\n",
    "        pident_matrix[j, i] = row[\"pident\"] / 100\n",
    "        evalue_matrix[i, j] = -np.log(row[\"evalue\"] + 1e-300)\n",
    "        evalue_matrix[j, i] = -np.log(row[\"evalue\"] + 1e-300)\n",
    "\n",
    "# Convert similarity to distance for silhouette calculation\n",
    "pident_distance = 1 - pident_matrix\n",
    "evalue_distance = 1 - (evalue_matrix / np.max(evalue_matrix))\n",
    "\n",
    "# Set diagonal to zero for both distance matrices\n",
    "np.fill_diagonal(pident_distance, 0)\n",
    "np.fill_diagonal(evalue_distance, 0)\n",
    "\n",
    "# Create labels based on cluster membership\n",
    "labels = np.full(num_proteins, -1)  # Initialize with -1 for unassigned\n",
    "for cluster_id, cluster in enumerate(filtered_clusters):\n",
    "    for member in cluster[\"members\"]:\n",
    "        if member in protein_index:\n",
    "            labels[protein_index[member]] = cluster_id  # Assign cluster ID as label\n",
    "\n",
    "# Filter out unassigned proteins\n",
    "valid_indices = labels != -1\n",
    "pident_distance = pident_distance[valid_indices][:, valid_indices]\n",
    "evalue_distance = evalue_distance[valid_indices][:, valid_indices]\n",
    "labels = labels[valid_indices]\n",
    "\n",
    "# Calculate silhouette scores for each sample based on pident and evalue distances\n",
    "pident_silhouette_samples = silhouette_samples(pident_distance, labels, metric=\"precomputed\")\n",
    "evalue_silhouette_samples = silhouette_samples(evalue_distance, labels, metric=\"precomputed\")\n",
    "\n",
    "# Create a DataFrame to store individual silhouette scores with cluster labels\n",
    "silhouette_df = pd.DataFrame({\n",
    "    \"protein\": [protein for protein, index in protein_index.items() if valid_indices[index]],\n",
    "    \"cluster\": labels,\n",
    "    \"pident_silhouette\": pident_silhouette_samples,\n",
    "    \"evalue_silhouette\": evalue_silhouette_samples\n",
    "})\n",
    "\n",
    "# Calculate average silhouette score for each cluster\n",
    "cluster_quality = silhouette_df.groupby(\"cluster\").agg({\n",
    "    \"pident_silhouette\": \"mean\",\n",
    "    \"evalue_silhouette\": \"mean\"\n",
    "}).reset_index()\n",
    "\n",
    "# Display results\n",
    "print(\"Overall Silhouette Score based on pident:\", silhouette_score(pident_distance, labels, metric=\"precomputed\"))\n",
    "print(\"Overall Silhouette Score based on evalue:\", silhouette_score(evalue_distance, labels, metric=\"precomputed\"))\n",
    "print(\"\\nAverage Silhouette Scores for Each Cluster:\")\n",
    "print(cluster_quality)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Average Silhouette Scores for Each Cluster:\n",
      "     cluster  pident_silhouette  evalue_silhouette\n",
      "0          0           0.200956           0.079232\n",
      "1          1           0.000000           0.000000\n",
      "2          2           0.000000           0.000000\n",
      "3          3           0.000000           0.000000\n",
      "4          4           0.000000           0.000000\n",
      "..       ...                ...                ...\n",
      "892      892           0.078613           0.066680\n",
      "893      893           0.000000           0.000000\n",
      "894      894           0.000000           0.000000\n",
      "895      895           0.038712           0.036874\n",
      "896      896           0.000000           0.000000\n",
      "\n",
      "[897 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import silhouette_score, silhouette_samples\n",
    "\n",
    "# Ensure diagonal elements are zero for silhouette calculation\n",
    "np.fill_diagonal(pident_distance, 0)\n",
    "np.fill_diagonal(evalue_distance, 0)\n",
    "\n",
    "# Filter out unassigned proteins and keep original protein IDs\n",
    "valid_protein_ids = [protein for protein, index in protein_index.items() if valid_indices[index]]\n",
    "labels = labels[valid_indices]\n",
    "pident_distance = pident_distance[valid_indices][:, valid_indices]\n",
    "evalue_distance = evalue_distance[valid_indices][:, valid_indices]\n",
    "\n",
    "# Calculate silhouette scores for each sample based on pident and evalue distances\n",
    "pident_silhouette_samples = silhouette_samples(pident_distance, labels, metric=\"precomputed\")\n",
    "evalue_silhouette_samples = silhouette_samples(evalue_distance, labels, metric=\"precomputed\")\n",
    "\n",
    "# Create a DataFrame to store protein IDs, cluster labels, and silhouette scores\n",
    "silhouette_df = pd.DataFrame({\n",
    "    \"protein\": valid_protein_ids,  # Proper protein IDs instead of indexes\n",
    "    \"cluster\": labels,\n",
    "    \"pident_silhouette\": pident_silhouette_samples,\n",
    "    \"evalue_silhouette\": evalue_silhouette_samples\n",
    "})\n",
    "\n",
    "# Calculate the average silhouette score for each cluster\n",
    "cluster_quality = silhouette_df.groupby(\"cluster\").agg({\n",
    "    \"pident_silhouette\": \"mean\",\n",
    "    \"evalue_silhouette\": \"mean\"\n",
    "}).reset_index()\n",
    "\n",
    "# Display results\n",
    "print(\"Average Silhouette Scores for Each Cluster:\")\n",
    "print(cluster_quality)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Average Silhouette Scores and Members for Each Cluster:\n",
      "     cluster  pident_silhouette  evalue_silhouette  \\\n",
      "0          0           0.200956           0.079232   \n",
      "1          1           0.000000           0.000000   \n",
      "2          2           0.000000           0.000000   \n",
      "3          3           0.000000           0.000000   \n",
      "4          4           0.000000           0.000000   \n",
      "..       ...                ...                ...   \n",
      "892      892           0.078613           0.066680   \n",
      "893      893           0.000000           0.000000   \n",
      "894      894           0.000000           0.000000   \n",
      "895      895           0.038712           0.036874   \n",
      "896      896           0.000000           0.000000   \n",
      "\n",
      "                                       cluster_members  \n",
      "0    [ajXP_036982891.1, drXP_045058106.2, ajXP_0369...  \n",
      "1    [ajXP_036982894.2, phXP_045690325.1, drXP_0244...  \n",
      "2    [ajXP_036984044.1, ajXP_036984045.1, pdXP_0283...  \n",
      "3    [ajXP_036984048.1, drXP_024428099.1, phXP_0456...  \n",
      "4    [ajXP_036988966.2, pdXP_028387370.1, phXP_0456...  \n",
      "..                                                 ...  \n",
      "892  [sbXP_066114806.1, sbXP_066114806.1, raXP_0159...  \n",
      "893  [sbXP_066118182.1, moXP_036129084.1, sbXP_0661...  \n",
      "894  [sbXP_066128774.1, drXP_053769435.1, moXP_0360...  \n",
      "895  [sbXP_066132642.1, pkXP_036292227.1, pkXP_0362...  \n",
      "896               [sbXP_066136446.1, moXP_036132855.1]  \n",
      "\n",
      "[897 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "# Create a dictionary that maps each cluster ID to the list of members (protein IDs) in that cluster\n",
    "cluster_members_dict = {\n",
    "    cluster_id: cluster[\"members\"]\n",
    "    for cluster_id, cluster in enumerate(filtered_clusters)\n",
    "}\n",
    "\n",
    "# Map each row in silhouette_df to its respective cluster members\n",
    "silhouette_df[\"cluster_members\"] = silhouette_df[\"cluster\"].map(cluster_members_dict)\n",
    "\n",
    "# Group by cluster to get average silhouette scores and list of cluster members\n",
    "cluster_quality = silhouette_df.groupby(\"cluster\").agg({\n",
    "    \"pident_silhouette\": \"mean\",\n",
    "    \"evalue_silhouette\": \"mean\",\n",
    "    \"cluster_members\": \"first\"  # List of members in each cluster\n",
    "}).reset_index()\n",
    "\n",
    "# Display results\n",
    "print(\"Average Silhouette Scores and Members for Each Cluster:\")\n",
    "print(cluster_quality)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Longevity-Associated Clusters with at least 3 Proteins, 3 Species, and Silhouette Score > 0.25 for either measure:\n",
      "     cluster                                            members  cluster_size  \\\n",
      "72        72  [efXP_008150328.2, mmXP_036185700.1, mbXP_0058...             5   \n",
      "82        82  [efXP_008158474.1, efXP_008158474.1, efXP_0081...             4   \n",
      "89        89  [efXP_027997859.1, mmXP_036194648.1, mlXP_0236...             4   \n",
      "90        90  [efXP_027997861.1, mmXP_036194650.1, mlXP_0236...             4   \n",
      "142      142  [mbXP_005860280.1, mbXP_005860280.1, mmXP_0361...             3   \n",
      "144      144  [mbXP_005860933.1, mlXP_006088898.1, mmXP_0362...             4   \n",
      "145      145  [mbXP_005861536.1, mlXP_006101181.1, mdXP_0595...             4   \n",
      "149      149  [mbXP_005865977.1, mmXP_036157647.1, efXP_0081...             5   \n",
      "151      151  [mbXP_005867728.1, mmXP_036198121.1, mbXP_0143...             5   \n",
      "153      153  [mbXP_005870616.1, mlXP_006089124.1, mmXP_0361...             5   \n",
      "160      160  [mbXP_005879218.1, mlXP_006088314.1, mdXP_0595...             3   \n",
      "164      164  [mbXP_005883600.1, mmXP_036175221.1, mdXP_0595...             3   \n",
      "188      188  [mdXP_059516699.1, mbXP_014389663.1, mdXP_0595...             3   \n",
      "197      197  [mdXP_059523946.1, mbXP_005873582.1, mmXP_0362...             3   \n",
      "201      201  [mdXP_059524585.1, mbXP_014402082.1, pkXP_0362...             5   \n",
      "210      210  [mdXP_059534430.1, mlXP_006088686.1, mbXP_0143...             4   \n",
      "224      224  [mdXP_059542043.1, mdXP_059542043.1, mmXP_0362...             4   \n",
      "232      232  [mdXP_059552241.1, mlXP_006107111.3, mbXP_0058...             3   \n",
      "258      258  [mlXP_006098793.1, mbXP_005859566.1, mmXP_0361...             4   \n",
      "266      266  [mlXP_006105219.1, mbXP_005861370.1, mdXP_0595...             4   \n",
      "268      268  [mlXP_006106810.1, mlXP_006106810.1, mmXP_0361...             3   \n",
      "270      270  [mlXP_014302388.1, mmXP_036212554.1, mbXP_0143...             3   \n",
      "293      293  [mlXP_023619702.1, mbXP_005876360.1, efXP_0545...             5   \n",
      "304      304  [mmXP_036161841.1, mbXP_005864605.2, mlXP_0060...             4   \n",
      "308      308  [mmXP_036167549.1, mdXP_059515173.1, mmXP_0361...             4   \n",
      "309      309  [mmXP_036167552.1, mdXP_059515174.1, mbXP_0058...             4   \n",
      "310      310  [mmXP_036167554.1, mlXP_006103070.1, mlXP_0061...             5   \n",
      "312      312  [mmXP_036167709.1, mbXP_014388848.1, mdXP_0595...             4   \n",
      "317      317  [mmXP_036169964.1, mdXP_059551869.1, mbXP_0058...             4   \n",
      "318      318  [mmXP_036170037.1, mdXP_059551317.1, mlXP_0061...             3   \n",
      "324      324  [mmXP_036181688.1, mbXP_005869436.1, mlXP_0060...             4   \n",
      "329      329  [mmXP_036187408.1, mbXP_005880991.1, mlXP_0060...             4   \n",
      "332      332  [mmXP_036192873.1, mdXP_059518463.1, mlXP_0236...             4   \n",
      "335      335  [mmXP_036193433.1, mbXP_005857519.1, mlXP_0236...             5   \n",
      "341      341  [mmXP_036199135.1, mlXP_014302574.1, mdXP_0595...             4   \n",
      "355      355  [mmXP_036211103.1, mmXP_036211104.1, efXP_0280...             7   \n",
      "560      560  [pgXP_039723820.1, pvXP_011365097.1, paXP_0248...             5   \n",
      "\n",
      "     num_species  short_living_percentage  long_living_percentage  \\\n",
      "72             5                     0.00                0.800000   \n",
      "82             3                     0.75                0.000000   \n",
      "89             4                     0.00                0.750000   \n",
      "90             4                     0.00                0.750000   \n",
      "142            3                     0.00                1.000000   \n",
      "144            4                     0.00                1.000000   \n",
      "145            4                     0.00                1.000000   \n",
      "149            5                     0.00                0.800000   \n",
      "151            3                     0.00                1.000000   \n",
      "153            4                     0.00                1.000000   \n",
      "160            3                     0.00                1.000000   \n",
      "164            3                     0.00                1.000000   \n",
      "188            3                     0.00                1.000000   \n",
      "197            3                     0.00                1.000000   \n",
      "201            5                     0.20                0.800000   \n",
      "210            4                     0.00                1.000000   \n",
      "224            4                     0.00                1.000000   \n",
      "232            3                     0.00                1.000000   \n",
      "258            4                     0.00                1.000000   \n",
      "266            4                     0.00                1.000000   \n",
      "268            3                     0.00                1.000000   \n",
      "270            3                     0.00                1.000000   \n",
      "293            5                     0.00                0.800000   \n",
      "304            4                     0.00                1.000000   \n",
      "308            3                     0.00                1.000000   \n",
      "309            4                     0.00                1.000000   \n",
      "310            4                     0.00                1.000000   \n",
      "312            4                     0.00                1.000000   \n",
      "317            4                     0.00                1.000000   \n",
      "318            3                     0.00                1.000000   \n",
      "324            4                     0.00                1.000000   \n",
      "329            4                     0.00                1.000000   \n",
      "332            4                     0.00                1.000000   \n",
      "335            5                     0.00                0.800000   \n",
      "341            4                     0.00                1.000000   \n",
      "355            4                     0.00                0.857143   \n",
      "560            3                     0.80                0.000000   \n",
      "\n",
      "     average_silhouette_pident  average_silhouette_evalue  \n",
      "72                    0.376297                   0.170704  \n",
      "82                    0.461724                   0.108921  \n",
      "89                    0.317000                   0.125195  \n",
      "90                    0.480000                   0.140122  \n",
      "142                   0.636333                   0.126859  \n",
      "144                   0.486167                   0.317017  \n",
      "145                   0.482833                   0.265391  \n",
      "149                   0.285700                   0.217794  \n",
      "151                   0.395300                   0.158338  \n",
      "153                   0.393400                   0.299051  \n",
      "160                   0.316667                   0.070515  \n",
      "164                   0.639333                   0.541102  \n",
      "188                   0.507278                   0.129837  \n",
      "197                   0.647667                   0.128280  \n",
      "201                   0.288300                   0.097113  \n",
      "210                   0.495667                   0.414857  \n",
      "224                   0.498000                   0.110267  \n",
      "232                   0.361117                   0.074959  \n",
      "258                   0.487167                   0.105803  \n",
      "266                   0.319500                   0.315418  \n",
      "268                   0.666667                   0.257415  \n",
      "270                   0.318333                   0.171592  \n",
      "293                   0.286133                   0.147208  \n",
      "304                   0.319333                   0.298072  \n",
      "308                   0.318833                   0.195587  \n",
      "309                   0.480833                   0.294821  \n",
      "310                   0.251237                   0.075902  \n",
      "312                   0.495000                   0.257678  \n",
      "317                   0.485500                   0.157882  \n",
      "318                   0.637667                   0.204047  \n",
      "324                   0.481000                   0.217566  \n",
      "329                   0.319333                   0.193612  \n",
      "332                   0.345414                   0.282064  \n",
      "335                   0.289000                   0.271299  \n",
      "341                   0.490500                   0.500000  \n",
      "355                   0.279947                   0.285714  \n",
      "560                   0.387200                   0.262380  \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Леонид\\AppData\\Local\\Temp\\ipykernel_24236\\1939553501.py:22: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  cluster_longevity = silhouette_df.groupby(\"cluster\").apply(\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Define longevity-associated prefixes\n",
    "short_living_prefixes = {'sb', 'pn', 'pk', 'mo', 'pd', 'pv', 'pa'}\n",
    "average_living_prefixes = {'en', 'ef', 'ph', 'aj', 'ra', 'pg'}\n",
    "long_living_prefixes = {'mb', 'md', 'ml', 'mm', 'dr', 'rf'}\n",
    "\n",
    "# Function to determine the longevity category based on prefix\n",
    "def get_longevity_category(prefix):\n",
    "    if prefix in short_living_prefixes:\n",
    "        return \"short\"\n",
    "    elif prefix in average_living_prefixes:\n",
    "        return \"average\"\n",
    "    elif prefix in long_living_prefixes:\n",
    "        return \"long\"\n",
    "    else:\n",
    "        return \"unknown\"\n",
    "\n",
    "# Add a new column to silhouette_df with the longevity category and species prefix\n",
    "silhouette_df[\"longevity\"] = silhouette_df[\"protein\"].apply(lambda x: get_longevity_category(x[:2]))\n",
    "silhouette_df[\"species_prefix\"] = silhouette_df[\"protein\"].apply(lambda x: x[:2])\n",
    "\n",
    "# Group by cluster and analyze the longevity distribution, species diversity, and silhouette score\n",
    "cluster_longevity = silhouette_df.groupby(\"cluster\").apply(\n",
    "    lambda df: pd.Series({\n",
    "        \"members\": df[\"cluster_members\"].iloc[0],  # List of cluster members\n",
    "        \"cluster_size\": len(df),  # Total number of proteins in the cluster\n",
    "        \"num_species\": df[\"species_prefix\"].nunique(),  # Number of unique species (prefixes)\n",
    "        \"short_living_percentage\": (df[\"longevity\"] == \"short\").mean(),\n",
    "        \"long_living_percentage\": (df[\"longevity\"] == \"long\").mean(),\n",
    "        \"average_silhouette_pident\": df[\"pident_silhouette\"].mean(),\n",
    "        \"average_silhouette_evalue\": df[\"evalue_silhouette\"].mean()\n",
    "    })\n",
    ").reset_index()\n",
    "\n",
    "# Filter clusters to meet the criteria:\n",
    "# - At least 3 proteins in the cluster\n",
    "# - At least 3 different species (prefixes)\n",
    "# - Either at least 75% short-living or long-living members\n",
    "# - At least one of the silhouette scores is higher than 0.25\n",
    "longevity_associated_clusters = cluster_longevity[\n",
    "    ((cluster_longevity[\"short_living_percentage\"] >= 0.75) | \n",
    "     (cluster_longevity[\"long_living_percentage\"] >= 0.75)) & \n",
    "    (cluster_longevity[\"cluster_size\"] >= 3) & \n",
    "    (cluster_longevity[\"num_species\"] >= 3) &\n",
    "    ((cluster_longevity[\"average_silhouette_pident\"] > 0.25) |\n",
    "     (cluster_longevity[\"average_silhouette_evalue\"] > 0.25))\n",
    "]\n",
    "\n",
    "# Display the longevity-associated clusters with silhouette scores\n",
    "print(\"Longevity-Associated Clusters with at least 3 Proteins, 3 Species, and Silhouette Score > 0.25 for either measure:\")\n",
    "print(longevity_associated_clusters)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "longevity_associated_clusters.to_csv(\"longevitysecretsclust.txt\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib qt5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import ast\n",
    "\n",
    "# Define longevity-associated prefixes\n",
    "short_living_prefixes = {'sb', 'pn', 'pk', 'mo', 'pd', 'pv', 'pa'}\n",
    "average_living_prefixes = {'en', 'ef', 'ph', 'aj', 'ra', 'pg'}\n",
    "long_living_prefixes = {'mb', 'md', 'ml', 'mm', 'dr', 'rf'}\n",
    "\n",
    "# Function to determine the longevity category based on prefix\n",
    "def get_longevity_category(prefix):\n",
    "    if prefix in short_living_prefixes:\n",
    "        return \"green\"\n",
    "    elif prefix in average_living_prefixes:\n",
    "        return \"yellow\"\n",
    "    elif prefix in long_living_prefixes:\n",
    "        return \"red\"\n",
    "    else:\n",
    "        return \"gray\"\n",
    "\n",
    "# Load your data\n",
    "alignment_df = pd.read_csv(\"secrets_cluster_aligns.tsv\", sep=\"\\t\", \n",
    "                           names=[\"query\", \"target\", \"evalue\", \"pident\", \"alnlen\", \"qstart\", \"qend\", \"tstart\", \"tend\"])\n",
    "\n",
    "aggregated_df = pd.read_csv(\"aggregated_SECRETS_clusters.tsv\", sep=\"\\t\")\n",
    "\n",
    "# Calculate silhouette scores, filter clusters, and prepare longevity_associated_clusters as in previous code\n",
    "# Assume longevity_associated_clusters is prepared and has the columns \"cluster\", \"members\", and \"average_silhouette_pident\"\n",
    "\n",
    "# Set up the plot with a grid layout\n",
    "plt.figure(figsize=(20, 20))  # Larger figure size for better spacing\n",
    "grid_size = int(np.ceil(np.sqrt(len(longevity_associated_clusters))))  # Determine grid size (e.g., 5x5 for 25 clusters)\n",
    "spacing = 7  # Adjust this value for more space between clusters\n",
    "\n",
    "for idx, (index, cluster) in enumerate(longevity_associated_clusters.iterrows()):\n",
    "    # Convert members column to list if it's in string format\n",
    "    if isinstance(cluster[\"members\"], str):\n",
    "        members = ast.literal_eval(cluster[\"members\"])  # Safely convert string to list\n",
    "    else:\n",
    "        members = cluster[\"members\"]\n",
    "    \n",
    "    silhouette_score = cluster[\"average_silhouette_pident\"]\n",
    "    \n",
    "    # Define grid position for each cluster\n",
    "    row, col = divmod(idx, grid_size)\n",
    "    cluster_position = (col * spacing, -row * spacing)  # Adjust y-axis to go downward for each row\n",
    "\n",
    "    circle_radius = 0.6 + 0.1 * len(members)  # Adjusted radius based on cluster size\n",
    "    circle = plt.Circle(cluster_position, circle_radius, color='black', fill=False, \n",
    "                        linewidth=silhouette_score * 4)  # Scale outline by silhouette score\n",
    "    plt.gca().add_artist(circle)\n",
    "    \n",
    "    # Set the first member as the representative\n",
    "    representative = members[0] if members else None\n",
    "\n",
    "    # Position proteins within the cluster circle\n",
    "    angles = np.linspace(0, 2 * np.pi, len(members), endpoint=False)\n",
    "    for i, protein in enumerate(members):\n",
    "        protein_color = get_longevity_category(protein[:2])  # Color by longevity\n",
    "        x = cluster_position[0] + circle_radius * 0.7 * np.cos(angles[i])\n",
    "        y = cluster_position[1] + circle_radius * 0.7 * np.sin(angles[i])\n",
    "\n",
    "        # Make the representative protein larger\n",
    "        marker_size = 12 if protein == representative else 8\n",
    "        plt.plot(x, y, 'o', color=protein_color, markersize=marker_size)\n",
    "        \n",
    "        # Add labels to each protein\n",
    "        plt.text(x, y, protein, fontsize=6, ha='center', va='center')\n",
    "\n",
    "        # Draw connections between all proteins in the cluster\n",
    "        for j in range(i + 1, len(members)):\n",
    "            x2 = cluster_position[0] + circle_radius * 0.7 * np.cos(angles[j])\n",
    "            y2 = cluster_position[1] + circle_radius * 0.7 * np.sin(angles[j])\n",
    "            plt.plot([x, x2], [y, y2], color=\"gray\", linewidth=1)  # Uniform line width for all connections\n",
    "\n",
    "# Display plot\n",
    "plt.axis('off')\n",
    "plt.title(\"Longevity-Associated Clusters Network (Grid Layout, Only Within-Cluster Connections)\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Леонид\\AppData\\Local\\Temp\\ipykernel_24236\\3806479941.py:41: MatplotlibDeprecationWarning: The get_cmap function was deprecated in Matplotlib 3.7 and will be removed in 3.11. Use ``matplotlib.colormaps[name]`` or ``matplotlib.colormaps.get_cmap()`` or ``pyplot.get_cmap()`` instead.\n",
      "  cmap = cm.get_cmap('viridis')\n"
     ]
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import ast\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "# Define longevity-associated prefixes\n",
    "short_living_prefixes = {'sb', 'pn', 'pk', 'mo', 'pd', 'pv', 'pa'}\n",
    "average_living_prefixes = {'en', 'ef', 'ph', 'aj', 'ra', 'pg'}\n",
    "long_living_prefixes = {'mb', 'md', 'ml', 'mm', 'dr', 'rf'}\n",
    "\n",
    "# Function to determine the longevity category based on prefix\n",
    "def get_longevity_category(prefix):\n",
    "    if prefix in short_living_prefixes:\n",
    "        return \"green\"\n",
    "    elif prefix in average_living_prefixes:\n",
    "        return \"yellow\"\n",
    "    elif prefix in long_living_prefixes:\n",
    "        return \"red\"\n",
    "    else:\n",
    "        return \"gray\"\n",
    "\n",
    "# Load your data\n",
    "alignment_df = pd.read_csv(\"secrets_cluster_aligns.tsv\", sep=\"\\t\", \n",
    "                           names=[\"query\", \"target\", \"evalue\", \"pident\", \"alnlen\", \"qstart\", \"qend\", \"tstart\", \"tend\"])\n",
    "\n",
    "aggregated_df = pd.read_csv(\"aggregated_SECRETS_clusters.tsv\", sep=\"\\t\")\n",
    "\n",
    "# Calculate silhouette scores, filter clusters, and prepare longevity_associated_clusters as in previous code\n",
    "# Assume longevity_associated_clusters is prepared and has the columns \"cluster\", \"members\", and \"average_silhouette_pident\"\n",
    "\n",
    "# Set up the plot with a grid layout\n",
    "fig, ax = plt.subplots(figsize=(20, 20))  # Larger figure size for better spacing\n",
    "grid_size = int(np.ceil(np.sqrt(len(longevity_associated_clusters))))  # Determine grid size (e.g., 5x5 for 25 clusters)\n",
    "spacing = 7  # Adjust this value for more space between clusters\n",
    "\n",
    "# Set up color mapping for silhouette scores\n",
    "norm = mcolors.Normalize(vmin=longevity_associated_clusters[\"average_silhouette_pident\"].min(),\n",
    "                         vmax=longevity_associated_clusters[\"average_silhouette_pident\"].max())\n",
    "cmap = cm.get_cmap('viridis')\n",
    "\n",
    "# Dummy scatter for color bar\n",
    "sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "sm.set_array([])  # This is needed to create a color bar\n",
    "\n",
    "for idx, (index, cluster) in enumerate(longevity_associated_clusters.iterrows()):\n",
    "    # Convert members column to list if it's in string format\n",
    "    if isinstance(cluster[\"members\"], str):\n",
    "        members = ast.literal_eval(cluster[\"members\"])  # Safely convert string to list\n",
    "    else:\n",
    "        members = cluster[\"members\"]\n",
    "    \n",
    "    silhouette_score = cluster[\"average_silhouette_pident\"]\n",
    "    \n",
    "    # Define grid position for each cluster\n",
    "    row, col = divmod(idx, grid_size)\n",
    "    cluster_position = (col * spacing, -row * spacing)  # Adjust y-axis to go downward for each row\n",
    "\n",
    "    circle_radius = 0.6 + 0.1 * len(members)  # Adjusted radius based on cluster size\n",
    "    outline_color = cmap(norm(silhouette_score))  # Map silhouette score to color\n",
    "    circle = plt.Circle(cluster_position, circle_radius, color=outline_color, fill=False, \n",
    "                        linewidth=2 + silhouette_score * 2)  # Outline width based on silhouette score\n",
    "    ax.add_artist(circle)\n",
    "    \n",
    "    # Set the first member as the representative\n",
    "    representative = members[0] if members else None\n",
    "\n",
    "    # Position proteins within the cluster circle\n",
    "    angles = np.linspace(0, 2 * np.pi, len(members), endpoint=False)\n",
    "    for i, protein in enumerate(members):\n",
    "        protein_color = get_longevity_category(protein[:2])  # Color by longevity\n",
    "        x = cluster_position[0] + circle_radius * 0.7 * np.cos(angles[i])\n",
    "        y = cluster_position[1] + circle_radius * 0.7 * np.sin(angles[i])\n",
    "\n",
    "        # Make the representative protein larger\n",
    "        marker_size = 8 #12 if protein == representative else 8\n",
    "        ax.plot(x, y, 'o', color=protein_color, markersize=marker_size)\n",
    "        \n",
    "        # Add labels to each protein\n",
    "        ax.text(x, y, protein, fontsize=6, ha='center', va='center')\n",
    "\n",
    "        # Draw connections between all proteins in the cluster\n",
    "        for j in range(i + 1, len(members)):\n",
    "            x2 = cluster_position[0] + circle_radius * 0.7 * np.cos(angles[j])\n",
    "            y2 = cluster_position[1] + circle_radius * 0.7 * np.sin(angles[j])\n",
    "            ax.plot([x, x2], [y, y2], color=\"gray\", linewidth=1)  # Uniform line width for all connections\n",
    "\n",
    "# Add color bar outside the plot\n",
    "cbar = fig.colorbar(sm, ax=ax, orientation='vertical', fraction=0.02, pad=0.04)\n",
    "cbar.set_label(\"Silhouette Score (pident-based)\", rotation=90)\n",
    "\n",
    "# Disable axis labels and ticks\n",
    "ax.axis('off')\n",
    "\n",
    "plt.title(\"Longevity-Associated Clusters Network (Grid Layout, Only Within-Cluster Connections)\")\n",
    "plt.show()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Леонид\\AppData\\Local\\Temp\\ipykernel_24236\\2385454948.py:40: MatplotlibDeprecationWarning: The get_cmap function was deprecated in Matplotlib 3.7 and will be removed in 3.11. Use ``matplotlib.colormaps[name]`` or ``matplotlib.colormaps.get_cmap()`` or ``pyplot.get_cmap()`` instead.\n",
      "  cmap = cm.get_cmap('viridis')\n"
     ]
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import ast\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "# Define longevity-associated prefixes\n",
    "short_living_prefixes = {'sb', 'pn', 'pk', 'mo', 'pd', 'pv', 'pa'}\n",
    "average_living_prefixes = {'en', 'ef', 'ph', 'aj', 'ra', 'pg'}\n",
    "long_living_prefixes = {'mb', 'md', 'ml', 'mm', 'dr', 'rf'}\n",
    "\n",
    "# Function to determine the longevity category based on prefix\n",
    "def get_longevity_category(prefix):\n",
    "    if prefix in short_living_prefixes:\n",
    "        return \"green\"\n",
    "    elif prefix in average_living_prefixes:\n",
    "        return \"yellow\"\n",
    "    elif prefix in long_living_prefixes:\n",
    "        return \"red\"\n",
    "    else:\n",
    "        return \"gray\"\n",
    "\n",
    "# Load your data\n",
    "alignment_df = pd.read_csv(\"secrets_cluster_aligns.tsv\", sep=\"\\t\", \n",
    "                           names=[\"query\", \"target\", \"evalue\", \"pident\", \"alnlen\", \"qstart\", \"qend\", \"tstart\", \"tend\"])\n",
    "\n",
    "aggregated_df = pd.read_csv(\"aggregated_SECRETS_clusters.tsv\", sep=\"\\t\")\n",
    "\n",
    "# Assume longevity_associated_clusters is prepared and has the columns \"cluster\", \"members\", and \"average_silhouette_pident\"\n",
    "\n",
    "# Set up the plot with a grid layout\n",
    "fig, ax = plt.subplots(figsize=(20, 20))\n",
    "grid_size = int(np.ceil(np.sqrt(len(longevity_associated_clusters))))\n",
    "spacing = 7\n",
    "\n",
    "# Set up color mapping for silhouette scores\n",
    "norm = mcolors.Normalize(vmin=longevity_associated_clusters[\"average_silhouette_pident\"].min(),\n",
    "                         vmax=longevity_associated_clusters[\"average_silhouette_pident\"].max())\n",
    "cmap = cm.get_cmap('viridis')\n",
    "\n",
    "# Dummy scatter for color bar\n",
    "sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "sm.set_array([])\n",
    "\n",
    "for idx, (index, cluster) in enumerate(longevity_associated_clusters.iterrows()):\n",
    "    # Convert members column to list if it's in string format\n",
    "    if isinstance(cluster[\"members\"], str):\n",
    "        members = list(set(ast.literal_eval(cluster[\"members\"])))  # Remove duplicates\n",
    "    else:\n",
    "        members = list(set(cluster[\"members\"]))\n",
    "\n",
    "    silhouette_score = cluster[\"average_silhouette_pident\"]\n",
    "    \n",
    "    # Define grid position for each cluster\n",
    "    row, col = divmod(idx, grid_size)\n",
    "    cluster_position = (col * spacing, -row * spacing)\n",
    "\n",
    "    circle_radius = 0.6 + 0.1 * len(members)\n",
    "    outline_color = cmap(norm(silhouette_score))\n",
    "    circle = plt.Circle(cluster_position, circle_radius, color=outline_color, fill=False, linewidth=2 + silhouette_score * 2)\n",
    "    ax.add_artist(circle)\n",
    "    \n",
    "    # Position proteins within the cluster circle\n",
    "    angles = np.linspace(0, 2 * np.pi, len(members), endpoint=False)\n",
    "    protein_positions = {}\n",
    "\n",
    "    for i, protein in enumerate(members):\n",
    "        protein_color = get_longevity_category(protein[:2])\n",
    "        x = cluster_position[0] + circle_radius * 0.7 * np.cos(angles[i])\n",
    "        y = cluster_position[1] + circle_radius * 0.7 * np.sin(angles[i])\n",
    "        protein_positions[protein] = (x, y)\n",
    "        ax.plot(x, y, 'o', color=protein_color, markersize=8)\n",
    "        ax.text(x, y, protein, fontsize=6, ha='center', va='center')\n",
    "\n",
    "    # Draw connections between all proteins in the cluster with a uniform line width\n",
    "    for i, protein1 in enumerate(members):\n",
    "        for j in range(i + 1, len(members)):\n",
    "            protein2 = members[j]\n",
    "            x1, y1 = protein_positions[protein1]\n",
    "            x2, y2 = protein_positions[protein2]\n",
    "            ax.plot([x1, x2], [y1, y2], color=\"gray\", linewidth=1)  # Uniform line width for all connections\n",
    "\n",
    "# Add color bar outside the plot\n",
    "cbar = fig.colorbar(sm, ax=ax, orientation='vertical', fraction=0.02, pad=0.04)\n",
    "cbar.set_label(\"Silhouette Score (pident-based)\", rotation=90)\n",
    "\n",
    "# Disable axis labels and ticks\n",
    "ax.axis('off')\n",
    "\n",
    "plt.title(\"Longevity-Associated Clusters Network (Grid Layout, Only Within-Cluster Connections)\")\n",
    "plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
